---
title: "metabolomics - intervention, Revision"
author: "TP"
date: "2024-03-24"
output: 
  html_document:
    toc: true
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 0. preperations
## install packages
```{r}
#devtools::install_github(repo = "samuel-marsh/scCustomize", ref = "develop")
```

```{r}

# RAMP-DB #==================
# install.packages("devtools")
# library(devtools)
# install_github("ncats/RAMP-DB")

##====================
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("KEGGREST")

# library(devtools)
# install_github("komorowskilab/metafetcher")
# library(metafetcher)
```

## Load packages
```{r}

library(variancePartition)
library(lme4)
library(emmeans)
library(graphlayouts)
library(ggnewscale)
library(ggforce)
#library(scCustomize)
library(scales)
library(wesanderson)
library(extrafont)
library(ggraph)
library(igraph)
library(tidygraph)
library(metaboliteIDmapping)
library(lattice)
library(ggpattern)
library(reshape2)
library(MKmisc)
library(magrittr)
#library(DT)
library(RaMP)
library(ggExtra)
library("FactoMineR")
library("factoextra")
library(fgsea)
library(enrichR)
library(fgsea)
library(clusterProfiler)
library(httr)
library(ggplot2)
library(dplyr)
library(ggpubr)
library(openxlsx)
library(naniar)
library(rstatix)
library(patchwork)
library(ggsci)
library(RColorBrewer)
library(tidyverse)
library(latrend)
#library(kml)
library(ggrepel)
library(IMIFA)
library(fcros)
library(easyalluvial)
library(ggridges)
library(ggtext)
library(UpSetR)
library(ComplexHeatmap)
library(pheatmap)
library(ggvenn)
```

## colors
```{r}
# col_timepoint
col_timepoint<- brewer.pal(n = 3, name = "Dark2")
names(col_timepoint)<-c("t0", "t1", "t2")

col_diet_diet <- c("#0564ad", "#b91c1c", "#f1c232")
names(col_diet_diet) <- c("rural_diet", "urban_diet", "MBEGE")
```


## Functions
### ... Getpathway_RAMPdb
```{r}
Getpathway_RAMPdb <- function(hmdbinput){
  require(httr)
  require(jsonlite)
  url = "https://rampdb.nih.gov/api/pathways-from-analytes"
  encode = "json"
  output <- list()
  
  payload = list(query_name = "myQuery",
                       analytes = hmdbinput)
        response = POST(url = url,
                        body = payload,
                        encode = encode)
        json = httr::content(response, as = "text")
        
        resultsall = fromJSON(json)
        return(resultsall$data)
}

```

### ... Get_chemical enrichment
```{r}
Getchemenrich_RAMPdb <- function(hmdbinput){
  require(httr)
  require(jsonlite)
  url = "https://rampdb.nih.gov/api/chemical-enrichment"
  encode = "json"
  output <- list()
  
  payload = list(query_name = "myQuery",
                       metabolites = hmdbinput)
        response = POST(url = url,
                        body = payload,
                        encode = encode)
        json = httr::content(response, as = "text")
        
        resultsall = fromJSON(json)
        return(resultsall$data)
}

```

### ... Get_chemical-classes
```{r}
Getchemclass_RAMPdb <- function(hmdbinput){
  require(httr)
  require(jsonlite)
  url = "https://rampdb.nih.gov/api/chemical-classes"
  encode = "json"
  output <- list()
  
  payload = list(query_name = "myQuery",
                       metabolites = hmdbinput)
        response = POST(url = url,
                        body = payload,
                        encode = encode)
        json = httr::content(response, as = "text")
        
        resultsall = fromJSON(json)
        return(resultsall$data)
}

```

### ... Get_fisher_pathway
```{r}
Getfisherpathway_RAMPdb <- function(hmdbinput){
  require(httr)
  require(jsonlite)
  url = "https://rampdb.nih.gov/api/combined-fisher-test"
  encode = "json"
  output <- list()
  
  payload = list(query_name = "myQuery",
                       analytes = hmdbinput)
        response = POST(url = url,
                        body = payload,
                        encode = encode)
        json = httr::content(response, as = "text")
        
        resultsall = fromJSON(json)
        return(resultsall$data)
}

```

### ... Getfoodb
```{r}
# Getpathway_RAMPdb <- function(foodbinput){
#   require(httr)
#   require(jsonlite)
#   url = "https://foodb.ca/api/v1/compoundreport/compound"
#   encode = "json"
#   output <- list()
# 
#   payload = list(query_name = "myQuery",
#                        compound_id = test)
# 
#   updateurl <- paste0(url,test, foodbinput,"&api_key=d9b6cec3eb561db6f920a0797ba49f3e")
# 
#   response = GET(url = updateurl,
#                        accept_json(),
# add_headers(Accept = 'application/json'),
# Authorization = 'Basic [API KEY]')
# 
#         json = httr::content(response, as = "text")
# 
#         resultsall = fromJSON(json)
#         return(resultsall$data)
# }

```

### ... fgsea_plot
```{r}
fgsea_plot <- function(fgseares, 
                       present_metabo_list,
                       background_metabo, 
                       p.adj = 0.05){
  tmp  <- fgseares

topPathways <- tmp[tmp$padj <p.adj,]$pathway

if (length(topPathways)>0){
  print(paste0("sig metabo enriched are ",topPathways))
  
  tmp <- tmp[which(tmp$pathway %in% topPathways),]
  
  plot(plotGseaTable(background_metabo[topPathways], present_metabo_list, tmp, 
              gseaParam=0.5) )
} else {
print("there are no sig metabo enriched")
}

}
```
### ... make df
```{r}
make_df <- function(a = NA, b = NA, z = NA) {
  data.frame(a = unlist(a), b = unlist(b), z = unlist(z))
}
```

### ... scalecolors
```{r}
scaleColors <- function(data = input_scale, # data to use
                        maxvalue = NULL, # value at which the color is fully red / blue
                        pal = "RdBu"
                        ){
  if(is.null(maxvalue)){
    maxvalue <- floor(min(abs(min(data)), max(data)))
  }
  

  
  if(max(data) > abs(min(data))){
    if(ceiling(max(data)) == maxvalue){
      myBreaks <- c(floor(-max(data)), seq(-maxvalue+0.2, maxvalue-0.2, 0.2),  ceiling(max(data)))
    } else{
      myBreaks <- c(floor(-max(data)), seq(-maxvalue, maxvalue, 0.2),  ceiling(max(data)))
    }
    paletteLength <- length(myBreaks)
    myColor <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 11, name = pal)))(paletteLength)
  } else {
    if(-floor(min(data)) == maxvalue){
      myBreaks <- c(floor(min(data)), seq(-maxvalue+0.2, maxvalue-0.2, 0.2),  ceiling(min(data)))
    } else{
      myBreaks <- c(floor(min(data)), seq(-maxvalue, maxvalue, 0.2),  ceiling(abs(min(data))))
    }
    paletteLength <- length(myBreaks)
    myColor <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 11, name = pal)))(paletteLength)
  }
  return(list(breaks = myBreaks, color = myColor))
}
```

### ... PCA
```{r}
plotPCAnew<-         function (pca_input = dds_vst,
                                ntop = "all",
                                xPC = 1, 
                                yPC = 2,
                                point_size=3,
                                anno_colour,
                                scale = FALSE, 
                                anno = sample_annotation, 
                                color,
                                shape = NULL,
                                title="PCA", 
                                label = NULL, 
                                label_subset = NULL, 
                                facet=NULL,
                                outliers = FALSE,
                                SD = 3, 
                                exclude = NULL,
                                outlier.ellipse= NULL,
                                color.ellipse= FALSE,
                                p.outlier=0.001, 
                                Scale = F
){
  
  
  if(!is.data.frame(pca_input) & !is.matrix(pca_input) ){
    vst_matrix <-as.matrix(assay(pca_input))
  }else{
    vst_matrix <- pca_input
  }
  
  
  if (!is.null(exclude)==TRUE){
    print(paste0("samples ",exclude,"are excluded from analysis"))
    #delete from vst_matrix
    vst_matrix<- vst_matrix[,which(!colnames(vst_matrix) %in% exclude)]
    #delete from anno
    anno<-anno[which(!rownames(anno) %in% exclude),]
  }
  
  if(ntop=="all"){
    pca <- prcomp(t(vst_matrix), scale. = Scale) 
  }else if (is.numeric(ntop)){
    # select the ntop genes by variance
    select <- order(rowVars(vst_matrix), decreasing=TRUE)[c(1:ntop)]
    pca <- prcomp(t(vst_matrix[select,]), scale. = Scale)
  } else {
    pca <- prcomp(t(vst_matrix[ntop,]), scale. = Scale)
  }
  
  #calculate explained variance per PC
  explVar <- pca$sdev^2/sum(pca$sdev^2)
  # transform variance to percent
  percentVar <- round(100 * explVar[c(xPC,yPC)], digits=1)
  
  # Define data for plotting  
  pcaData <- data.frame(xPC=pca$x[,xPC], 
                        yPC=pca$x[,yPC], 
                        color = anno[[color]],
                        name= as.character(anno$ID),
                        stringsAsFactors = F)
  
  #plot PCA
  
  if (!is.null(facet)==TRUE){
    pcaData$facet_var<-anno[[facet]]
  } else {
    pcaData<-pcaData
  }
  
  if(is.factor(pcaData$color) || is.character(pcaData$color)|| is.integer(pcaData$color)){
    if(is.null(shape)==TRUE){
      pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC)) +
        geom_point(size =point_size, aes(color= color))
    }else{
      pcaData$shape = anno[[shape]]
      pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC)) +
        geom_point(size =point_size, aes(color=color, shape=shape)) +
        scale_shape_discrete(name=shape)
    }
    
    if(is.null(anno_colour[1])==TRUE){
      pca_plot <- pca_plot + scale_color_discrete(name=color)
    }else{
      pca_plot <- pca_plot + scale_color_manual(values=anno_colour, name=color)
    }
    
  }else if(is.numeric(pcaData$color)){
    if(is.null(shape) == TRUE){
      pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC)) +
        geom_point(size =point_size, aes(color=color)) +
        scale_color_gradientn(colours = bluered(100),name=color)
    }else{
      pcaData$shape = anno[[shape]]
      pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC)) +
        geom_point(size =point_size, aes(colour=color, shape=shape)) +
        scale_color_gradientn(colours = bluered(100),name=color)+
        scale_shape_discrete(name=shape)
    }
  }
  
  # outliers
  if (outliers==TRUE){
    print ("only outliers will be labeled")
    
    #Mahalanobis distance - add this
    pcaData$m_dist<-mahalanobis(pcaData[,c("xPC","yPC")], colMeans(pcaData[,c("xPC","yPC")]), cov(pcaData[,c("xPC","yPC")]))
    pcaData$p <- pchisq(pcaData$m_dist, df=SD, lower.tail=FALSE)
    
    # #PC1 outliers
    # table1<-as.data.frame(pcaData[,xPC])
    # rownames(table1)<-pcaData$name
    # table1$outlier<-NA
    # 
    # PC1_mean<-mean(as.numeric(table1[,1]))
    # PC1_SD<-sd(as.numeric(table1[,1]))
    # PC1_up<-PC1_mean+SD*PC1_SD
    # PC1_dn<-PC1_mean-SD*PC1_SD
    # 
    # table1[,2]<-apply(table1,1,function(z){
    #   if (z[1]>PC1_up | z[1]<PC1_dn){
    #     "outlier"
    #   } else {
    #     ""
    #   } 
    # })
    # 
    # 
    # outliers_PC1<-rownames(table1[which(table1$outlier=="outlier"),])
    # 
    # #PC2 outliers
    # table2<-as.data.frame(pcaData[,yPC])
    # rownames(table2)<-pcaData$name
    # table2$outlier<-NA
    # 
    # PC2_mean<-mean(as.numeric(table2[,1]))
    # PC2_SD<-sd(as.numeric(table2[,1]))
    # PC2_up<-PC2_mean+SD*PC2_SD
    # PC2_dn<-PC2_mean-SD*PC2_SD
    # 
    # table2[,2]<-apply(table2,1,function(z){
    #   if (z[1]>PC2_up | z[1]<PC2_dn){
    #     "outlier"
    #   } else {
    #     ""
    #   } 
    # })
    # 
    # outliers_PC2<-rownames(table2[which(table2$outlier=="outlier"),])
    # 
    # add column label for the pcaData
    pcaData$label <- ifelse(pcaData$p<p.outlier,pcaData$name,"")
    
    #adjust title
    title = paste0(title," outliers labeled at SD=",SD )
    pca_plot <- pca_plot + 
      geom_text_repel(data = pcaData, aes(label = label), nudge_x = 2, nudge_y = 2, colour = "black") 
  }
  
  # adds a label to the plot. To label only specific points, put them in the arument label_subset
  if (outliers == FALSE & !is.null(label) == TRUE){
    pcaData$label <- anno[[label]]
    if(!is.null(label_subset) == TRUE){
      pcaData_labeled <- pcaData[pcaData$label %in% label_subset,]
    } else {
      pcaData_labeled <- pcaData
    }
    pca_plot <- pca_plot + 
      geom_text_repel(data = pcaData_labeled, aes(label = label), nudge_x = 2, nudge_y = 2, colour = "black") 
  }
  
  pca_plot <- pca_plot+
    xlab(paste0("PC ",xPC, ": ", percentVar[1], "% variance")) +
    ylab(paste0("PC ",yPC,": ", percentVar[2], "% variance")) +
    coord_fixed()+
    theme_classic()+        
    theme(aspect.ratio = 1)+
    ggtitle(title)
  
  
  
  #add facet if required
  if (is.null(facet)==TRUE){
    pca_plot<-pca_plot
  } else {
    pca_plot<-pca_plot + facet_wrap(.~facet_var)
  }
  
  #add ellipse if required
  if(!is.null(outlier.ellipse)==TRUE){
    pca_plot<-pca_plot + stat_ellipse(type = "norm",level = outlier.ellipse)
  }
  
  #add groups ellipse
  if (color.ellipse==TRUE){
    pca_plot <- pca_plot + stat_ellipse(type = "norm",aes(color=color))
  }
  
  pca_plot
}

# PCA explained variance plot

PCAexpvar<- function (input = dds_vst, 
                      xlimit = 10,
                      ylimit = 100
                      ){
  
  # Extract the eigenvalues/variances of the principal dimensions
  if(is.matrix(input) | is.data.frame(input)){
    eigenvalue <- get_eig(prcomp(t(as.matrix(input))))
  } else {
    eigenvalue <- get_eig(prcomp(t(assay(input))))
  }
  
  eigenvalue$dim <- as.numeric(c(1:nrow(eigenvalue)))
  
  pp<- ggplot(eigenvalue[c(1:xlimit),], aes(factor(dim), variance.percent, label=paste0(round(variance.percent,2),"%")))+ 
    geom_bar(stat = "identity", fill="gray", color="black")+
    #geom_line(aes(dim, variance.percent)) +
    #geom_point(aes(dim, variance.percent)) +
    #geom_line(aes(dim, cumulative.variance.percent), colour= "grey") + 
    #geom_point(aes(dim, cumulative.variance.percent), colour= "grey") + 
    scale_x_continuous(breaks=c(1:nrow(eigenvalue)))+
    ylab("Percentage of explained variances") +
    theme_bw()+
    scale_x_discrete(name= "Dimensions")+
    ylim(0,ylimit)+
    geom_text(vjust = -1)
  
  pp
  
}
```

#### .. connection ramp
```{r}
#' Set Connection Parameters or RaMP
#'
#' @param dbname the name of the database (by default a string that is database name including all tables
#' @param username a string that is username for database (default: root)
#' @param conpass password for database (string)
#' @param host a string that stand for host
#' @param socket optional, location of mySQL.sock file (useful when running RaMP on remote clusters)
#'
#' @examples
#' \dontrun{
#' pkg.globals <- setConnectionToRaMP(dbname="ramp2",username="root",conpass="",host = "localhost")
#' }
#' @export
setConnectionToRaMP_new <- function(dbname="ramp",
                       username = "root",
                       conpass = "",
                       host ="localhost",
		       socket = ""){
  pkg.globals <- new.env()
  pkg.globals$dbname=dbname
  pkg.globals$username=username
  pkg.globals$conpass=conpass
  pkg.globals$host=host
  if(socket ==""){
      pkg.globals$socket=NULL
  }else{
      pkg.globals$socket = socket
  }
  return(pkg.globals)
}

#' Connect to RaMP database (requires mysql password when running locally)
#'
#' @examples
#' \dontrun{
#' pkg.globals <- setConnectionToRaMP(dbname="ramp2",username="root",conpass="",host = "localhost")
#' con <- connectToRaMP()
#' }
#' @return MySQL connection based on given dbname,username,password, and host
#' @export
connectToRaMP <- function() {
  if(!exists("pkg.globals")) {
	stop("Be sure the run the setConnectionToRaMP() and assign it to pkg.globals");
  }
  if(!is.null(get("socket",pkg.globals))) {
  	con <- RMariaDB::dbConnect(
	    drv = RMariaDB::MariaDB(),
	    dbname = get("dbname",pkg.globals),
	    username = get("username",pkg.globals),
	    password = get("conpass",pkg.globals),
	    host = get("host",pkg.globals),
	    unix.socket = get("socket",pkg.globals)
	)
  } else {
       con <- RMariaDB::dbConnect(
            drv = RMariaDB::MariaDB(),
            dbname = get("dbname",pkg.globals),
            username = get("username",pkg.globals),
            password = get("conpass",pkg.globals),
            host = get("host",pkg.globals)
	)
   }
 return(con)
}
```

### ... mixed model function overall
```{r}
mixed_model_overall_function <- function(anno , 
                     data , 
                     fixed.main = c("timepoint"), 
                     fixed.covariates = c("BMI_t0","age"), # variables to add as co-variates, or NULL
                     random.effect = c("PID"),
                     id.col = "ID", Method= "REML" # OR ML
                      ){
  
  # create list to store results
  tmplist_mix <- list()
  
  ## choose only relevant columns from annodf
  idx <- match(c(id.col, random.effect, fixed.main, fixed.covariates), colnames(anno))
  names(idx) <- c(id.col, random.effect, fixed.main, fixed.covariates)
  
  ## check is some variable names are missing in the annotation table
  if (sum(sapply(idx, is.na)) >0){
    ## find which variables is missing
    which_missing <- names(idx)[is.na(idx)]
    msg <- paste0("The variable ", which_missing, " is missing from the annotation table")
    stop(msg)
  }
         
  tmp_annodf <- anno[,idx]
  
  ## name rownames as id.col
  rownames(tmp_annodf) <- tmp_annodf[[id.col]]
  
  ## match rownames tmp_annodf and data
  tmpdata_df <- data[match(rownames(tmp_annodf), rownames(data)),]
  
  if(identical(rownames(tmpdata_df), rownames(tmp_annodf))){
    print("tables rows are identical")
  } else {
    print("check function")
  }
  
  ## formulat parameters
  
  response_name <- "y"
  
      if (length(fixed.covariates)>0){
      model_formula <- as.formula(paste0(response_name, " ~ ", paste("fixed.main", paste(fixed.covariates, collapse = " + "), sep = " + ")))
      } else {
        model_formula <- as.formula(paste0(response_name, " ~ fixed.main"))
      }
  
  tmplist_mix[["model_results"]] <- apply(tmpdata_df, 2, function(x){
    ## combine x with the anno
    tmpx <- cbind(x, tmp_annodf)
    
    ## leave only complete cases
    #tmpx <- tmpx[complete.cases(tmpx),]
    
    ## change colnames
   
    ## comparisons to be made
    if(length(random.effect) == 1){
      random.effect <- random.effect
    } else {
      random.effect <- random.effect[1]
      print("random.effect had more than 1 parameter, only first one will be used")
    }
    
     # first column for response will be y
    colnames(tmpx)[1] <- response_name
    # random effect column 
    colnames(tmpx)[which(colnames(tmpx) %in% random.effect[1])] <- "random.eff"
    # main fixed effect
    colnames(tmpx)[which(colnames(tmpx) %in% fixed.main[1])] <- "fixed.main"
    
    
    
      tmpx <- as.data.frame(tmpx)

      nlme::lme(fixed =  model_formula , 
                            data =  testdf , random = ~ 1|random.eff, 
                method = Method)
      }, simplify=FALSE, USE.NAMES=TRUE)     
  
  
  ## store in the tmplist_mix also the data frames used for the calculations
  tmplist_mix[["data"]] <- tmpdata_df
  tmplist_mix[["annotation"]] <- tmp_annodf
  tmplist_mix[["parameters"]] <- list("id.col" = id.col, 
                                      "response_name" = response_name, 
                                      "random.effect" = random.effect, 
                                      "fixed.covariates" = fixed.covariates, 
                                      "fixed.main" = fixed.main, 
                                      "model_formulat" = model_formula)
  
  print(model_formula)
  
 return(tmplist_mix) 
}
```



#### ... Mixed model function
```{r}
mixed_model_function <- function(anno , 
                     data , 
                     fixed.main = c("timepoint"), 
                     fixed.covariates = c("BMI_t0","age"), # variables to add as co-variates, or NULL
                     random.effect = c("PID"),
                     id.col = "ID", Method= "ML" # OR REML
                      ){
  
  # create list to store results
  tmplist_mix <- list()
  
  ## choose only relevant columns from annodf
  idx <- match(c(id.col, random.effect, fixed.main, fixed.covariates), colnames(anno))
  names(idx) <- c(id.col, random.effect, fixed.main, fixed.covariates)
  
  ## check is some variable names are missing in the annotation table
  if (sum(sapply(idx, is.na)) >0){
    ## find which variables is missing
    which_missing <- names(idx)[is.na(idx)]
    msg <- paste0("The variable ", which_missing, " is missing from the annotation table")
    stop(msg)
  }
         
  tmp_annodf <- anno[,idx]
  
  ## name rownames as id.col
  rownames(tmp_annodf) <- tmp_annodf[[id.col]]
  
  ## match rownames tmp_annodf and data
  tmpdata_df <- data[match(rownames(tmp_annodf), rownames(data)),]
  
  if(identical(rownames(tmpdata_df), rownames(tmp_annodf))){
    print("tables rows are identical")
  } else {
    print("check function")
  }
  
  ## formulat parameters
  
  response_name <- "y"
  
      if (length(fixed.covariates)>0){
      model_formula <- as.formula(paste0(response_name, " ~ ", paste("fixed.main", paste(fixed.covariates, collapse = " + "), sep = " + ")))
      } else {
        model_formula <- as.formula(paste0(response_name, " ~ fixed.main"))
      }
  
  tmplist_mix[["model_results"]] <- apply(tmpdata_df, 2, function(x){
    ## combine x with the anno
    tmpx <- cbind(x, tmp_annodf)
    
    ## leave only complete cases
    tmpx <- tmpx[complete.cases(tmpx),]
    
    ## change colnames
   
    ## comparisons to be made
    if(length(random.effect) == 1){
      random.effect <- random.effect
    } else {
      random.effect <- random.effect[1]
      print("random.effect had more than 1 parameter, only first one will be used")
    }
    
     # first column for response will be y
    #protein_name <- colnames(tmpx)[1]
    
    colnames(tmpx)[1] <- response_name
    # random effect column 
    colnames(tmpx)[which(colnames(tmpx) %in% random.effect[1])] <- "random.eff"
    # main fixed effect
    colnames(tmpx)[which(colnames(tmpx) %in% fixed.main[1])] <- "fixed.main"
    
    
      if (is.factor(tmpx$fixed.main)){
        
      tmp_mycomparisons <-levels(tmpx$fixed.main)
      tmp_mycomparisons <- as.data.frame(t(combn(tmp_mycomparisons, 2)))
      tmp_mycomparisons <- paste0(tmp_mycomparisons[,1],"_vs_", tmp_mycomparisons[,2])
      
      } else if (is.character(tmpx$fixed.main)){
        
      tmp_mycomparisons <-unique(tmpx$fixed.main)
      tmp_mycomparisons <- as.data.frame(t(combn(tmp_mycomparisons, 2)))
      tmp_mycomparisons <- paste0(tmp_mycomparisons[,1],"_vs_", tmp_mycomparisons[,2])
      
      } else {
        print("random.factor is not chr or factor and cannot be used")
      }
      
    
      tmpx <- as.data.frame(tmpx)
      
      
      
     sapply(tmp_mycomparisons, function(y){
      values_to_use <- str_split_fixed(y, "_vs_", 2)[1,]
      
      testdf <- subset(tmpx, fixed.main %in% values_to_use)
      
      ## take only PIDs with 2 timepoints
      # testdf <- as.data.frame(testdf %>% group_by(random.eff) %>% mutate(n_count = n()))
      # testdf <- subset(testdf, n_count == 2)
      
     
      # Try to fit the model and catch errors or warnings
      result <- tryCatch({
        model <- nlme::lme(fixed =  model_formula , 
                            data =  testdf , random = ~ 1|random.eff, 
                method = Method)
        
        return(model)
        }, error = function(e) {
          #message("Model could not be created for ", protein_name)
          return(NULL)
        }, warning = function(w) {
        return(model)
      })
      
      result
      
      }, simplify=FALSE, USE.NAMES=TRUE)     
    
  })
  
  
  ## store in the tmplist_mix also the data frames used for the calculations
  tmplist_mix[["data"]] <- tmpdata_df
  tmplist_mix[["annotation"]] <- tmp_annodf
  tmplist_mix[["parameters"]] <- list("id.col" = id.col, 
                                      "response_name" = response_name, 
                                      "random.effect" = random.effect, 
                                      "fixed.covariates" = fixed.covariates, 
                                      "fixed.main" = fixed.main, 
                                      "model_formulat" = model_formula)
  
  print(model_formula)
  
 return(tmplist_mix) 
}
```

#### ... Mixed model function - lmerTest
```{r}
mixed_model_function_lmerTest <- function(anno , 
                     data , 
                     fixed.main = c("timepoint"), 
                     fixed.covariates = c("BMI_t0","age"), # variables to add as co-variates, or NULL
                     random.effect = c("PID"),
                     id.col = "ID", REML.= TRUE
                      ){
  
  # create list to store results
  tmplist_mix <- list()
  
  ## choose only relevant columns from annodf
  idx <- match(c(id.col, random.effect, fixed.main, fixed.covariates), colnames(anno))
  names(idx) <- c(id.col, random.effect, fixed.main, fixed.covariates)
  
  ## check is some variable names are missing in the annotation table
  if (sum(sapply(idx, is.na)) >0){
    ## find which variables is missing
    which_missing <- names(idx)[is.na(idx)]
    msg <- paste0("The variable ", which_missing, " is missing from the annotation table")
    stop(msg)
  }
         
  tmp_annodf <- anno[,idx]
  
  ## name rownames as id.col
  rownames(tmp_annodf) <- tmp_annodf[[id.col]]
  
  ## match rownames tmp_annodf and data
  tmpdata_df <- data[match(rownames(tmp_annodf), rownames(data)),]
  
  if(identical(rownames(tmpdata_df), rownames(tmp_annodf))){
    print("tables rows are identical")
  } else {
    print("check function")
  }
  
  ## formulat parameters
  
  response_name <- "y"
  
      if (length(fixed.covariates)>0){
      model_formula <- as.formula(paste0(response_name, " ~ ", paste("fixed.main", paste(fixed.covariates, collapse = " + "), "(1|random.eff)", sep = " + ")))
      } else {
        model_formula <- as.formula(paste0(response_name, " ~ fixed.main"))
      }
  
  tmplist_mix[["model_results"]] <- apply(tmpdata_df, 2, function(x){
    ## combine x with the anno
    tmpx <- cbind(x, tmp_annodf)
    
    ## leave only complete cases
    tmpx <- tmpx[complete.cases(tmpx),]
    
    ## change colnames
   
    ## comparisons to be made
    if(length(random.effect) == 1){
      random.effect <- random.effect
    } else {
      random.effect <- random.effect[1]
      print("random.effect had more than 1 parameter, only first one will be used")
    }
    
     # first column for response will be y
    
    orig_var <- colnames(tmpx)[1]
    
    colnames(tmpx)[1] <- response_name
    # random effect column 
    colnames(tmpx)[which(colnames(tmpx) %in% random.effect[1])] <- "random.eff"
    # main fixed effect
    colnames(tmpx)[which(colnames(tmpx) %in% fixed.main[1])] <- "fixed.main"
    
    
     
      tmpx <- as.data.frame(tmpx)
      
      ## factor fixed.main
      tmpx$fixed.main <- factor(tmpx$fixed.main, levels = unique(tmpx$fixed.main))
      
      
      
      
      
    model <- lmerTest::lmer(data =  tmpx , formula = model_formula, REML = REML.)
        
    # Check for singular fit
    if (lme4::isSingular(model, tol = 1e-4)) {
      warning("Singular fit detected for ", orig_var)
      return(NULL)
    } else {
      return(model)
    }


    
  })
  
  
  ## store in the tmplist_mix also the data frames used for the calculations
  tmplist_mix[["data"]] <- tmpdata_df
  tmplist_mix[["annotation"]] <- tmp_annodf
  tmplist_mix[["parameters"]] <- list("id.col" = id.col, 
                                      "response_name" = response_name, 
                                      "random.effect" = random.effect, 
                                      "fixed.covariates" = fixed.covariates, 
                                      "fixed.main" = fixed.main, 
                                      "model_formulat" = model_formula)
  
  print(model_formula)
  
 return(tmplist_mix) 
}
```



#### ... Summary function mixed.model
```{r}
summary_table_mixed.mod_function <- function(input ,
                                    response_name = "protein",
                                    p.adj.method = "BH", 
                                    p.adj_threshold = 0.05, 
                                    fc_threshold = 1.2,
                                    p.adj.level = "response", # or "comparison" or "all"
                                    orig.data.log = TRUE # if original data is log-scaled TRUE, if not FALSE, if it's a mixed - you cannot rely on the fold change claculations and have to do it separetly
){
  
  ## create empty list to store results
  list_summary_table_res <- list()
  
  # 1 - create summary table -----------------------------------------------------------
  ## perform anova to get sig
  tmp_results <- lapply(input[["model_results"]], function(x){
  lapply(x, function(y) {
    if(is.null(y)){
     data.frame(p = NA, direction = NA) 
    } else{
    tmpdf <- as.data.frame(y[["coefficients"]][["fixed"]])
    data.frame(p = anova(y)["fixed.main","p-value"], 
                    direction = ifelse(sign(tmpdf[grepl("fixed.main", rownames(tmpdf)),1])>0,"up", "down"))
    }
  })
}) 
  
  # rbind the data frames within each comparison
  comb_results <- lapply(tmp_results, function(res){
    do.call(rbind, Map(cbind, comparison = names(res), res))
  })
  
  comb_results <- do.call(rbind, Map(cbind, variable = names(comb_results), comb_results))
  
  
  # remove parameters that were not tested (i.e. have NA in their p-value col and direction) and store their identity in a seperate data frame
  
  na_pars <- as.character(comb_results[is.na(comb_results$p),]$variable)
  
  comb_results <- comb_results[!is.na(comb_results$p),]
  
  # adjust-pvalue
  if(p.adj.level == "response"){
    comb_results <- as.data.frame(comb_results %>% group_by(variable) %>% adjust_pvalue(., p.col = "p", method = p.adj.method) %>% add_significance())
  } else if (p.adj.level == "comparison"){
    comb_results <- as.data.frame(comb_results %>% group_by(comparison) %>% adjust_pvalue(., p.col = "p", method = p.adj.method) %>% add_significance())
  } else if(p.adj.level == "all"){
    comb_results <- as.data.frame(comb_results %>% adjust_pvalue(., p.col = "p", method = p.adj.method) %>% add_significance())
  } else {
    print("No proper level for p.adj was selected, so no adj was made")
  }
  
  
  
  # create merged column to help match the FC results in the next section
  comb_results$merged <- paste0(comb_results$variable, "_", comb_results$comparison)
  
  # 2 - calculate and add log2FC and log2(paired-FC) to summary table -----------
  
  tmpdata_df <- t(input[["data"]])
  tmpanno_df <- input[["annotation"]]
  
  fixed.main <- input[["parameters"]][["fixed.main"]]
  id.col <- input[["parameters"]][["id.col"]]
  
  
  # create data from to store results
  all_fc_res <- data.frame()
  
  for (i in unique(comb_results$comparison)){
    comp <- str_split_fixed(i, pattern =  "_vs_", n= 2)[1,]
  
  cont_ids <- as.character(unique(tmpanno_df[which(tmpanno_df[[fixed.main]] == comp[1]), id.col]))
  test_ids <- as.character(unique(tmpanno_df[which(tmpanno_df[[fixed.main]] == comp[2]), id.col]))
  
  
  tmpdata_df_subset <- tmpdata_df[,match(c(cont_ids, test_ids), colnames(tmpdata_df))]
  
  log2option <- ifelse(orig.data.log == TRUE, 0, 1)
  
    fc_res <-  fcros::fcrosFCmat(tmpdata_df_subset, cont_ids, test_ids, log2.opt = log2option)
    
    tmpdf_fc_res <- data.frame(variable.fc = fc_res$idnames, FC = fc_res$FC, FC2 = fc_res$FC2, comparison.fc = i)
    tmpdf_fc_res$merged <- paste0(tmpdf_fc_res$variable.fc,"_", tmpdf_fc_res$comparison.fc)
    
    tmpdf_fc_res$log2FC <- log2(tmpdf_fc_res$FC)
    tmpdf_fc_res$log2FC2 <- log2(tmpdf_fc_res$FC2)
    
    all_fc_res <- rbind(all_fc_res, tmpdf_fc_res)
    
  }
  
    # add the FC results to the summarytable
    
    comb_results <- merge(comb_results , all_fc_res, by = "merged", all.x = T) 

   
    
  # 3 - determine regulation ----------------------------------------------------
    
    # regulation
  comb_results$regulation <- ifelse(comb_results$p.adj <= p.adj_threshold & abs(comb_results$log2FC) >= log2(fc_threshold), comb_results$direction, "ns")
        
  
  # 4 - create contrasts count ---------------------------------------------------
  
  contrast.count <- comb_results %>%
                              group_by(comparison) %>% 
                              summarize(contrast.count = sum(regulation != "ns" ))

  
  contrast.count <- as.data.frame(contrast.count)  

contrast.count$contrast.count <- paste0(as.character(contrast.count$comparison ),"\n[",contrast.count$contrast.count,"]" )

#contrast.count <- tibble::deframe(contrast.count)

print(contrast.count)

comb_results$x_label <- as.character(contrast.count[match(comb_results$comparison, contrast.count$comparison),"contrast.count"])

  # 5 - store results in a list and return at the end of the function -------------
  
 # rename column of variable to the reponse_name
  colnames(comb_results)[which(colnames(comb_results) == "variable")] <- as.character(response_name)

  list_summary_table_res[["summary_table"]] <- comb_results
  list_summary_table_res[["contrast_count"]] <- contrast.count
  list_summary_table_res[["parameters_not_modelled"]] <- na_pars
  
  
  return(list_summary_table_res)

}
```


### ... summary lmerTest mixed model
```{r}
summary_mixedmod_lmerTest <- function(mixmod_input ,
                            response_name = "protein",
                            p.adj.method.lmer = "none",
                            p.adj.method.overall = "BH",  
                            p.adj_threshold = 0.05, 
                            fc_threshold = 1.2,
                            p.adj.level = "response", # or "comparison" or "all"
                            orig.data.log = TRUE # if original data is log-scaled TRUE, if not FALSE, if it's a mixed - you cannot rely on the fold change claculations and have to do it separetly
){
  
  ## create empty list to store results
  list_summary_table_res <- list()
  
  # 1 - create summary table -----------------------------------------------------------
  ## use lmerTest  to get p-values 
  
  
  tmp_results <- lapply(mixmod_input$model_results, function(x){
  if(!is.null(x)){
   tmp <- as.data.frame(ls_means(x, which = "fixed.main", pairwise = TRUE, adjust = p.adj.method.lmer))
   colnames(tmp)[ncol(tmp)] <- c("p")
   tmp
  }
  })
  
  # remove parameters that were not tested (i.e. have NA in their p-value col and direction) and store their identity in a seperate data frame
  
  na_pars <- unlist(lapply(tmp_results, is.null))
  na_pars <- names(na_pars[which(na_pars == TRUE)])
  tmp_results <- tmp_results[which(!names(tmp_results) %in% na_pars)]
  
  
  comb_results <- do.call(rbind, Map(cbind, variable = names(tmp_results), tmp_results))

  
  
  
  comb_results$contrast <- gsub(" - ","_vs_", comb_results$levels)
  
  ## determine direction
  comb_results$direction <- ifelse(comb_results[["t value"]]<0, "down","up")
  
  ## create merged variable
  comb_results$merged <- paste0(comb_results$variable,"_", comb_results$contrast)
  
  p.col <- "p.adj"
  # adjust-pvalue
  if(p.adj.level == "response"){
    comb_results <- as.data.frame(comb_results %>% group_by(variable) %>% adjust_pvalue(., p.col = "p", method = p.adj.method.overall) %>% add_significance())
  } else if (p.adj.level == "comparison"){
    comb_results <- as.data.frame(comb_results %>% group_by(contrast) %>% adjust_pvalue(., p.col = "p", method = p.adj.method.overall) %>% add_significance())
  } else if(p.adj.level == "all"){
    comb_results <- as.data.frame(comb_results %>% adjust_pvalue(., p.col = "p", method = p.adj.method.overall) %>% add_significance())
  } else if (p.adj.level == "none") {
    comb_results <- comb_results
    p.col <- "p"
    }else{
    print("No proper level for p.adj was selected, so no adj was made")
  }
  
  
  
  
  # 2 - calculate and add log2FC and log2(paired-FC) to summary table -----------
  
  tmpdata_df <- t(mixmod_input[["data"]])
  tmpanno_df <- mixmod_input[["annotation"]]
  
  fixed.main <- mixmod_input[["parameters"]][["fixed.main"]]
  id.col <- mixmod_input[["parameters"]][["id.col"]]
  
  
  # create data from to store results
  all_fc_res <- data.frame()
  
  for (i in unique(comb_results$contrast)){
    comp <- str_split_fixed(i, pattern =  "_vs_", n= 2)[1,]
  
  cont_ids <- as.character(unique(tmpanno_df[which(tmpanno_df[[fixed.main]] == comp[1]), id.col]))
  test_ids <- as.character(unique(tmpanno_df[which(tmpanno_df[[fixed.main]] == comp[2]), id.col]))
  
  
  tmpdata_df_subset <- tmpdata_df[,match(c(cont_ids, test_ids), colnames(tmpdata_df))]
  
  log2option <- ifelse(orig.data.log == TRUE, 0, 1)
  
    fc_res <-  fcros::fcrosFCmat(tmpdata_df_subset, cont_ids, test_ids, log2.opt = log2option)
    
    tmpdf_fc_res <- data.frame(variable.fc = fc_res$idnames, FC = fc_res$FC, FC2 = fc_res$FC2, comparison.fc = i)
    tmpdf_fc_res$merged <- paste0(tmpdf_fc_res$variable.fc,"_", tmpdf_fc_res$comparison.fc)
    
    tmpdf_fc_res$log2FC <- log2(tmpdf_fc_res$FC)
    tmpdf_fc_res$log2FC2 <- log2(tmpdf_fc_res$FC2)
    
    all_fc_res <- rbind(all_fc_res, tmpdf_fc_res)
    
  }
  
    # add the FC results to the summarytable
    
    comb_results <- merge(comb_results , all_fc_res, by = "merged", all.x = T) 

  
    
  # 3 - determine regulation ----------------------------------------------------
    
    # regulation
  comb_results$regulation <- ifelse(comb_results[[p.col]] <= p.adj_threshold & abs(comb_results$log2FC) >= log2(fc_threshold), comb_results$direction, "ns")
        
  
  # 4 - create contrasts count ---------------------------------------------------
  
  contrast.count <- comb_results %>%
                              group_by(contrast) %>% 
                              summarize(contrast.count = sum(regulation != "ns" ))

  
  contrast.count <- as.data.frame(contrast.count)  

contrast.count$contrast.count <- paste0(as.character(contrast.count$contrast ),"\n[",contrast.count$contrast.count,"]" )

#contrast.count <- tibble::deframe(contrast.count)

print(contrast.count)

comb_results$x_label <- as.character(contrast.count[match(comb_results$contrast, contrast.count$contrast),"contrast.count"])

  # 5 - store results in a list and return at the end of the function -------------
  
 # rename column of variable to the reponse_name
  colnames(comb_results)[which(colnames(comb_results) == "variable")] <- as.character(response_name)

  list_summary_table_res[["summary_table"]] <- comb_results
  list_summary_table_res[["contrast_count"]] <- contrast.count
  list_summary_table_res[["parameters_not_modelled"]] <- na_pars
  
  return(list_summary_table_res)

  
}

```





# --------------------------------------------
# 1. Load data
## ** metabolomics
```{r}
metabo <- read.delim("./240323_metabolomics_ionmatrix_withannotation_avg.txt", header = T)
rownames(metabo) <- metabo$ionIdx
head(metabo)
```

```{r}
metabo_mtx <- metabo[,c(7:ncol(metabo))]
rownames(metabo_mtx) <- metabo$ionMz
rownames(metabo_mtx)[1:10]
all(sapply(metabo_mtx, is.numeric))
```
```{r}
head(metabo_mtx)
```

```{r}
dim(metabo_mtx)
```

```{r}
metabo_anno <- metabo[,c(1:6)]
metabo_anno$ion_name <- paste0("ion_",metabo_anno$ionIdx)
head(metabo_anno)
```

## Godfrey annotation
```{r}
sampls_info <- read.delim("../Metadata/240320_data_godfrey.txt", header = T)

colnames(sampls_info)
head(sampls_info)
```


## ion annotation
```{r}
ion_anno <- read.delim("./240324_ion_anno.txt", header = T)

head(ion_anno)
```
### add database
```{r}
ion_anno$database <- "KEGG"
ion_anno[grepl("CHEBI", ion_anno$CompoundID),"database"] <- "ChEBI"
ion_anno[grepl("HMDB", ion_anno$CompoundID),"database"] <- "HMDB"
```

### add source annotation
```{r}
## how many samples are from which database
count_ids_perdatabase <- ion_anno %>%
                          group_by(database) %>%
                          summarize(count.idx = n_distinct(ionMz))

count_ids_perdatabase <- as.data.frame(count_ids_perdatabase)

ggplot(count_ids_perdatabase, aes(x= database, y= count.idx))+
  geom_bar(stat = "identity")+
  geom_text(aes(label = count.idx), color = "white", vjust = 1)+
  theme(aspect.ratio = 1.5)
```
```{r}
list_database <- split(
  ion_anno$ionMz, 
  ion_anno$database
)

UpSetR::upset(fromList(list_database))
```
*--> unique peaks for each database will remain with this database*
*--> peaks that are shared between ChEBI and KEGG or HMDB --> will be annotated by KEGG or HMDB*
*--> the rest - which are shared with HMDB will be annotated by HMDB*

```{r}
database_mtx <- as.data.frame( ComplexHeatmap::list_to_matrix(list_database))

head(database_mtx)
```

```{r}
database_mtx[database_mtx$ChEBI == 1,"ChEBI"] <- "ChEBI"
database_mtx[database_mtx$ChEBI == 0,"ChEBI"] <- ""

database_mtx[database_mtx$HMDB == 1,"HMDB"] <- "HMDB"
database_mtx[database_mtx$HMDB == 0,"HMDB"] <- ""

database_mtx[database_mtx$KEGG == 1,"KEGG"] <- "KEGG"
database_mtx[database_mtx$KEGG == 0,"KEGG"] <- ""

head(database_mtx)

```

```{r}
database_mtx$intersect <- apply(database_mtx,1,paste, collapse = "")
database_mtx$ionMz <- rownames(database_mtx)

head(database_mtx)
```

### Annoate by HMDB
```{r}
annotate_list <- split(
  database_mtx$ionMz,
  database_mtx$intersect
)

str(annotate_list)
```

```{r}
ion_annotation_tal <- list()

## 1. chebi
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$ChEBI,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(row_number()==1)

tmp$annotation <- "ChEBI"

ion_annotation_tal[["ChEBI"]] <- tmp

# ==========================================

## 2. KEGG
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$KEGG,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(row_number()==1)

tmp$annotation <- "KEGG"

ion_annotation_tal[["KEGG"]] <- tmp

# ==========================================

## 3. HMDB
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$HMDB,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(row_number()==1)

tmp$annotation <- "HMDB"

ion_annotation_tal[["HMDB"]] <- tmp

# ==========================================

## 4. ChEBIKEGG
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$ChEBIKEGG,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(database == "KEGG") %>%
  filter(row_number()==1)

tmp$annotation <- "ChEBIKEGG"

ion_annotation_tal[["ChEBIKEGG"]] <- tmp

# ==========================================

## 5. ChEBIHMDB
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$ChEBIHMDB,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(database == "HMDB") %>%
  filter(row_number()==1)

tmp$annotation <- "ChEBIHMDB"

ion_annotation_tal[["ChEBIHMDB"]] <- tmp

# ===========================================

## 6. HMDBKEGG
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$HMDBKEGG,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(database == "HMDB") %>%
  filter(row_number()==1)

tmp$annotation <- "HMDBKEGG"

ion_annotation_tal[["HMDBKEGG"]] <- tmp

# ============================================

## 7. ChEBIHMDBKEGG
tmp <- NULL
tmp <- ion_anno[ion_anno$ionMz %in% annotate_list$ChEBIHMDBKEGG,]

tmp <- tmp %>%
  group_by(ionMz) %>%
  filter(database == "HMDB") %>%
  filter(row_number()==1)

tmp$annotation <- "ChEBIHMDBKEGG"

ion_annotation_tal[["ChEBIHMDBKEGG"]] <- tmp
```

```{r}
melt_annotation_tal <- bind_rows(ion_annotation_tal)
melt_annotation_tal <- as.data.frame(melt_annotation_tal)
rownames(melt_annotation_tal) <- melt_annotation_tal$ionMz
## add ionIndx 
melt_annotation_tal <- melt_annotation_tal[order(melt_annotation_tal$ionMz, decreasing = F),]
melt_annotation_tal$ionInx <- c(1:nrow(melt_annotation_tal))

## as.data.frame
melt_annotation_tal <- as.data.frame(melt_annotation_tal)

nrow(melt_annotation_tal) == length(unique(ion_anno$ionMz))
```

```{r}
ggplot(melt_annotation_tal, aes(x= database))+
  geom_bar(stat = "count")+
  geom_text(vjust =1, color = "white", aes(label = after_stat(count)), stat = "count")
```

### ... use manual converstion to HMDB from metaboanalyst
was done on 27.3.24
* CHEBI annotations had no results
* KEGG annotations were partical
```{r}
kegg_to_hmdb <- read.csv("./ref_metabolites_tal_HMDB_strat/240327_keggtohmdb_metaboanalyst.csv")
colnames(kegg_to_hmdb)[1] <- "kegg_id"


```
```{r}
nrow(kegg_to_hmdb[!is.na(kegg_to_hmdb$HMDB),])
melt_annotation_tal$kegg_to_hmdb <- kegg_to_hmdb[match(melt_annotation_tal$CompoundID, kegg_to_hmdb$kegg_id),"HMDB"]
```


which KEGG ids were not found at all? 
these will be removed from the metabolites 
```{r}
kegg_miss <- kegg_to_hmdb[is.na(kegg_to_hmdb$KEGG),"kegg_id"]
length(kegg_miss)
kegg_miss[1:10]
```
```{r}
melt_annotation_tal$kegg_miss <- NA
melt_annotation_tal[which(melt_annotation_tal$CompoundID %in% kegg_miss),"kegg_miss"] <- "missing"
```

since all CHEBI were not found in the database

### ... NEW anno: compound ID kegg to hmdb and exclude CHEBI
```{r}
## exclude chebi
melt_annotation_tal_new <- melt_annotation_tal[which(melt_annotation_tal$database != "ChEBI"),]

melt_annotation_tal_new <- melt_annotation_tal_new[is.na(melt_annotation_tal_new$kegg_miss),]

melt_annotation_tal_new$CompoundID_new <- ifelse(
  melt_annotation_tal_new$database == "HMDB", melt_annotation_tal_new$CompoundID,
  ifelse(melt_annotation_tal_new$database == "KEGG", melt_annotation_tal_new$kegg_to_hmdb,NA)
)

## exclude those with no new CompoundID
melt_annotation_tal_new <- melt_annotation_tal_new[!is.na(melt_annotation_tal_new$CompoundID_new),]

melt_annotation_tal_new$database_new <- ifelse(grepl("HMDB", melt_annotation_tal_new$CompoundID_new), "HMDB", "error")

head(melt_annotation_tal_new)
```

```{r}
ggplot(melt_annotation_tal_new, aes(x= database_new))+
  geom_bar(stat = "count")+
  geom_text(vjust =-1, color = "red", aes(label = after_stat(count)), stat = "count")
```
## New metabo_mtx 
```{r}
metabo_mtx_full <- metabo[,c(7:ncol(metabo))]
rownames(metabo_mtx_full) <- metabo$ionMz
head(metabo_mtx_full)
```

```{r}
metabo_mtx <- metabo_mtx_full[which(rownames(metabo_mtx_full) %in% as.character(melt_annotation_tal_new$ionMz)),]


head(metabo_mtx)
```
```{r}
all(as.character(melt_annotation_tal_new$ionMz) %in% rownames(metabo_mtx))
```

## ** Import enrichment sets

## ... import the sets
```{r}
# temp = paste0("./SMDP/smpdb_metabolites.csv/",list.files(path = "./SMDP/smpdb_metabolites.csv/", pattern="\\.csv$"))
# myfiles = lapply(temp, read.csv)
# 
# length(myfiles)
```
```{r}
#saveRDS(myfiles,"./SMDP/smpdb_metabolites.rds")
```

```{r}
myfiles <- readRDS("./SMDP/smpdb_metabolites.rds")
```

##.. filter the HMDB ids and name
```{r}
smdp_list <- lapply(myfiles, function(x){
  as.character(x[["HMDB.ID"]])
})

names(smdp_list) <- lapply(myfiles, function(x){
  unique(as.character(x[["Pathway.Name"]]))
})


```

```{r}
metabolic <- unlist(lapply(myfiles, function(x){
  unique(x[["Pathway.Subject"]]) %in% c("Metabolic","Signaling","Protein","Physiological")
}))

smdp_list_metabolic <- smdp_list[metabolic]

length(smdp_list_metabolic)
names(smdp_list_metabolic)[1:10]
```
### .. fischer enrichment  universe
```{r}
universe <- melt_annotation_tal_new$CompoundID_new
length(universe)
universe[1:10]
all(grepl("HMDB",universe))
universe <- universe[which(grepl("HMDB", universe))]
all(grepl("HMDB",universe))
```
## ** Import Pathbank
```{r}
pathbkank <- read_csv("./pathbank/pathbank_all_metabolites.csv/pathbank_all_metabolites.csv")


```


```{r}
colnames(pathbkank) <- gsub(" ","_",colnames(pathbkank))
head(pathbkank)
```
### intersect food metabo
import the converstion to other ids 
```{r}
interesect_metabo_ids <- read_csv("./pathbank/240401_HMDB_name_map_foodmetabo.csv")
```
```{r}
head(interesect_metabo_ids)
```
```{r}
dim(interesect_metabo_ids)
```

## ** Metadata new
```{r}
metadata_new <- read.csv("../Metadata/Sneha/2024-07-12_sample_table_Tzintervention.csv", row.names = 1)
rownames(metadata_new) <- metadata_new$PID
head(metadata_new)
```
```{r}
vis_miss(metadata_new)
```
```{r fig.height=10, fig.width=5}
gg_miss_var(metadata_new)
```
#### .. add columns 
```{r}
unique(metadata_new$Type_Intervention)
```


```{r}
unique(metadata_new$Type_Intervention)
```
diet
```{r}
metadata_new$diet <- ifelse(metadata_new$Type_Intervention == "Banana beer","MBEGE",
                            ifelse(metadata_new$Type_Intervention == "switch to western diet", "urban_diet",
                                   ifelse(metadata_new$Type_Intervention == "switch to traditional diet", "rural_diet", 
                                          ifelse(metadata_new$Type_Intervention == "Control remain on traditional diet", "controls_r", 
                                                 ifelse(metadata_new$Type_Intervention == "control remain on western diet", "controls_u", NA)))))

metadata_new$diet <- factor(metadata_new$diet, levels = c("urban_diet","rural_diet","MBEGE","controls_u","controls_r"))
unique(metadata_new$diet)
```
bmi
```{r}
densityplot(metadata_new$Bwt_Baseline)
```
```{r}
densityplot(metadata_new$Bwt_PostIntervention)
```
```{r}
densityplot(metadata_new$Bwt_afterBaseline)
```
```{r}
densityplot(metadata_new$Height)
```

```{r}
densityplot(metadata_new$BMI_t0)
```
```{r}
densityplot(metadata_new$BMI_t1)
```
```{r}
densityplot(metadata_new$BMI_t2)
```
```{r}
head(metadata_new$BMI_t2[order(metadata_new$BMI_t2, decreasing = F)])
```

# ----------------------
## create sample table
```{r}
sample_table_full <- data.frame(ID = colnames(metabo_mtx))

sample_table_full$type <- "sample"
sample_table_full[grepl("pSS", sample_table_full$ID), "type"] <- "pSS"

ggplot(sample_table_full, aes(x= type))+
  geom_bar(stat= "count")+theme(aspect.ratio = 1.5)+
  geom_text(color = "white", aes(label = after_stat(count)), stat = "count", vjust = 1)
```
```{r}
sample_table <- subset(sample_table_full, type !="pSS")
dim(sample_table)
```
```{r}
## column diet
sample_table$diet <- NA
sample_table[grepl("DSR", sample_table$ID),"diet"] <- "urban_diet"
sample_table[grepl("DSU", sample_table$ID),"diet"] <- "rural_diet"
sample_table[grepl("DSM", sample_table$ID),"diet"] <- "MBEGE"

## column timepoint
sample_table$timepoint <- paste0("t",str_split_fixed(sample_table$ID,"_", 2)[,2])

## PID
sample_table$PID <- str_split_fixed(sample_table$ID,"_", 2)[,1]

head(sample_table)


```


## control samples
```{r}
unique(sampls_info$Intervention )

controls_u <- sampls_info[sampls_info$Intervention == "control remain on western diet","PID"]
controls_r <- sampls_info[sampls_info$Intervention == "Control remain on traditional diet","PID"]

sample_table$diet <- ifelse(sample_table$PID %in% controls_u, "controls_u", sample_table$diet)

sample_table$diet <- ifelse(sample_table$PID %in% controls_r, "controls_r", sample_table$diet)

unique(sample_table$diet)
```


### .. factor sample table
```{r}
# diet
sample_table$diet <- factor(sample_table$diet, levels = c("urban_diet","rural_diet","MBEGE","controls_r","controls_u"))

#timepoint
sample_table$timepoint <- factor(sample_table$timepoint, levels = c("t0","t1","t2"))

```
# 2. Injection order
```{r }
injections <- read.delim("./240324_injection_order.txt", header = T)

injections$type <- "sample"

injections[grepl("pSS",injections$ID),"type"] <- "pSS"

injections$y <- ifelse(injections$type == "sample",1,2)

ggplot(injections, aes(x= order, y= y, color= type))+
  geom_point(size = 1)+
  scale_color_manual(values = c("sample" = "black", "pSS" = "red"))+
  ylim(0,5)
```
```{r}
pSS_mean <- metabo_mtx[,which(colnames(metabo_mtx) %in% sample_table_full$type == "pSS")]
```


# 3. Pre processing
https://labplan.ie/content/uploads/2018/01/Biocrates-Tutorial-MetaboAnalyst-V01.pdf
I chose to only perform log2 transformation without pareto scaling to avoid problems with data analysis. 

## Normalize and scale data
```{r fig.width=20}
#boxplot metabolites before normalization
boxplot(metabo_mtx)
```


```{r fig.width=15}
#200 metabolites example plot before normalization
boxplot(t(metabo_mtx[1:200,]))
```


```{r}
#log2 normalization of metabolites
metabo_log2<-apply(metabo_mtx,2, function(x){log2(x+1)})
rownames(metabo_log2)<-rownames(metabo_mtx)
```

```{r}
## visualize distribution of samples after log2 transformation
hist(metabo_log2, breaks = 50, freq=FALSE)
curve(dnorm(x,mean=mean(metabo_log2),sd=sd(metabo_log2)), add=TRUE,col="red")
```
```{r}
vis_miss(as.data.frame(metabo_log2))
```

```{r}
#glog2 normalization of metabolites
metabo_glog2<-apply(metabo_mtx,2, glog2)
rownames(metabo_glog2)<-rownames(metabo_mtx)
```

```{r}
## visualize distribution of samples after log2 transformation
hist(metabo_glog2, breaks = 50, freq=FALSE)
curve(dnorm(x,mean=mean(metabo_glog2),sd=sd(metabo_glog2)), add=TRUE,col="red")
```




```{r}
#pareto scaling of metabolites
metabo_log2_pareto<-apply(metabo_log2,2, pareto_scale )
rownames(metabo_log2_pareto)<-rownames(metabo_log2)
```

```{r}
## visualize distribution of samples after log2 transformation and pareto scaling
hist(metabo_log2_pareto, breaks = 50, freq=FALSE)
curve(dnorm(x,mean=mean(metabo_log2_pareto),sd=sd(metabo_log2_pareto)), add=TRUE,col="red")
```

```{r}
## visualize distribution of samples after log2 transformation and pareto scaling: qq plot
# input_qq<-reshape2::melt(metabo_log2_pareto)
# qqPlot(input_qq$value)
```

### ...Log2 only: metabolites boxplot

```{r fig.width=20}
topvar_metabo<- order(Rfast::colVars(metabo_log2), decreasing = T)[1:50]
boxplot(metabo_log2[,topvar_metabo])
```


# 4. PCA
## Urban diet PCA
```{r}
sample_table_urbandiet <- subset(sample_table, diet == "urban_diet")
rownames(sample_table_urbandiet) <- sample_table_urbandiet$ID

metabo_log2_urbandiet <-metabo_log2[,match(sample_table_urbandiet$ID, colnames(metabo_log2))]

identical(sample_table_urbandiet$ID, colnames(metabo_log2_urbandiet))

pca.urbandiet <- plotPCAnew(pca_input = metabo_log2_urbandiet, 
           ntop = "all", Scale = T,
           xPC = 1, yPC = 2, point_size = 4, 
           anno_colour = col_timepoint, 
           scale = T, anno= sample_table_urbandiet, 
           color = "timepoint", color.ellipse = T, title = "urban diet")+border() + theme_bw()+ theme(aspect.ratio = 1)

pca.urbandiet
```
### ... pca loadings urban diet
var.coord = loadings * the component standard deviations
var.cos2 = var.coord^2
var.contrib. The contribution of a variable to a given principal component is (in percentage) : (var.cos2 * 100) / (total cos2 of the component)
http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/112-pca-principal-component-analysis-essentials/
```{r}
 pca_urbandiet <- prcomp(t(metabo_log2_urbandiet), scale. = T) 
pca_urbandiet <- data.frame(pca_urbandiet$x[,c(1:2)])
pca_urbandiet$timepoint <- sample_table_urbandiet[match(rownames(pca_urbandiet), sample_table_urbandiet$ID),"timepoint"]

ggplot(pca_urbandiet, aes(x= PC1, y= PC2, color = timepoint))+
  geom_point(size = 3)+stat_ellipse(type = "norm", aes(color = timepoint))+
    theme_bw()+
  theme(aspect.ratio = 1)+
  scale_color_manual(values = col_timepoint)

```
```{r}
 pca_urbandiet <- prcomp(t(metabo_log2_urbandiet), scale. = T)
urbandiet_pca.var <-as.data.frame(pca_urbandiet$rotation)
urbandiet_pca.var <- urbandiet_pca.var[order(urbandiet_pca.var$PC1, decreasing = T),]
urbandiet_pca.var50 <- list(top50 = rownames(urbandiet_pca.var)[c(1:50)], 
                            low50 = rownames(urbandiet_pca.var)[c((nrow(urbandiet_pca.var)-49):nrow(urbandiet_pca.var))])

 urbandiet_pca.var50 <- lapply(urbandiet_pca.var50, function(x){
   names(x) <- melt_annotation_tal_new[match(x, as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]
 })

 urbandiet_pca.var50
```
```{r}
murbandiet_pca.var50_hmdb <- lapply(urbandiet_pca.var50, function(x){
  melt_annotation_tal_new[match(x, melt_annotation_tal_new$CompoundName),"CompoundID_new"]
})

head(murbandiet_pca.var50_hmdb)
```


```{r}
list_topvar_urbandiet <- list()
  set.seed(34)

for (i in names(murbandiet_pca.var50_hmdb)){
  print(i)
   tmp <- fora(pathways = smdp_list, 
       genes = murbandiet_pca.var50_hmdb[[i]], # genes in cluster i
       universe = universe, # all genes
       minSize = 3, 
       maxSize = 500)
  
  tmp <- as.data.frame(tmp)
  
  tmp$comparison <- i
  
  tmp <- tmp[order(tmp$overlap, decreasing = T),]
  
  list_topvar_urbandiet[[i]] <- tmp
  
}
  
list_topvar_urbandiet$low50
```

```{r fig.height=3, fig.width=3}
res.pca_urbandiet <- PCA(t(metabo_log2_urbandiet), scale.unit = T, graph = F)

fviz_pca_ind(res.pca_urbandiet,
             geom.ind = "point", # show points only (nbut not "text")
             col.ind = sample_table_urbandiet$timepoint, # color by groups
             palette = c("#00AFBB", "#E7B800", "#FC4E07"),
             addEllipses = TRUE, # Concentration ellipses
             legend.title = "Groups"
             )

res.pca_urbandiet

```
```{r}
var_urbandiet <- get_pca_var(res.pca_urbandiet)
var_urbandiet
var.contrib_urbandiet <- var_urbandiet$contrib
var.contrib_urbandiet <- as.data.frame(var.contrib_urbandiet)
var.contrib_urbandiet$CompoundName_new <- melt_annotation_tal_new[match(rownames(var.contrib_urbandiet),as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

```
```{r}
var.contrib_urbandiet <- var.contrib_urbandiet[order(var.contrib_urbandiet$Dim.1, decreasing = T),]
head(var.contrib_urbandiet, 6)
```


## Rural diet PCA
```{r}
sample_table_ruraldiet <- subset(sample_table, diet == "rural_diet")
rownames(sample_table_ruraldiet) <- sample_table_ruraldiet$ID

metabo_log2_ruraldiet <-metabo_log2[,match(sample_table_ruraldiet$ID, colnames(metabo_log2))]

identical(sample_table_ruraldiet$ID, colnames(metabo_log2_ruraldiet))

pca.ruraldiet <- plotPCAnew(pca_input = metabo_log2_ruraldiet, 
           ntop = "all", Scale = T,
           xPC = 1, yPC = 2, point_size = 4, 
           anno_colour = col_timepoint, 
           scale = T, anno= sample_table_ruraldiet, 
           color = "timepoint", color.ellipse = T, title = "rural diet")+border() + theme_bw()+ theme(aspect.ratio = 1)

pca.ruraldiet
```



## MBEGE diet PCA

```{r}
sample_table_mbege <- subset(sample_table, diet == "MBEGE")
rownames(sample_table_mbege) <- sample_table_mbege$ID

metabo_log2_mbege <-metabo_log2[,match(sample_table_mbege$ID, colnames(metabo_log2))]

identical(sample_table_mbege$ID, colnames(metabo_log2_mbege))

pca.mbege <- plotPCAnew(pca_input = metabo_log2_mbege, 
           ntop = "all", Scale = T,
           xPC = 1, yPC = 2, point_size = 4, 
           anno_colour = col_timepoint, 
           scale = T, anno= sample_table_mbege, 
           color = "timepoint", color.ellipse = T, title = "mbege diet")+border() + theme_bw()+ theme(aspect.ratio = 1)

pca.mbege
```

# ===============================================
# 5. Mutliple t-test
## SET FC threshold
```{r}
fc_threshold <- 1.2
```


# -----------------------------
# 5.1. Urban diet
## data distribution
```{r}
data_melt_urbandiet <- as.data.frame(t(metabo_log2_urbandiet))
data_melt_urbandiet$ID <- rownames(data_melt_urbandiet)

data_melt_urbandiet <- reshape2::melt(data_melt_urbandiet, id.var = "ID")
data_melt_urbandiet$timepoint <- sample_table[match(data_melt_urbandiet$ID, sample_table$ID),"timepoint"]
data_melt_urbandiet$PID <- sample_table[match(data_melt_urbandiet$ID, sample_table$ID),"PID"]

head(data_melt_urbandiet)
```


```{r}
# ur <- data_melt_urbandiet %>% group_by(variable)
# 
# list_test <- group_split(ur)
# names(list_test) <-group_keys(ur)$variable
# 
# 
# ## create a list with each one as stim_cyto
# 
# sp.test <- lapply(list_test, function(x){
#   x %>%
#   group_by(timepoint) %>%
#   shapiro_test(value)
# })
# 
# for (i in names(sp.test)){
#   sp.test[[i]]$variable <- i
# }
# 
# 
# sp.test <- bind_rows(sp.test)
```

```{r}
# sp.test$sig <- ifelse(sp.test$p <= 0.05, "no normal dist", "normal")
# 
# ggplot(sp.test, aes(x=timepoint, fill= sig))+
#   geom_bar(stat = "count", position = "fill")+
#   theme(aspect.ratio = 2)
```






# ----------------------
## Variance partition: Urban diet
### ... add info from metadata_new
```{r}
cols_2_add <- c("Age","BMI_t0","activity_sum_rank_scaled","PID")

sample_table_urbandiet <- merge(sample_table_urbandiet, metadata_new[,cols_2_add], by = "PID", all.x = T, all.y = F)
rownames(sample_table_urbandiet) <- sample_table_urbandiet$ID

head(sample_table_urbandiet)
```

### step1: find metabolites contributing to more than 5% variation for each time comparison
```{r}
form <- ~ Age + (1 | PID) + timepoint_numeric + BMI_t0 + activity_sum_rank_scaled

varpart_features_urbandiet <- list()

for ( i in c("t0_vs_t1","t1_vs_t2","t0_vs_t2")){
  tp1 <- str_split_fixed(i,"_vs_",2)[,1]
  tp2 <- str_split_fixed(i,"_vs_",2)[,2]
  
  tmpinfo <- subset(sample_table_urbandiet, timepoint %in% c(tp1,tp2))
  
  ## make timepoint variable into numeric
  tmpinfo$timepoint_numeric <- as.numeric(gsub("t","",tmpinfo$timepoint))
  
  tmpdata <- metabo_log2_urbandiet[,match(tmpinfo$ID, colnames(metabo_log2_urbandiet))]
  
  if(identical(tmpinfo$ID, colnames(tmpdata))){
    print("identical")
  } else{
    stop("error, info and data samples not ordered correctly")
  }
  
  
  tmpvarpart <- fitExtractVarPartModel(tmpdata, form, tmpinfo)
  
  # sort variables (i.e. columns) by median fraction
#       of variance explained
tmpvarpart <- sortCols(tmpvarpart)

varpart_features_urbandiet[[i]]$varpart <- tmpvarpart
varpart_features_urbandiet[[i]]$info <- tmpinfo
varpart_features_urbandiet[[i]]$data <- tmpdata

}
```

### step2: plot results
```{r}
## colors 
colors_vars <- c("Residuals" = "lightgray",
                 "timepoint_numeric"  = "turquoise",
                 "Age" = "purple",
                 "BMI_t0" = "cornflowerblue",
                 "PID" = "lightyellow",
                 "activity_sum_rank_scaled" = "tomato3")
```

```{r}
list_parvar_plots_urbandiet <- list()

for (i in names(varpart_features_urbandiet)){
  print(i)
  
tmp_df <- as.data.frame(varpart_features_urbandiet[[i]]$varpart)
tmp_df$ionMz <- rownames(tmp_df)
tmp_df <- reshape2::melt(tmp_df, id.vars = "ionMz")

timepoint_var <- subset(tmp_df, variable == "timepoint_numeric")$value
names(timepoint_var) <- as.character(subset(tmp_df, variable == "timepoint_numeric")$ionMz)

tmp_df$timepoint <- timepoint_var[match(tmp_df$ionMz, names(timepoint_var))]
tmp_df$value100 <- 100*tmp_df$value

## factor
tmp_df$variable <- factor(tmp_df$variable, levels = rev(c("timepoint_numeric","Age","BMI_t0","activity_sum_rank_scaled", "PID","Residuals")))


tmp_df <- tmp_df[order(tmp_df$timepoint, decreasing = T),]

tmp_df$compoundname <- melt_annotation_tal_new[match(as.character(tmp_df$ionMz), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

# barplot
tmpplot <- ggplot(subset(tmp_df, timepoint >0.05), aes(x = reorder(compoundname, timepoint), y = value100, fill = variable))+
  geom_bar(stat = "identity", position = "stack")+
  coord_flip()+
  scale_fill_manual(values = colors_vars)+
 xlab("")+
  ylab("% Explained variance")+
  ggtitle("t0vst1 Top contributions (>5%) - timepoint")


list_parvar_plots_urbandiet[["df"]][[i]] <- tmp_df
list_parvar_plots_urbandiet[["plot"]][[i]] <- tmpplot
}
```
```{r fig.width=20, fig.height=10}
wrap_plots(list_parvar_plots_urbandiet$plot, nrow = 1, guides = "collect")
```

### step3: extract metabolites contributing >5% for downstream analysis
```{r}
for (i in names(list_parvar_plots_urbandiet$df)){
  print(i)
  
  tmp_df <- NULL
  
  tmp_df <- list_parvar_plots_urbandiet$df[[i]]
  
  var_metabo <- unique(tmp_df[tmp_df$timepoint>0.05,"ionMz"])

  print(paste0(round(length(var_metabo)*100/length(unique(tmp_df$ionMz)),digits = 2),"% metabolites taken for analysis for comparison ",i))
  
  
  ## add to varpart_features_urbandiet
  varpart_features_urbandiet[[i]]$vars <- var_metabo
}
```



### step4: perform dream analysis
```{r}
list_dream_urbandiet <- list()

for (i in names(varpart_features_urbandiet)){
  print(i)
  
  myinfo <- varpart_features_urbandiet[[i]]$info
  ## factor for the variables present
  myinfo$timepoint <- factor(myinfo$timepoint, levels = unique(myinfo$timepoint))
  
  mydata <- varpart_features_urbandiet[[i]]$data
  ## filter variables
  mydata <- mydata[which(rownames(mydata) %in% as.character( varpart_features_urbandiet[[i]]$vars)),]
  
  mycontrasts <- c(
    t0_vs_t1 = "timepointt0 - timepointt1",
    t0_vs_t2 = "timepointt0 - timepointt2",
    t1_vs_t2 = "timepointt1 - timepointt2"
  )
  
  
  L <- makeContrastsDream(~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID) , myinfo,
  contrasts = mycontrasts[names(mycontrasts) == i]
)

# Visualize contrast matrix
plotContrasts(L)

fitmm_dream <- dream(mydata, ~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID), myinfo, L)

fitmm_dream <- eBayes(fitmm_dream)

dream_res <- variancePartition::topTable(fitmm_dream, coef = i, number = Inf, adjust.method = "BH", sort.by = "p", lfc = 0)

dream_res$ionMz <- rownames(dream_res)

print(nrow(dream_res[dream_res$adj.P.Val <=0.05,]))

list_dream_urbandiet[[i]] <- dream_res
}
```
```{r}
res_dream_urbandiet <- do.call(rbind, Map(cbind, comparison = names(list_dream_urbandiet),list_dream_urbandiet))

head(res_dream_urbandiet)
```
```{r}
## add sig
res_dream_urbandiet <- as.data.frame(res_dream_urbandiet %>% add_significance(p.col = "adj.P.Val"))
```

### step5: calculate paire fc

```{r}
pairedFC_urbandiet_diet <- data.frame()

for(i in names(varpart_features_urbandiet)){
  print(i)
  
  ctr_condition = str_split_fixed(i, "_vs_",2)[,1]
  test_condition = str_split_fixed(i, "_vs_",2)[,2]
  
  print(ctr_condition)
  print(test_condition)
  
  ctrids <- rownames(subset(varpart_features_urbandiet[[i]]$info, 
                            as.character(timepoint) == ctr_condition))
  testids <- rownames(subset(varpart_features_urbandiet[[i]]$info, 
                             as.character(timepoint) == test_condition))

  
  
  tmpfc <- fcros::fcrosFCmat(varpart_features_urbandiet[[i]]$data, ctrids, testids, log2.opt = 0)

  tmp_pairedFC <- data.frame(ionMz = tmpfc$idnames, 
                                       FC = as.numeric(tmpfc$FC), 
                                       FC2 = as.numeric(tmpfc$FC2))

  tmp_pairedFC$log2FC <- log2(tmp_pairedFC$FC)
  tmp_pairedFC$log2FC2 <- log2(tmp_pairedFC$FC2)

  tmp_pairedFC$direction <- ifelse(tmp_pairedFC$log2FC2>0,"up",  "down")

  tmp_pairedFC$comparison <- i
  
  pairedFC_urbandiet_diet <- rbind(pairedFC_urbandiet_diet, tmp_pairedFC)
}
```
### step6: combine dream and paireFC results

```{r}
### add paired FC
pairedFC_urbandiet_diet$merged <- paste0(pairedFC_urbandiet_diet$ionMz,"_", pairedFC_urbandiet_diet$comparison)

res_dream_urbandiet$merged <- paste0(res_dream_urbandiet$ionMz,"_", res_dream_urbandiet$comparison)

cols_2_add <- c("FC","FC2","log2FC","log2FC2","direction","merged")

res_dream_urbandiet <- merge(res_dream_urbandiet, pairedFC_urbandiet_diet[,cols_2_add], by = "merged")

res_dream_urbandiet$regulation <- ifelse(res_dream_urbandiet$adj.P.Val.signif != "ns" & abs(res_dream_urbandiet$log2FC2) > log2(fc_threshold), res_dream_urbandiet$direction, "ns")

rownames(res_dream_urbandiet) <- res_dream_urbandiet$merged

res_dream_urbandiet$CompoundName <- melt_annotation_tal_new[match(as.character(res_dream_urbandiet$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

res_dream_urbandiet$HMDB <- melt_annotation_tal_new[match(as.character(res_dream_urbandiet$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundID_new"]

res_dream_urbandiet$label <- ifelse(res_dream_urbandiet$adj.P.Val.signif !="ns", str_trunc(res_dream_urbandiet$CompoundName,24),NA)

head(res_dream_urbandiet)
```

### step7: list of sig metabolites
```{r}
list_sig_dream_urbandiet_diet <- list()

for(i in unique(res_dream_urbandiet$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet_diet[[i]] <- split(
    subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$ionMz,
     subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet_diet)
```
```{r}
list_sig_dream_urbandiet_diet_name <- list()

for(i in unique(res_dream_urbandiet$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet_diet_name[[i]] <- split(
    subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$CompoundName,
     subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet_diet_name)
```
```{r}
list_sig_dream_urbandiet_diet_hmdb <- list()

for(i in unique(res_dream_urbandiet$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet_diet_hmdb[[i]] <- split(
    subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$HMDB,
     subset(res_dream_urbandiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet_diet_hmdb)
```

### step8 compare to previous res - SKIP
### step9: plots
count
```{r}
count_da_urbandiet <- as.data.frame(res_dream_urbandiet %>% 
                                      group_by(comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_urbandiet <- reshape2::melt(count_da_urbandiet)

count_da_urbandiet
```

```{r}
count_da_urbandiet$comparison <- factor(count_da_urbandiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(count_da_urbandiet, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity",color = "black", position = "stack")+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))+
  theme_bw()+
  ggtitle("Urban diet")+
  theme(aspect.ratio = 1.5)
```

# ==================================================
# 5.2. Rural diet
## data distribution
```{r}
data_melt_ruraldiet <- as.data.frame(t(metabo_log2_ruraldiet))
data_melt_ruraldiet$ID <- rownames(data_melt_ruraldiet)

data_melt_ruraldiet <- reshape2::melt(data_melt_ruraldiet, id.var = "ID")
data_melt_ruraldiet$timepoint <- sample_table[match(data_melt_ruraldiet$ID, sample_table$ID),"timepoint"]
data_melt_ruraldiet$PID <- sample_table[match(data_melt_ruraldiet$ID, sample_table$ID),"PID"]

head(data_melt_ruraldiet)
```






# ------------------------------------------
## Variance partition - Rural diet

### ... add info from metadata_new
```{r}
cols_2_add <- c("Age","BMI_t0","activity_sum_rank_scaled","PID")

sample_table_ruraldiet <- merge(sample_table_ruraldiet, metadata_new[,cols_2_add], by = "PID", all.x = T, all.y = F)
rownames(sample_table_ruraldiet) <- sample_table_ruraldiet$ID

head(sample_table_ruraldiet)
```

### step1: find metabolites contributing to more than 5% variation for each time comparison
```{r}
form <- ~ Age + (1 | PID) + timepoint_numeric + BMI_t0 + activity_sum_rank_scaled

varpart_features_ruraldiet <- list()

for ( i in c("t0_vs_t1","t1_vs_t2","t0_vs_t2")){
  tp1 <- str_split_fixed(i,"_vs_",2)[,1]
  tp2 <- str_split_fixed(i,"_vs_",2)[,2]
  
  tmpinfo <- subset(sample_table_ruraldiet, timepoint %in% c(tp1,tp2))
  
  ## make timepoint variable into numeric
  tmpinfo$timepoint_numeric <- as.numeric(gsub("t","",tmpinfo$timepoint))
  
  tmpdata <- metabo_log2_ruraldiet[,match(tmpinfo$ID, colnames(metabo_log2_ruraldiet))]
  
  if(identical(tmpinfo$ID, colnames(tmpdata))){
    print("identical")
  } else{
    stop("error, info and data samples not ordered correctly")
  }
  
  
  tmpvarpart <- fitExtractVarPartModel(tmpdata, form, tmpinfo)
  
  # sort variables (i.e. columns) by median fraction
#       of variance explained
tmpvarpart <- sortCols(tmpvarpart)

varpart_features_ruraldiet[[i]]$varpart <- tmpvarpart
varpart_features_ruraldiet[[i]]$info <- tmpinfo
varpart_features_ruraldiet[[i]]$data <- tmpdata

}
```

### step2: plot results
```{r}
## colors 
colors_vars <- c("Residuals" = "lightgray",
                 "timepoint_numeric"  = "turquoise",
                 "Age" = "purple",
                 "BMI_t0" = "cornflowerblue",
                 "PID" = "lightyellow",
                 "activity_sum_rank_scaled" = "tomato3")
```

```{r}
list_parvar_plots_ruraldiet <- list()

for (i in names(varpart_features_ruraldiet)){
  print(i)
  
tmp_df <- as.data.frame(varpart_features_ruraldiet[[i]]$varpart)
tmp_df$ionMz <- rownames(tmp_df)
tmp_df <- reshape2::melt(tmp_df, id.vars = "ionMz")

timepoint_var <- subset(tmp_df, variable == "timepoint_numeric")$value
names(timepoint_var) <- as.character(subset(tmp_df, variable == "timepoint_numeric")$ionMz)

tmp_df$timepoint <- timepoint_var[match(tmp_df$ionMz, names(timepoint_var))]
tmp_df$value100 <- 100*tmp_df$value

## factor
tmp_df$variable <- factor(tmp_df$variable, levels = rev(c("timepoint_numeric","Age","BMI_t0","activity_sum_rank_scaled", "PID","Residuals")))


tmp_df <- tmp_df[order(tmp_df$timepoint, decreasing = T),]

tmp_df$compoundname <- melt_annotation_tal_new[match(as.character(tmp_df$ionMz), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

# barplot
tmpplot <- ggplot(subset(tmp_df, timepoint >0.05), aes(x = reorder(compoundname, timepoint), y = value100, fill = variable))+
  geom_bar(stat = "identity", position = "stack")+
  coord_flip()+
  scale_fill_manual(values = colors_vars)+
 xlab("")+
  ylab("% Explained variance")+
  ggtitle("t0vst1 Top contributions (>5%) - timepoint")


list_parvar_plots_ruraldiet[["df"]][[i]] <- tmp_df
list_parvar_plots_ruraldiet[["plot"]][[i]] <- tmpplot
}
```
```{r fig.width=20, fig.height=10}
wrap_plots(list_parvar_plots_ruraldiet$plot, nrow = 1, guides = "collect")
```

### step3: extract metabolites contributing >5% for downstream analysis
```{r}
for (i in names(list_parvar_plots_ruraldiet$df)){
  print(i)
  
  tmp_df <- NULL
  
  tmp_df <- list_parvar_plots_ruraldiet$df[[i]]
  
  var_metabo <- unique(tmp_df[tmp_df$timepoint>0.05,"ionMz"])

  print(paste0(round(length(var_metabo)*100/length(unique(tmp_df$ionMz)),digits = 2),"% metabolites taken for analysis for comparison ",i))
  
  
  ## add to varpart_features_ruraldiet
  varpart_features_ruraldiet[[i]]$vars <- var_metabo
}
```



### step4: perform dream analysis
```{r}
list_dream_ruraldiet <- list()

for (i in names(varpart_features_ruraldiet)){
  print(i)
  
  myinfo <- varpart_features_ruraldiet[[i]]$info
  ## factor for the variables present
  myinfo$timepoint <- factor(myinfo$timepoint, levels = unique(myinfo$timepoint))
  
  mydata <- varpart_features_ruraldiet[[i]]$data
  ## filter variables
  mydata <- mydata[which(rownames(mydata) %in% as.character( varpart_features_ruraldiet[[i]]$vars)),]
  
  mycontrasts <- c(
    t0_vs_t1 = "timepointt0 - timepointt1",
    t0_vs_t2 = "timepointt0 - timepointt2",
    t1_vs_t2 = "timepointt1 - timepointt2"
  )
  
  
  L <- makeContrastsDream(~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID) , myinfo,
  contrasts = mycontrasts[names(mycontrasts) == i]
)

# Visualize contrast matrix
plotContrasts(L)

fitmm_dream <- dream(mydata, ~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID), myinfo, L)

fitmm_dream <- eBayes(fitmm_dream)

dream_res <- variancePartition::topTable(fitmm_dream, coef = i, number = Inf, adjust.method = "BH", sort.by = "p", lfc = 0)

dream_res$ionMz <- rownames(dream_res)

print(nrow(dream_res[dream_res$adj.P.Val <=0.05,]))

list_dream_ruraldiet[[i]] <- dream_res
}
```
```{r}
res_dream_ruraldiet <- do.call(rbind, Map(cbind, comparison = names(list_dream_ruraldiet),list_dream_ruraldiet))

head(res_dream_ruraldiet)
```
```{r}
## add sig
res_dream_ruraldiet <- as.data.frame(res_dream_ruraldiet %>% add_significance(p.col = "adj.P.Val"))
```

### step5: calculate paire fc

```{r}
pairedFC_ruraldiet_diet <- data.frame()

for(i in names(varpart_features_ruraldiet)){
  print(i)
  
  ctr_condition = str_split_fixed(i, "_vs_",2)[,1]
  test_condition = str_split_fixed(i, "_vs_",2)[,2]
  
  print(ctr_condition)
  print(test_condition)
  
  ctrids <- rownames(subset(varpart_features_ruraldiet[[i]]$info, 
                            as.character(timepoint) == ctr_condition))
  testids <- rownames(subset(varpart_features_ruraldiet[[i]]$info, 
                             as.character(timepoint) == test_condition))

  
  
  tmpfc <- fcros::fcrosFCmat(varpart_features_ruraldiet[[i]]$data, ctrids, testids, log2.opt = 0)

  tmp_pairedFC <- data.frame(ionMz = tmpfc$idnames, 
                                       FC = as.numeric(tmpfc$FC), 
                                       FC2 = as.numeric(tmpfc$FC2))

  tmp_pairedFC$log2FC <- log2(tmp_pairedFC$FC)
  tmp_pairedFC$log2FC2 <- log2(tmp_pairedFC$FC2)

  tmp_pairedFC$direction <- ifelse(tmp_pairedFC$log2FC2>0,"up",  "down")

  tmp_pairedFC$comparison <- i
  
  pairedFC_ruraldiet_diet <- rbind(pairedFC_ruraldiet_diet, tmp_pairedFC)
}
```
### step6: combine dream and paireFC results

```{r}
### add paired FC
pairedFC_ruraldiet_diet$merged <- paste0(pairedFC_ruraldiet_diet$ionMz,"_", pairedFC_ruraldiet_diet$comparison)

res_dream_ruraldiet$merged <- paste0(res_dream_ruraldiet$ionMz,"_", res_dream_ruraldiet$comparison)

cols_2_add <- c("FC","FC2","log2FC","log2FC2","direction","merged")

res_dream_ruraldiet <- merge(res_dream_ruraldiet, pairedFC_ruraldiet_diet[,cols_2_add], by = "merged")

res_dream_ruraldiet$regulation <- ifelse(res_dream_ruraldiet$adj.P.Val.signif != "ns" & abs(res_dream_ruraldiet$log2FC2) > log2(fc_threshold), res_dream_ruraldiet$direction, "ns")

rownames(res_dream_ruraldiet) <- res_dream_ruraldiet$merged

res_dream_ruraldiet$CompoundName <- melt_annotation_tal_new[match(as.character(res_dream_ruraldiet$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

res_dream_ruraldiet$HMDB <- melt_annotation_tal_new[match(as.character(res_dream_ruraldiet$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundID_new"]

res_dream_ruraldiet$label <- ifelse(res_dream_ruraldiet$adj.P.Val.signif !="ns", str_trunc(res_dream_ruraldiet$CompoundName,24),NA)

head(res_dream_ruraldiet)
```

### step7: list of sig metabolites
```{r}
list_sig_dream_ruraldiet_diet <- list()

for(i in unique(res_dream_ruraldiet$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet_diet[[i]] <- split(
    subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$ionMz,
     subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet_diet)
```
```{r}
list_sig_dream_ruraldiet_diet_name <- list()

for(i in unique(res_dream_ruraldiet$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet_diet_name[[i]] <- split(
    subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$CompoundName,
     subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet_diet_name)
```
```{r}
list_sig_dream_ruraldiet_diet_hmdb <- list()

for(i in unique(res_dream_ruraldiet$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet_diet_hmdb[[i]] <- split(
    subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$HMDB,
     subset(res_dream_ruraldiet, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet_diet_hmdb)
```

### step8 compare to previous res - SKIP

### step9: plots
count
```{r}
count_da_ruraldiet <- as.data.frame(res_dream_ruraldiet %>% 
                                      group_by(comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_ruraldiet <- reshape2::melt(count_da_ruraldiet)

count_da_ruraldiet
```

```{r}
count_da_ruraldiet$comparison <- factor(count_da_ruraldiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(count_da_ruraldiet, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity",color = "black", position = "stack")+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))+
  theme_bw()+
  ggtitle("Urban diet")+
  theme(aspect.ratio = 1.5)
```


# =================================
# 5.3. Mbege
## data distribution
```{r}
data_melt_mbege <- as.data.frame(t(metabo_log2_mbege))
data_melt_mbege$ID <- rownames(data_melt_mbege)

data_melt_mbege <- reshape2::melt(data_melt_mbege, id.var = "ID")
data_melt_mbege$timepoint <- sample_table[match(data_melt_mbege$ID, sample_table$ID),"timepoint"]
data_melt_mbege$PID <- sample_table[match(data_melt_mbege$ID, sample_table$ID),"PID"]

head(data_melt_mbege)
```

### ... repeated paired t-test
```{r}
## order by time point and PID
data_melt_mbege <- data_melt_mbege[order(data_melt_mbege$timepoint, data_melt_mbege$PID, decreasing = F),]

head(data_melt_mbege)
```

```{r}
ggplot(sample_table_mbege, aes(x= timepoint, y= PID, color = diet))+
  geom_point()+
  geom_line(aes(group = PID))
```
```{r}
table(sample_table_mbege$timepoint)
```







# ===================

## Variance partition - MBEGE

### ... add info from metadata_new
```{r}
cols_2_add <- c("Age","BMI_t0","activity_sum_rank_scaled","PID")

sample_table_mbege <- merge(sample_table_mbege, metadata_new[,cols_2_add], by = "PID", all.x = T, all.y = F)
rownames(sample_table_mbege) <- sample_table_mbege$ID

head(sample_table_mbege)
```

### step1: find metabolites contributing to more than 5% variation for each time comparison
```{r}
form <- ~ Age + (1 | PID) + timepoint_numeric + BMI_t0 + activity_sum_rank_scaled

varpart_features_mbege <- list()

for ( i in c("t0_vs_t1","t1_vs_t2","t0_vs_t2")){
  tp1 <- str_split_fixed(i,"_vs_",2)[,1]
  tp2 <- str_split_fixed(i,"_vs_",2)[,2]
  
  tmpinfo <- subset(sample_table_mbege, timepoint %in% c(tp1,tp2))
  
  ## make timepoint variable into numeric
  tmpinfo$timepoint_numeric <- as.numeric(gsub("t","",tmpinfo$timepoint))
  
  tmpdata <- metabo_log2_mbege[,match(tmpinfo$ID, colnames(metabo_log2_mbege))]
  
  if(identical(tmpinfo$ID, colnames(tmpdata))){
    print("identical")
  } else{
    stop("error, info and data samples not ordered correctly")
  }
  
  
  tmpvarpart <- fitExtractVarPartModel(tmpdata, form, tmpinfo)
  
  # sort variables (i.e. columns) by median fraction
#       of variance explained
tmpvarpart <- sortCols(tmpvarpart)

varpart_features_mbege[[i]]$varpart <- tmpvarpart
varpart_features_mbege[[i]]$info <- tmpinfo
varpart_features_mbege[[i]]$data <- tmpdata

}
```

### step2: plot results
```{r}
## colors 
colors_vars <- c("Residuals" = "lightgray",
                 "timepoint_numeric"  = "turquoise",
                 "Age" = "purple",
                 "BMI_t0" = "cornflowerblue",
                 "PID" = "lightyellow",
                 "activity_sum_rank_scaled" = "tomato3")
```

```{r}
list_parvar_plots_mbege <- list()

for (i in names(varpart_features_mbege)){
  print(i)
  
tmp_df <- as.data.frame(varpart_features_mbege[[i]]$varpart)
tmp_df$ionMz <- rownames(tmp_df)
tmp_df <- reshape2::melt(tmp_df, id.vars = "ionMz")

timepoint_var <- subset(tmp_df, variable == "timepoint_numeric")$value
names(timepoint_var) <- as.character(subset(tmp_df, variable == "timepoint_numeric")$ionMz)

tmp_df$timepoint <- timepoint_var[match(tmp_df$ionMz, names(timepoint_var))]
tmp_df$value100 <- 100*tmp_df$value

## factor
tmp_df$variable <- factor(tmp_df$variable, levels = rev(c("timepoint_numeric","Age","BMI_t0","activity_sum_rank_scaled", "PID","Residuals")))


tmp_df <- tmp_df[order(tmp_df$timepoint, decreasing = T),]

tmp_df$compoundname <- melt_annotation_tal_new[match(as.character(tmp_df$ionMz), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

# barplot
tmpplot <- ggplot(subset(tmp_df, timepoint >0.05), aes(x = reorder(compoundname, timepoint), y = value100, fill = variable))+
  geom_bar(stat = "identity", position = "stack")+
  coord_flip()+
  scale_fill_manual(values = colors_vars)+
 xlab("")+
  ylab("% Explained variance")+
  ggtitle("t0vst1 Top contributions (>5%) - timepoint")


list_parvar_plots_mbege[["df"]][[i]] <- tmp_df
list_parvar_plots_mbege[["plot"]][[i]] <- tmpplot
}
```
```{r fig.width=20, fig.height=10}
wrap_plots(list_parvar_plots_mbege$plot, nrow = 1, guides = "collect")
```

### step3: extract metabolites contributing >5% for downstream analysis
```{r}
for (i in names(list_parvar_plots_mbege$df)){
  print(i)
  
  tmp_df <- NULL
  
  tmp_df <- list_parvar_plots_mbege$df[[i]]
  
  var_metabo <- unique(tmp_df[tmp_df$timepoint>0.05,"ionMz"])

  print(paste0(round(length(var_metabo)*100/length(unique(tmp_df$ionMz)),digits = 2),"% metabolites taken for analysis for comparison ",i))
  
  
  ## add to varpart_features_mbege
  varpart_features_mbege[[i]]$vars <- var_metabo
}
```



### step4: perform dream analysis
```{r}
list_dream_mbege <- list()

for (i in names(varpart_features_mbege)){
  print(i)
  
  myinfo <- varpart_features_mbege[[i]]$info
  ## factor for the variables present
  myinfo$timepoint <- factor(myinfo$timepoint, levels = unique(myinfo$timepoint))
  
  mydata <- varpart_features_mbege[[i]]$data
  ## filter variables
  mydata <- mydata[which(rownames(mydata) %in% as.character( varpart_features_mbege[[i]]$vars)),]
  
  mycontrasts <- c(
    t0_vs_t1 = "timepointt0 - timepointt1",
    t0_vs_t2 = "timepointt0 - timepointt2",
    t1_vs_t2 = "timepointt1 - timepointt2"
  )
  
  
  L <- makeContrastsDream(~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID) , myinfo,
  contrasts = mycontrasts[names(mycontrasts) == i]
)

# Visualize contrast matrix
plotContrasts(L)

fitmm_dream <- dream(mydata, ~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID), myinfo, L)

fitmm_dream <- eBayes(fitmm_dream)

dream_res <- variancePartition::topTable(fitmm_dream, coef = i, number = Inf, adjust.method = "BH", sort.by = "p", lfc = 0)

dream_res$ionMz <- rownames(dream_res)

print(nrow(dream_res[dream_res$adj.P.Val <=0.05,]))

list_dream_mbege[[i]] <- dream_res
}
```
```{r}
res_dream_mbege <- do.call(rbind, Map(cbind, comparison = names(list_dream_mbege),list_dream_mbege))

head(res_dream_mbege)
```
```{r}
## add sig
res_dream_mbege <- as.data.frame(res_dream_mbege %>% add_significance(p.col = "adj.P.Val"))
```

### step5: calculate paire fc

```{r}
pairedFC_mbege_diet <- data.frame()

for(i in names(varpart_features_mbege)){
  print(i)
  
  ctr_condition = str_split_fixed(i, "_vs_",2)[,1]
  test_condition = str_split_fixed(i, "_vs_",2)[,2]
  
  print(ctr_condition)
  print(test_condition)
  
  ctrids <- rownames(subset(varpart_features_mbege[[i]]$info, 
                            as.character(timepoint) == ctr_condition))
  testids <- rownames(subset(varpart_features_mbege[[i]]$info, 
                             as.character(timepoint) == test_condition))

  
  
  tmpfc <- fcros::fcrosFCmat(varpart_features_mbege[[i]]$data, ctrids, testids, log2.opt = 0)

  tmp_pairedFC <- data.frame(ionMz = tmpfc$idnames, 
                                       FC = as.numeric(tmpfc$FC), 
                                       FC2 = as.numeric(tmpfc$FC2))

  tmp_pairedFC$log2FC <- log2(tmp_pairedFC$FC)
  tmp_pairedFC$log2FC2 <- log2(tmp_pairedFC$FC2)

  tmp_pairedFC$direction <- ifelse(tmp_pairedFC$log2FC2>0,"up",  "down")

  tmp_pairedFC$comparison <- i
  
  pairedFC_mbege_diet <- rbind(pairedFC_mbege_diet, tmp_pairedFC)
}
```
### step6: combine dream and paireFC results

```{r}
### add paired FC
pairedFC_mbege_diet$merged <- paste0(pairedFC_mbege_diet$ionMz,"_", pairedFC_mbege_diet$comparison)

res_dream_mbege$merged <- paste0(res_dream_mbege$ionMz,"_", res_dream_mbege$comparison)

cols_2_add <- c("FC","FC2","log2FC","log2FC2","direction","merged")

res_dream_mbege <- merge(res_dream_mbege, pairedFC_mbege_diet[,cols_2_add], by = "merged")

res_dream_mbege$regulation <- ifelse(res_dream_mbege$adj.P.Val.signif != "ns" & abs(res_dream_mbege$log2FC2) > log2(fc_threshold), res_dream_mbege$direction, "ns")

rownames(res_dream_mbege) <- res_dream_mbege$merged

res_dream_mbege$CompoundName <- melt_annotation_tal_new[match(as.character(res_dream_mbege$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

res_dream_mbege$HMDB <- melt_annotation_tal_new[match(as.character(res_dream_mbege$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundID_new"]

res_dream_mbege$label <- ifelse(res_dream_mbege$adj.P.Val.signif !="ns", str_trunc(res_dream_mbege$CompoundName,24),NA)

head(res_dream_mbege)
```

### step7: list of sig metabolites
```{r}
list_sig_dream_mbege_diet <- list()

for(i in unique(res_dream_mbege$comparison)){
  
  print(i)
  
  list_sig_dream_mbege_diet[[i]] <- split(
    subset(res_dream_mbege, regulation != "ns" & comparison == i)$ionMz,
     subset(res_dream_mbege, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_mbege_diet)
```
```{r}
list_sig_dream_mbege_diet_name <- list()

for(i in unique(res_dream_mbege$comparison)){
  
  print(i)
  
  list_sig_dream_mbege_diet_name[[i]] <- split(
    subset(res_dream_mbege, regulation != "ns" & comparison == i)$CompoundName,
     subset(res_dream_mbege, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_mbege_diet_name)
```
```{r}
list_sig_dream_mbege_diet_hmdb <- list()

for(i in unique(res_dream_mbege$comparison)){
  
  print(i)
  
  list_sig_dream_mbege_diet_hmdb[[i]] <- split(
    subset(res_dream_mbege, regulation != "ns" & comparison == i)$HMDB,
     subset(res_dream_mbege, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_mbege_diet_hmdb)
```

### step8 compare to previous res- SKIP

### step9: plots
count
```{r}
count_da_mbege <- as.data.frame(res_dream_mbege %>% 
                                      group_by(comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_mbege <- reshape2::melt(count_da_mbege)

count_da_mbege
```

```{r}
count_da_mbege$comparison <- factor(count_da_mbege$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(count_da_mbege, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity",color = "black", position = "stack")+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))+
  theme_bw()+
  ggtitle("Mbege")+
  theme(aspect.ratio = 1.5)
```
# ========================================


## Export metabolites names and analysis metabolites
```{r}


## urban diet
tmp_analysis_urban <- lapply(varpart_features_urbandiet, "[[","vars")
tmp_analysis_urban<-ComplexHeatmap::list_to_matrix(tmp_analysis_urban)
colnames(tmp_analysis_urban) <- paste0("WD: ",colnames(tmp_analysis_urban))


## rural diet
tmp_analysis_rural <- lapply(varpart_features_ruraldiet, "[[","vars")
tmp_analysis_rural<-ComplexHeatmap::list_to_matrix(tmp_analysis_rural)
colnames(tmp_analysis_rural) <- paste0("TD: ",colnames(tmp_analysis_rural))

## Mbege
tmp_analysis_mbege <- lapply(varpart_features_mbege, "[[","vars")
tmp_analysis_mbege<-ComplexHeatmap::list_to_matrix(tmp_analysis_mbege)
colnames(tmp_analysis_mbege) <- paste0("Fermented: ",colnames(tmp_analysis_mbege))

## all

tmp_analysis_metabolites <- merge(tmp_analysis_urban, tmp_analysis_rural, by = "row.names", all.x = T, all.y = T)
rownames(tmp_analysis_metabolites) <- tmp_analysis_metabolites$Row.names
tmp_analysis_metabolites$Row.names <- NULL

tmp_analysis_metabolites <- merge(tmp_analysis_metabolites, tmp_analysis_mbege, by = "row.names", all.x = T, all.y = T)
rownames(tmp_analysis_metabolites) <- tmp_analysis_metabolites$Row.names
tmp_analysis_metabolites$Row.names <- NULL

## replace 1 with X and NA or 0 with ""
tmp_analysis_metabolites[tmp_analysis_metabolites == 1] <- "X"
tmp_analysis_metabolites[tmp_analysis_metabolites == 0] <- ""
tmp_analysis_metabolites[is.na(tmp_analysis_metabolites)] <- ""

## add annotation
tmp_analysis_metabolites$compoundName <- melt_annotation_tal_new[match(rownames(tmp_analysis_metabolites), as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]
tmp_analysis_metabolites$ionMz <- rownames(tmp_analysis_metabolites)
tmp_analysis_metabolites$HMDB_ID <- melt_annotation_tal_new[match(rownames(tmp_analysis_metabolites), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

head(tmp_analysis_metabolites)
```

```{r}
#write.xlsx(tmp_analysis_metabolites, paste0("./",Sys.Date(),"_metabolomics_used_for_analysis.xlsx"), asTable = T)
```

#-------------------------------------
## Combine all results dream: Diets
### .. combine rea_dream tables
```{r}
res_dream_all <- list(urban_diet = res_dream_urbandiet, 
                      rural_diet = res_dream_ruraldiet, 
                      MBEGE = res_dream_mbege)
```
```{r}
res_dream_all <- do.call(rbind, Map(cbind, diet = names(res_dream_all), res_dream_all))

head(res_dream_all)
```
### ... add chemical class
##### .... chemical class
```{r}
tmpinput <- paste0("hmdb:", unique(res_dream_all$HMDB))
rampdb__all_class <- Getchemclass_RAMPdb(hmdbinput =  tmpinput)

length(tmpinput)
rampdb__all_class$HMDB <- gsub("hmdb:","", rampdb__all_class$sourceId)
length(unique(rampdb__all_class$sourceId))
head(rampdb__all_class)
```
```{r}
rampdb__all_class_ClassyFire_class <- subset(rampdb__all_class, class_level_name == "ClassyFire_super_class")

res_dream_all$class_name <- rampdb__all_class_ClassyFire_class[match(res_dream_all$HMDB, rampdb__all_class_ClassyFire_class$HMDB),"class_name"]

head(res_dream_all[,c("HMDB","class_name")])
```

```{r}
#stop("stop")
```

# =========================================

# 6. Plot all sig metabolites together

## 6.1. All sig metabolites
##### ... all sig df
```{r}
all_sig_df <- subset(res_dream_all, regulation != "ns")
head(all_sig_df)
```




##### ... all sig list
```{r}

all_sig_hmdb <- unique(subset(all_sig_df, regulation != "ns")$HMDB)

# all_sig_hmdb <- list()
# 
# for (i in unique(all_sig_df$diet)){
#   print(i)
#   
#   for (j in unique(all_sig_df$comparison)){
#     all_sig_hmdb[[i]][[j]] <- split(
#       subset(all_sig_df, diet ==i & comparison ==j & regulation !="ns")$HMDB,
#       subset(all_sig_df, diet ==i & comparison ==j & regulation !="ns")$regulation
#     )
#   }
# }

str(all_sig_hmdb)
```
```{r}
all_sig_ionMz <- subset(all_sig_df, regulation != "ns")

all_sig_ionMz <- split(all_sig_ionMz$ionMz, all_sig_ionMz$diet)

str(all_sig_ionMz)
```
## 6.2. annotation
```{r}
anno_pathways <- data.frame()
for (i in unique(res_dream_all$diet)){
  print(i)
  
  for (j in unique(res_dream_all$comparison)){
    dfup <- Getpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "up")$HMDB)))
    dfup$regulation <- "up"
    
  dfdown <- Getpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "down")$HMDB)))
  dfdown$regulation <- "down"
  
  tmpdf <- rbind(dfup, dfdown)
  tmpdf$comparison <- j
  tmpdf$diet <- i
  
  anno_pathways<- rbind(anno_pathways, tmpdf)
  }

}


head(anno_pathways)
```


## 6.3. plots
#### volcano plot
#### ... metabolites to label
```{r}
## label top 20
labels1 <- subset(res_dream_all, regulation != "ns" & !(CompoundName %in% c("Paracetamol sulfate","Dihydroisomorphine-3-glucuronide"))) %>%
  group_by(diet, comparison, direction) %>%
  slice_max(abs(log2FC2), n = 10) %>%
  mutate(label = CompoundName)

labels2 <- subset(res_dream_all, regulation != "ns" & !(CompoundName %in% c("Paracetamol sulfate","Dihydroisomorphine-3-glucuronide"))) %>%
  group_by(diet, comparison, direction) %>%
  slice_min(abs(adj.P.Val), n = 10) %>%
  mutate(label = CompoundName)

labels <- rbind(labels1, labels2)

labels <- as.data.frame(labels)

labels$ionMz_comp_diet <- paste0(labels$ionMz,"_", labels$comparison,"_", labels$diet)

labels <- labels[!duplicated(labels$ionMz_comp_diet),]
# 
# labels_urbandiet_text <- melt_ttest_res_urbandiet %>%
#   group_by(comparison, direction) %>%
#   slice_max(abs(log2FC2), n = 20) %>%
#   mutate(metabo_tranc = stringr::str_trunc(metabolite, 13) ) %>%
#   summarize(label_tranc = paste(metabo_tranc, collapse = ", \n"))

head(labels)

```

```{r fig.width=10, fig.height=10}
plotdf <- res_dream_all

plotdf$diet <- factor(plotdf$diet, levels = c("urban_diet","rural_diet","MBEGE"))

plotdf$ionMz_comp_diet <- paste0(plotdf$ionMz,"_", plotdf$comparison, "_", plotdf$diet)

plotdf$label <- labels[match(plotdf$ionMz_comp_diet, labels$ionMz_comp_diet),"label"]

plotdf$label_tranc <- str_trunc(plotdf$label,24)

plotdf$fill <- plotdf$regulation

plotdf$color3 <- ifelse(is.na(plotdf$label_tranc),plotdf$fill, "black")

plotdf$regulation <- factor(plotdf$regulation, levels = c("up","down","ns"))

volcano.all <- 
  plotdf %>%
  arrange(adj.P.Val)%>%
  ggplot(., aes(x= log2FC2,y= -log10(adj.P.Val), fill = fill))+
  geom_point( size = 1.5, aes(alpha = -adj.P.Val, color = color3), shape = 21)+
  facet_grid(comparison~diet, scales = "free_y")+
  theme_bw()+
  theme(axis.text = element_text())+
  scale_fill_manual(values = c("ns" = "lightgray", "up" = "tomato3", "down" = "cornflowerblue" ), na.value = "lightgray")+
  scale_color_manual(values = c("ns" = "lightgray", "up" = "tomato3", "down" = "cornflowerblue", "black" = "black"), na.value = "lightgray")+
  xlim(-4,4)+geom_hline(yintercept = -log10(0.05), linetype = "dashed")+
  geom_vline(xintercept = c(-log2(fc_threshold), log2(fc_threshold)), linetype = "dashed", color = "black")+
  geom_vline(xintercept = 0)+
  geom_text_repel(aes(label= label_tranc, x= log2FC2, y= -log10(adj.P.Val)) , color = "black", size = 2, max.overlaps = 50)

volcano.all

```
#### traingle plots of top metabolites
```{r}
## label top 5
labels1 <- subset(res_dream_all, regulation != "ns" & !(CompoundName %in% c("Paracetamol sulfate","Dihydroisomorphine-3-glucuronide"))) %>%
  group_by(diet, comparison, direction) %>%
  slice_max(abs(log2FC2), n = 8) %>%
  mutate(label = CompoundName)

labels2 <- subset(res_dream_all, regulation != "ns" & !(CompoundName %in% c("Paracetamol sulfate","Dihydroisomorphine-3-glucuronide"))) %>%
  group_by(diet, comparison, direction) %>%
  slice_min(abs(adj.P.Val), n = 6) %>%
  mutate(label = CompoundName)

labels <- as.data.frame(rbind(labels1, labels2))

#labels <- as.data.frame(labels2)

#labels <- as.data.frame(labels1)

labels$ionMz_comp_diet <- paste0(labels$ionMz,"_", labels$comparison,"_", labels$diet)

labels <- labels[!duplicated(labels$ionMz_comp_diet),]
# 
# labels_urbandiet_text <- melt_ttest_res_urbandiet %>%
#   group_by(comparison, direction) %>%
#   slice_max(abs(log2FC2), n = 20) %>%
#   mutate(metabo_tranc = stringr::str_trunc(metabolite, 13) ) %>%
#   summarize(label_tranc = paste(metabo_tranc, collapse = ", \n"))

head(labels)

## include additional metabolites 

```

```{r fig.width=15, fig.height=8}
plotdf <- subset(res_dream_all, regulation !="ns") 

plotdf$diet <- factor(plotdf$diet, levels = c("urban_diet","rural_diet","MBEGE"))

plotdf$ionMz_comp_diet <- paste0(plotdf$ionMz,"_", plotdf$comparison, "_", plotdf$diet)

labels_manuscript <- c("")

plotdf$label <- labels[match(plotdf$ionMz_comp_diet, labels$ionMz_comp_diet),"label"]

## keep only labels that are not NA
plotdf <- subset(plotdf, !is.na(label))

plotdf$label_tranc <- str_trunc(plotdf$label,24)

plotdf$comparison <- factor(plotdf$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

#plotdf$y <- str_trunc(plotdf$CompoundName,18)

max.val <- round(max(abs(plotdf$log2FC2), na.rm = T),1)*1.2

plotdf$class_name <- factor(plotdf$class_name, levels = unique(plotdf$class_name))



plotdf <- plotdf[order(plotdf$class_name),]

plotdf$class_numeric <- as.numeric(plotdf$class_name)

## replace NA with the highest number
plotdf[is.na(plotdf$class_numeric),"class_numeric"] <- (max(plotdf$class_numeric, na.rm = T)+1)

plotdf <- plotdf[order(plotdf$log2FC2, decreasing = T),]

plotdf$pathway <- anno_pathways[match(plotdf$HMDB, gsub("hmdb:","",anno_pathways$inputId)),"pathwayName"]

plotdf$anno <- ifelse(is.na(plotdf$pathway),as.character(plotdf$class_name), plotdf$pathway)

#plotdf$class_name <- factor(plotdf$class_name, levels = unique(plotdf$class_name))

## ==============
#  colors_class <- pals::polychrome()
# colors_class <- colors_class[c((length(colors_class) - length( unique(plotdf$class_name))+1): length(colors_class))]
# 
# # library(RColorBrewer)
# # n <- length( unique(plotdf$class_name))
# # 
# # qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
# # col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
# 
# # 
 colors_class <- RColorBrewer::brewer.pal(n = 12, "Paired")

 names(colors_class) <- unique(plotdf$class_name)

## ===============

# urban diet
p1_urbandiet <- 
  ggplot(subset(plotdf, diet == "urban_diet") , aes(x= comparison, y= reorder(label_tranc, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 3)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Urban diet")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_urbandiet <-  
  ggplot(subset(plotdf, diet == "urban_diet"), aes(x= "class", y= reorder(label_tranc, -class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())+
  scale_fill_manual(values = colors_class, na.value = "gray88")

## ===============

# rural diet
p1_ruraldiet <-  
  ggplot(subset(plotdf, diet == "rural_diet"), aes(x= comparison, y= reorder(label_tranc, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 3)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Rural diet")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_ruraldiet <-  
  ggplot(subset(plotdf, diet == "rural_diet"), aes(x= "class", y= reorder(label_tranc, -class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())+
  scale_fill_manual(values = colors_class, na.value = "gray88")

## ===============

# Mbege
p1_mbege <- 
  ggplot(subset(plotdf, diet == "MBEGE"), aes(x= comparison, y= reorder(label_tranc, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 3)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Mbege")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_mbege <-  
  ggplot(subset(plotdf, diet == "MBEGE"), aes(x= "class", y= reorder(label_tranc,-class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())+
  scale_fill_manual(values = colors_class, na.value = "gray88")

p2_urbandiet + p1_urbandiet + p2_ruraldiet + p1_ruraldiet + p2_mbege + p1_mbege + plot_layout(nrow = 1, ncol = 6, widths = c(0.1, 3,0.1, 3, 0.1, 3 ), guides = "collect")




```
```{r}
legend_plot <- ggplot(plotdf, aes(x= "class", y= reorder(label_tranc,-class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text(), legend.position =  "right")+
  scale_fill_manual(values = colors_class, na.value = "gray88")

legend_plot
```


```{r fig.width=15, fig.height=8}
plotdf <- subset(res_dream_all, regulation !="ns") 

plotdf$diet <- factor(plotdf$diet, levels = c("urban_diet","rural_diet","MBEGE"))

plotdf$ionMz_comp_diet <- paste0(plotdf$ionMz,"_", plotdf$comparison, "_", plotdf$diet)

labels_manuscript <- c("")

plotdf$label <- labels[match(plotdf$ionMz_comp_diet, labels$ionMz_comp_diet),"label"]

## keep only labels that are not NA
plotdf <- subset(plotdf, !is.na(label))

plotdf$label_tranc <- str_trunc(plotdf$label,24)

plotdf$comparison <- factor(plotdf$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

#plotdf$y <- str_trunc(plotdf$CompoundName,18)

max.val <- round(max(abs(plotdf$log2FC2), na.rm = T),1)*1.2

plotdf$class_name <- factor(plotdf$class_name, levels = unique(plotdf$class_name))



plotdf <- plotdf[order(plotdf$class_name),]

plotdf$class_numeric <- as.numeric(plotdf$class_name)

## replace NA with the highest number
plotdf[is.na(plotdf$class_numeric),"class_numeric"] <- (max(plotdf$class_numeric, na.rm = T)+1)

plotdf <- plotdf[order(plotdf$log2FC2, decreasing = T),]
## ===============

# urban diet
p1_urbandiet <- 
  ggplot(subset(plotdf, diet == "urban_diet") , aes(x= comparison, y= reorder(HMDB, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 4)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Urban diet")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_urbandiet <-  
  ggplot(subset(plotdf, diet == "urban_diet"), aes(x= "class", y= reorder(HMDB, -class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())

## ===============

# rural diet
p1_ruraldiet <-  
  ggplot(subset(plotdf, diet == "rural_diet"), aes(x= comparison, y= reorder(HMDB, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 4)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Rural diet")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_ruraldiet <-  
  ggplot(subset(plotdf, diet == "rural_diet"), aes(x= "class", y= reorder(HMDB, -class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())

## ===============

# Mbege
p1_mbege <- 
  ggplot(subset(plotdf, diet == "MBEGE"), aes(x= comparison, y= reorder(HMDB, -class_numeric)))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 4)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  theme_bw()+ggtitle("Mbege")+
  theme(axis.title = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())

p2_mbege <-  
  ggplot(subset(plotdf, diet == "MBEGE"), aes(x= "class", y= reorder(HMDB,-class_numeric), fill = class_name))+
  geom_tile()+
  theme_void()+
  theme(axis.text.y = element_text())

p2_urbandiet + p1_urbandiet + p2_ruraldiet + p1_ruraldiet + p2_mbege + p1_mbege + plot_layout(nrow = 1, ncol = 6, widths = c(0.1, 3,0.1, 3, 0.1, 3 ), guides = "collect")
```
### DE by class
```{r}
plotdf <- subset(res_dream_all, regulation !="ns") 

## create colors vector -----------------------------------
names_class_all <- c(unique(plotdf$class_name))
names(names_class_all) <- colors_class[match(names_class_all, names(colors_class))]

names(names_class_all)[is.na(names(names_class_all))] <- RColorBrewer::brewer.pal(n = 12, "Set2")

colors_class2 <- names(names_class_all)
names(colors_class2) <- names_class_all


## factor variables ----------------------------------------
plotdf$diet <- factor(plotdf$diet, levels = c("urban_diet","rural_diet","MBEGE"))

ggplot(plotdf, aes(x= diet, fill = class_name))+
  geom_bar(stat = "count", color = "black")+
  scale_fill_manual(values = colors_class2)


ggplot(plotdf, aes(x= diet, fill = class_name))+
  geom_bar(stat = "count", color = "black", position = "fill")+
  scale_fill_manual(values = colors_class2)

ggplot(plotdf, aes(x= diet, fill = class_name))+
  geom_bar(stat = "count", color = "black", position = "fill")+
  scale_fill_manual(values = colors_class2)+
  coord_polar(theta = "y")+
  facet_wrap(.~diet, scales = "free")+
  theme_void()+border()
```


# =====================================
# 7. Controls

## 7.1. Rural controls
### PCA Rural controls
```{r}
sample_table_urbandiet.ctr <- subset(sample_table, diet == "controls_r")
rownames(sample_table_urbandiet.ctr) <- sample_table_urbandiet.ctr$ID

metabo_log2_urbandiet.ctr <-metabo_log2[,match(sample_table_urbandiet.ctr$ID, colnames(metabo_log2))]

identical(sample_table_urbandiet.ctr$ID, colnames(metabo_log2_urbandiet.ctr))

pca.urbandiet.ctr <- plotPCAnew(pca_input = metabo_log2_urbandiet.ctr, 
           ntop = "all", Scale = T,
           xPC = 1, yPC = 2, point_size = 4, 
           anno_colour = col_timepoint, 
           scale = T, anno= sample_table_urbandiet.ctr, 
           color = "timepoint", color.ellipse = T, title = "urban diet controls (rural controls)")+border() + theme_bw()+ theme(aspect.ratio = 1)

pca.urbandiet.ctr
```


## DREAM analysis: controls_r
## Variance partition - rural controls

### ... add info from metadata_new
```{r}
cols_2_add <- c("Age","BMI_t0","activity_sum_rank_scaled","PID")

sample_table_urbandiet.ctr <- merge(sample_table_urbandiet.ctr, metadata_new[,cols_2_add], by = "PID", all.x = T, all.y = F)
rownames(sample_table_urbandiet.ctr) <- sample_table_urbandiet.ctr$ID

head(sample_table_urbandiet.ctr)
```

### step1: find metabolites contributing to more than 5% variation for each time comparison
```{r}
form <- ~ Age + (1 | PID) + timepoint_numeric + BMI_t0 + activity_sum_rank_scaled

varpart_features_urbandiet.ctr <- list()

for ( i in c("t0_vs_t1","t1_vs_t2","t0_vs_t2")){
  tp1 <- str_split_fixed(i,"_vs_",2)[,1]
  tp2 <- str_split_fixed(i,"_vs_",2)[,2]
  
  tmpinfo <- subset(sample_table_urbandiet.ctr, timepoint %in% c(tp1,tp2))
  
  ## make timepoint variable into numeric
  tmpinfo$timepoint_numeric <- as.numeric(gsub("t","",tmpinfo$timepoint))
  
  tmpdata <- metabo_log2_urbandiet.ctr[,match(tmpinfo$ID, colnames(metabo_log2_urbandiet.ctr))]
  
  if(identical(tmpinfo$ID, colnames(tmpdata))){
    print("identical")
  } else{
    stop("error, info and data samples not ordered correctly")
  }
  
  
  tmpvarpart <- fitExtractVarPartModel(tmpdata, form, tmpinfo)
  
  # sort variables (i.e. columns) by median fraction
#       of variance explained
tmpvarpart <- sortCols(tmpvarpart)

varpart_features_urbandiet.ctr[[i]]$varpart <- tmpvarpart
varpart_features_urbandiet.ctr[[i]]$info <- tmpinfo
varpart_features_urbandiet.ctr[[i]]$data <- tmpdata

}
```

### step2: plot results
```{r}
## colors 
colors_vars <- c("Residuals" = "lightgray",
                 "timepoint_numeric"  = "turquoise",
                 "Age" = "purple",
                 "BMI_t0" = "cornflowerblue",
                 "PID" = "lightyellow",
                 "activity_sum_rank_scaled" = "tomato3")
```

```{r}
list_parvar_plots_urbandiet.ctr <- list()

for (i in names(varpart_features_urbandiet.ctr)){
  print(i)
  
tmp_df <- as.data.frame(varpart_features_urbandiet.ctr[[i]]$varpart)
tmp_df$ionMz <- rownames(tmp_df)
tmp_df <- reshape2::melt(tmp_df, id.vars = "ionMz")

timepoint_var <- subset(tmp_df, variable == "timepoint_numeric")$value
names(timepoint_var) <- as.character(subset(tmp_df, variable == "timepoint_numeric")$ionMz)

tmp_df$timepoint <- timepoint_var[match(tmp_df$ionMz, names(timepoint_var))]
tmp_df$value100 <- 100*tmp_df$value

## factor
tmp_df$variable <- factor(tmp_df$variable, levels = rev(c("timepoint_numeric","Age","BMI_t0","activity_sum_rank_scaled", "PID","Residuals")))


tmp_df <- tmp_df[order(tmp_df$timepoint, decreasing = T),]

tmp_df$compoundname <- melt_annotation_tal_new[match(as.character(tmp_df$ionMz), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

# barplot
tmpplot <- ggplot(subset(tmp_df, timepoint >0.05), aes(x = reorder(compoundname, timepoint), y = value100, fill = variable))+
  geom_bar(stat = "identity", position = "stack")+
  coord_flip()+
  scale_fill_manual(values = colors_vars)+
 xlab("")+
  ylab("% Explained variance")+
  ggtitle("t0vst1 Top contributions (>5%) - timepoint")


list_parvar_plots_urbandiet.ctr[["df"]][[i]] <- tmp_df
list_parvar_plots_urbandiet.ctr[["plot"]][[i]] <- tmpplot
}
```
```{r fig.width=20, fig.height=10}
wrap_plots(list_parvar_plots_urbandiet.ctr$plot, nrow = 1, guides = "collect")
```

### step3: extract metabolites contributing >5% for downstream analysis
```{r}
for (i in names(list_parvar_plots_urbandiet.ctr$df)){
  print(i)
  
  tmp_df <- NULL
  
  tmp_df <- list_parvar_plots_urbandiet.ctr$df[[i]]
  
  var_metabo <- unique(tmp_df[tmp_df$timepoint>0.05,"ionMz"])

  print(paste0(round(length(var_metabo)*100/length(unique(tmp_df$ionMz)),digits = 2),"% metabolites taken for analysis for comparison ",i))
  
  
  ## add to varpart_features_urbandiet.ctr
  varpart_features_urbandiet.ctr[[i]]$vars <- var_metabo
}
```



### step4: perform dream analysis
```{r}
list_dream_urbandiet.ctr <- list()

for (i in names(varpart_features_urbandiet.ctr)){
  print(i)
  
  myinfo <- varpart_features_urbandiet.ctr[[i]]$info
  ## factor for the variables present
  myinfo$timepoint <- factor(myinfo$timepoint, levels = unique(myinfo$timepoint))
  
  mydata <- varpart_features_urbandiet.ctr[[i]]$data
  ## filter variables
  mydata <- mydata[which(rownames(mydata) %in% as.character( varpart_features_urbandiet.ctr[[i]]$vars)),]
  
  mycontrasts <- c(
    t0_vs_t1 = "timepointt0 - timepointt1",
    t0_vs_t2 = "timepointt0 - timepointt2",
    t1_vs_t2 = "timepointt1 - timepointt2"
  )
  
  
  L <- makeContrastsDream(~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID) , myinfo,
  contrasts = mycontrasts[names(mycontrasts) == i]
)

# Visualize contrast matrix
plotContrasts(L)

fitmm_dream <- dream(mydata, ~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID), myinfo, L)

fitmm_dream <- eBayes(fitmm_dream)

dream_res <- variancePartition::topTable(fitmm_dream, coef = i, number = Inf, adjust.method = "BH", sort.by = "p", lfc = 0)

dream_res$ionMz <- rownames(dream_res)

print(nrow(dream_res[dream_res$adj.P.Val <=0.05,]))

list_dream_urbandiet.ctr[[i]] <- dream_res
}
```
```{r}
res_dream_urbandiet.ctr <- do.call(rbind, Map(cbind, comparison = names(list_dream_urbandiet.ctr),list_dream_urbandiet.ctr))

head(res_dream_urbandiet.ctr)
```
```{r}
## add sig
res_dream_urbandiet.ctr <- as.data.frame(res_dream_urbandiet.ctr %>% add_significance(p.col = "adj.P.Val"))
```

### step5: calculate paire fc

```{r}
pairedFC_urbandiet.ctr_diet <- data.frame()

for(i in names(varpart_features_urbandiet.ctr)){
  print(i)
  
  ctr_condition = str_split_fixed(i, "_vs_",2)[,1]
  test_condition = str_split_fixed(i, "_vs_",2)[,2]
  
  print(ctr_condition)
  print(test_condition)
  
  ctrids <- rownames(subset(varpart_features_urbandiet.ctr[[i]]$info, 
                            as.character(timepoint) == ctr_condition))
  testids <- rownames(subset(varpart_features_urbandiet.ctr[[i]]$info, 
                             as.character(timepoint) == test_condition))

  
  
  tmpfc <- fcros::fcrosFCmat(varpart_features_urbandiet.ctr[[i]]$data, ctrids, testids, log2.opt = 0)

  tmp_pairedFC <- data.frame(ionMz = tmpfc$idnames, 
                                       FC = as.numeric(tmpfc$FC), 
                                       FC2 = as.numeric(tmpfc$FC2))

  tmp_pairedFC$log2FC <- log2(tmp_pairedFC$FC)
  tmp_pairedFC$log2FC2 <- log2(tmp_pairedFC$FC2)

  tmp_pairedFC$direction <- ifelse(tmp_pairedFC$log2FC2>0,"up",  "down")

  tmp_pairedFC$comparison <- i
  
  pairedFC_urbandiet.ctr_diet <- rbind(pairedFC_urbandiet.ctr_diet, tmp_pairedFC)
}
```
### step6: combine dream and paireFC results

```{r}
### add paired FC
pairedFC_urbandiet.ctr_diet$merged <- paste0(pairedFC_urbandiet.ctr_diet$ionMz,"_", pairedFC_urbandiet.ctr_diet$comparison)

res_dream_urbandiet.ctr$merged <- paste0(res_dream_urbandiet.ctr$ionMz,"_", res_dream_urbandiet.ctr$comparison)

cols_2_add <- c("FC","FC2","log2FC","log2FC2","direction","merged")

res_dream_urbandiet.ctr <- merge(res_dream_urbandiet.ctr, pairedFC_urbandiet.ctr_diet[,cols_2_add], by = "merged")

res_dream_urbandiet.ctr$regulation <- ifelse(res_dream_urbandiet.ctr$adj.P.Val.signif != "ns" & abs(res_dream_urbandiet.ctr$log2FC2) > log2(fc_threshold), res_dream_urbandiet.ctr$direction, "ns")

rownames(res_dream_urbandiet.ctr) <- res_dream_urbandiet.ctr$merged

res_dream_urbandiet.ctr$CompoundName <- melt_annotation_tal_new[match(as.character(res_dream_urbandiet.ctr$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

res_dream_urbandiet.ctr$HMDB <- melt_annotation_tal_new[match(as.character(res_dream_urbandiet.ctr$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundID_new"]

res_dream_urbandiet.ctr$label <- ifelse(res_dream_urbandiet.ctr$adj.P.Val.signif !="ns", str_trunc(res_dream_urbandiet.ctr$CompoundName,24),NA)

head(res_dream_urbandiet.ctr)
```

### step7: list of sig metabolites
```{r}
list_sig_dream_urbandiet.ctr_diet <- list()

for(i in unique(res_dream_urbandiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet.ctr_diet[[i]] <- split(
    subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$ionMz,
     subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet.ctr_diet)
```
```{r}
list_sig_dream_urbandiet.ctr_diet_name <- list()

for(i in unique(res_dream_urbandiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet.ctr_diet_name[[i]] <- split(
    subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$CompoundName,
     subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet.ctr_diet_name)
```
```{r}
list_sig_dream_urbandiet.ctr_diet_hmdb <- list()

for(i in unique(res_dream_urbandiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_urbandiet.ctr_diet_hmdb[[i]] <- split(
    subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$HMDB,
     subset(res_dream_urbandiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_urbandiet.ctr_diet_hmdb)
```

### step8 compare to previous res



### step9: plots
count
```{r}
count_da_urbandiet.ctr <- as.data.frame(res_dream_urbandiet.ctr %>% 
                                      group_by(comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_urbandiet.ctr <- reshape2::melt(count_da_urbandiet.ctr)

count_da_urbandiet.ctr
```

```{r}
count_da_urbandiet.ctr$comparison <- factor(count_da_urbandiet.ctr$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(count_da_urbandiet.ctr, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity",color = "black", position = "stack")+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))+
  theme_bw()+
  ggtitle("urbandiet.ctr")+
  theme(aspect.ratio = 1.5)
```

# --------------------------------
## 6.2. ruraldiet.ctr: Urban controls
### PCA Urban controls
```{r}
sample_table_ruraldiet.ctr <- subset(sample_table, diet == "controls_u")
rownames(sample_table_ruraldiet.ctr) <- sample_table_ruraldiet.ctr$ID

metabo_log2_ruraldiet.ctr <-metabo_log2[,match(sample_table_ruraldiet.ctr$ID, colnames(metabo_log2))]

identical(sample_table_ruraldiet.ctr$ID, colnames(metabo_log2_ruraldiet.ctr))

pca.ruraldiet.ctr <- plotPCAnew(pca_input = metabo_log2_ruraldiet.ctr, 
           ntop = "all", Scale = T,
           xPC = 1, yPC = 2, point_size = 4, 
           anno_colour = col_timepoint, 
           scale = T, anno= sample_table_ruraldiet.ctr, 
           color = "timepoint", color.ellipse = T, title = "urban/mbege diet controls (urban controls)")+border() + theme_bw()+ theme(aspect.ratio = 1)

pca.ruraldiet.ctr
```


## DREAM analysis: controls_u
## Variance partition - urban controls

### ... add info from metadata_new
```{r}
cols_2_add <- c("Age","BMI_t0","activity_sum_rank_scaled","PID")

sample_table_ruraldiet.ctr <- merge(sample_table_ruraldiet.ctr, metadata_new[,cols_2_add], by = "PID", all.x = T, all.y = F)
rownames(sample_table_ruraldiet.ctr) <- sample_table_ruraldiet.ctr$ID

head(sample_table_ruraldiet.ctr)
```

### step1: find metabolites contributing to more than 5% variation for each time comparison
```{r}
form <- ~ Age + (1 | PID) + timepoint_numeric + BMI_t0 + activity_sum_rank_scaled

varpart_features_ruraldiet.ctr <- list()

for ( i in c("t0_vs_t1","t1_vs_t2","t0_vs_t2")){
  tp1 <- str_split_fixed(i,"_vs_",2)[,1]
  tp2 <- str_split_fixed(i,"_vs_",2)[,2]
  
  tmpinfo <- subset(sample_table_ruraldiet.ctr, timepoint %in% c(tp1,tp2))
  
  ## make timepoint variable into numeric
  tmpinfo$timepoint_numeric <- as.numeric(gsub("t","",tmpinfo$timepoint))
  
  tmpdata <- metabo_log2_ruraldiet.ctr[,match(tmpinfo$ID, colnames(metabo_log2_ruraldiet.ctr))]
  
  if(identical(tmpinfo$ID, colnames(tmpdata))){
    print("identical")
  } else{
    stop("error, info and data samples not ordered correctly")
  }
  
  
  tmpvarpart <- fitExtractVarPartModel(tmpdata, form, tmpinfo)
  
  # sort variables (i.e. columns) by median fraction
#       of variance explained
tmpvarpart <- sortCols(tmpvarpart)

varpart_features_ruraldiet.ctr[[i]]$varpart <- tmpvarpart
varpart_features_ruraldiet.ctr[[i]]$info <- tmpinfo
varpart_features_ruraldiet.ctr[[i]]$data <- tmpdata

}
```

### step2: plot results
```{r}
## colors 
colors_vars <- c("Residuals" = "lightgray",
                 "timepoint_numeric"  = "turquoise",
                 "Age" = "purple",
                 "BMI_t0" = "cornflowerblue",
                 "PID" = "lightyellow",
                 "activity_sum_rank_scaled" = "tomato3")
```

```{r}
list_parvar_plots_ruraldiet.ctr <- list()

for (i in names(varpart_features_ruraldiet.ctr)){
  print(i)
  
tmp_df <- as.data.frame(varpart_features_ruraldiet.ctr[[i]]$varpart)
tmp_df$ionMz <- rownames(tmp_df)
tmp_df <- reshape2::melt(tmp_df, id.vars = "ionMz")

timepoint_var <- subset(tmp_df, variable == "timepoint_numeric")$value
names(timepoint_var) <- as.character(subset(tmp_df, variable == "timepoint_numeric")$ionMz)

tmp_df$timepoint <- timepoint_var[match(tmp_df$ionMz, names(timepoint_var))]
tmp_df$value100 <- 100*tmp_df$value

## factor
tmp_df$variable <- factor(tmp_df$variable, levels = rev(c("timepoint_numeric","Age","BMI_t0","activity_sum_rank_scaled", "PID","Residuals")))


tmp_df <- tmp_df[order(tmp_df$timepoint, decreasing = T),]

tmp_df$compoundname <- melt_annotation_tal_new[match(as.character(tmp_df$ionMz), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"]

# barplot
tmpplot <- ggplot(subset(tmp_df, timepoint >0.05), aes(x = reorder(compoundname, timepoint), y = value100, fill = variable))+
  geom_bar(stat = "identity", position = "stack")+
  coord_flip()+
  scale_fill_manual(values = colors_vars)+
 xlab("")+
  ylab("% Explained variance")+
  ggtitle("t0vst1 Top contributions (>5%) - timepoint")


list_parvar_plots_ruraldiet.ctr[["df"]][[i]] <- tmp_df
list_parvar_plots_ruraldiet.ctr[["plot"]][[i]] <- tmpplot
}
```
```{r fig.width=20, fig.height=10}
wrap_plots(list_parvar_plots_ruraldiet.ctr$plot, nrow = 1, guides = "collect")
```

### step3: extract metabolites contributing >5% for downstream analysis
```{r}
for (i in names(list_parvar_plots_ruraldiet.ctr$df)){
  print(i)
  
  tmp_df <- NULL
  
  tmp_df <- list_parvar_plots_ruraldiet.ctr$df[[i]]
  
  var_metabo <- unique(tmp_df[tmp_df$timepoint>0.05,"ionMz"])

  print(paste0(round(length(var_metabo)*100/length(unique(tmp_df$ionMz)),digits = 2),"% metabolites taken for analysis for comparison ",i))
  
  
  ## add to varpart_features_ruraldiet.ctr
  varpart_features_ruraldiet.ctr[[i]]$vars <- var_metabo
}
```



### step4: perform dream analysis
```{r}
list_dream_ruraldiet.ctr <- list()

for (i in names(varpart_features_ruraldiet.ctr)){
  print(i)
  
  myinfo <- varpart_features_ruraldiet.ctr[[i]]$info
  ## factor for the variables present
  myinfo$timepoint <- factor(myinfo$timepoint, levels = unique(myinfo$timepoint))
  
  mydata <- varpart_features_ruraldiet.ctr[[i]]$data
  ## filter variables
  mydata <- mydata[which(rownames(mydata) %in% as.character( varpart_features_ruraldiet.ctr[[i]]$vars)),]
  
  mycontrasts <- c(
    t0_vs_t1 = "timepointt0 - timepointt1",
    t0_vs_t2 = "timepointt0 - timepointt2",
    t1_vs_t2 = "timepointt1 - timepointt2"
  )
  
  
  L <- makeContrastsDream(~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID) , myinfo,
  contrasts = mycontrasts[names(mycontrasts) == i]
)

# Visualize contrast matrix
plotContrasts(L)

fitmm_dream <- dream(mydata, ~ 0 + timepoint + BMI_t0 + Age + activity_sum_rank_scaled + (1|PID), myinfo, L)

fitmm_dream <- eBayes(fitmm_dream)

dream_res <- variancePartition::topTable(fitmm_dream, coef = i, number = Inf, adjust.method = "BH", sort.by = "p", lfc = 0)

dream_res$ionMz <- rownames(dream_res)

print(nrow(dream_res[dream_res$adj.P.Val <=0.05,]))

list_dream_ruraldiet.ctr[[i]] <- dream_res
}
```
```{r}
res_dream_ruraldiet.ctr <- do.call(rbind, Map(cbind, comparison = names(list_dream_ruraldiet.ctr),list_dream_ruraldiet.ctr))

head(res_dream_ruraldiet.ctr)
```
```{r}
## add sig
res_dream_ruraldiet.ctr <- as.data.frame(res_dream_ruraldiet.ctr %>% add_significance(p.col = "adj.P.Val"))
```

### step5: calculate paire fc

```{r}
pairedFC_ruraldiet.ctr_diet <- data.frame()

for(i in names(varpart_features_ruraldiet.ctr)){
  print(i)
  
  ctr_condition = str_split_fixed(i, "_vs_",2)[,1]
  test_condition = str_split_fixed(i, "_vs_",2)[,2]
  
  print(ctr_condition)
  print(test_condition)
  
  ctrids <- rownames(subset(varpart_features_ruraldiet.ctr[[i]]$info, 
                            as.character(timepoint) == ctr_condition))
  testids <- rownames(subset(varpart_features_ruraldiet.ctr[[i]]$info, 
                             as.character(timepoint) == test_condition))

  
  
  tmpfc <- fcros::fcrosFCmat(varpart_features_ruraldiet.ctr[[i]]$data, ctrids, testids, log2.opt = 0)

  tmp_pairedFC <- data.frame(ionMz = tmpfc$idnames, 
                                       FC = as.numeric(tmpfc$FC), 
                                       FC2 = as.numeric(tmpfc$FC2))

  tmp_pairedFC$log2FC <- log2(tmp_pairedFC$FC)
  tmp_pairedFC$log2FC2 <- log2(tmp_pairedFC$FC2)

  tmp_pairedFC$direction <- ifelse(tmp_pairedFC$log2FC2>0,"up",  "down")

  tmp_pairedFC$comparison <- i
  
  pairedFC_ruraldiet.ctr_diet <- rbind(pairedFC_ruraldiet.ctr_diet, tmp_pairedFC)
}
```
### step6: combine dream and paireFC results

```{r}
### add paired FC
pairedFC_ruraldiet.ctr_diet$merged <- paste0(pairedFC_ruraldiet.ctr_diet$ionMz,"_", pairedFC_ruraldiet.ctr_diet$comparison)

res_dream_ruraldiet.ctr$merged <- paste0(res_dream_ruraldiet.ctr$ionMz,"_", res_dream_ruraldiet.ctr$comparison)

cols_2_add <- c("FC","FC2","log2FC","log2FC2","direction","merged")

res_dream_ruraldiet.ctr <- merge(res_dream_ruraldiet.ctr, pairedFC_ruraldiet.ctr_diet[,cols_2_add], by = "merged")

res_dream_ruraldiet.ctr$regulation <- ifelse(res_dream_ruraldiet.ctr$adj.P.Val.signif != "ns" & abs(res_dream_ruraldiet.ctr$log2FC2) > log2(fc_threshold), res_dream_ruraldiet.ctr$direction, "ns")

rownames(res_dream_ruraldiet.ctr) <- res_dream_ruraldiet.ctr$merged

res_dream_ruraldiet.ctr$CompoundName <- melt_annotation_tal_new[match(as.character(res_dream_ruraldiet.ctr$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

res_dream_ruraldiet.ctr$HMDB <- melt_annotation_tal_new[match(as.character(res_dream_ruraldiet.ctr$ionMz), as.character(melt_annotation_tal_new$ionMz)),"CompoundID_new"]

res_dream_ruraldiet.ctr$label <- ifelse(res_dream_ruraldiet.ctr$adj.P.Val.signif !="ns", str_trunc(res_dream_ruraldiet.ctr$CompoundName,24),NA)

head(res_dream_ruraldiet.ctr)
```

### step7: list of sig metabolites
```{r}
list_sig_dream_ruraldiet.ctr_diet <- list()

for(i in unique(res_dream_ruraldiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet.ctr_diet[[i]] <- split(
    subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$ionMz,
     subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet.ctr_diet)
```
```{r}
list_sig_dream_ruraldiet.ctr_diet_name <- list()

for(i in unique(res_dream_ruraldiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet.ctr_diet_name[[i]] <- split(
    subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$CompoundName,
     subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet.ctr_diet_name)
```
```{r}
list_sig_dream_ruraldiet.ctr_diet_hmdb <- list()

for(i in unique(res_dream_ruraldiet.ctr$comparison)){
  
  print(i)
  
  list_sig_dream_ruraldiet.ctr_diet_hmdb[[i]] <- split(
    subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$HMDB,
     subset(res_dream_ruraldiet.ctr, regulation != "ns" & comparison == i)$regulation
    )
  
}

str(list_sig_dream_ruraldiet.ctr_diet_hmdb)
```

### step8 compare to previous res: SKIP



### step9: plots
count
```{r}
count_da_ruraldiet.ctr <- as.data.frame(res_dream_ruraldiet.ctr %>% 
                                      group_by(comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_ruraldiet.ctr <- reshape2::melt(count_da_ruraldiet.ctr)

count_da_ruraldiet.ctr
```

```{r}
count_da_ruraldiet.ctr$comparison <- factor(count_da_ruraldiet.ctr$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(count_da_ruraldiet.ctr, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity",color = "black", position = "stack")+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))+
  theme_bw()+
  ggtitle("ruraldiet.ctr")+
  theme(aspect.ratio = 1.5)
```
## 6.3. Compare diet with ctr

### List DAM all
```{r}

list_dam_all_dream <- list(urban_diet = unlist(list_sig_dream_urbandiet_diet, recursive = F), 
                           rural_diet = unlist(list_sig_dream_ruraldiet_diet, recursive = F), 
                           MBEGE = unlist(list_sig_dream_mbege_diet, recursive = F), 
                           rural_diet_ctr = unlist(list_sig_dream_ruraldiet.ctr_diet, recursive = F), 
                           urban_diet_ctr = unlist(list_sig_dream_urbandiet.ctr_diet, recursive = F))

list_dam_all_dream <- unlist(list_dam_all_dream, recursive = F)

str(list_dam_all_dream)
```


```{r}
dam_all_dream_df <- as.data.frame(do.call(rbind, Map(cbind, comp = names(list_dam_all_dream), list_dam_all_dream)))

dam_all_dream_df$diet <- str_split_fixed(dam_all_dream_df$comp,"\\.", 3)[,1]
dam_all_dream_df$comparison <- str_split_fixed(dam_all_dream_df$comp,"\\.", 3)[,2]
dam_all_dream_df$regulation <- str_split_fixed(dam_all_dream_df$comp,"\\.", 3)[,3]
dam_all_dream_df$treat <- ifelse(grepl("ctr",dam_all_dream_df$diet),"control","diet")

head(dam_all_dream_df)
```

```{r}
dam_all_dream_df$main_diet <- ifelse(dam_all_dream_df$diet == "urban_diet_ctr","urban_diet", 
                                     ifelse(dam_all_dream_df$diet == "rural_diet_ctr", "rural_diet", 
                                            dam_all_dream_df$diet))

unique(dam_all_dream_df$main_diet)
```
```{r}
## add compoundName
dam_all_dream_df$CompoundName <- melt_annotation_tal_new[match(as.character(dam_all_dream_df$V2), as.character(melt_annotation_tal_new$ionMz)),"CompoundName"]

dam_all_dream_df$CompoundName[1:5]
```
```{r}
## urban diet
tmp <- subset(dam_all_dream_df, main_diet == "urban_diet")

tmp_sum <- as.data.frame(tmp %>% 
                           group_by(CompoundName, diet, comparison) %>% 
                           summarize(diet.count = sum(treat == "diet"), 
                                     ctr.count = sum(treat == "control")))

## keep only rows with ctr == 1
tmp_sum_urbandiet <- subset(tmp_sum, ctr.count == 1 & diet.count == 1)
dim(tmp_sum_urbandiet)
```
```{r}
## rural diet
tmp <- subset(dam_all_dream_df, main_diet == "rural_diet")

tmp_sum <- as.data.frame(tmp %>% 
                           group_by(CompoundName, diet, comparison) %>% 
                           summarize(diet.count = sum(treat == "diet"), 
                                     ctr.count = sum(treat == "control")))
## keep only rows with ctr == 1
tmp_sum_ruraldiet <- subset(tmp_sum, ctr.count == 1 & diet.count == 1)
dim(tmp_sum_ruraldiet)
```
```{r}
## Mbege
tmp <- subset(dam_all_dream_df, diet %in% c("MBEGE","rural_diet_ctr"))

tmp_sum <- as.data.frame(tmp %>% 
                           group_by(CompoundName, diet, comparison) %>% 
                           summarize(diet.count = sum(treat == "diet"), 
                                     ctr.count = sum(treat == "control")))
## keep only rows with ctr == 1
tmp_sum_mbege <- subset(tmp_sum, ctr.count == 1 & diet.count == 1)
dim(tmp_sum_mbege)
```
### .. upset plots
### ... > urbandiet
```{r}
tmp <- list_dam_all_dream[grepl("urban_diet", names(list_dam_all_dream))]

tmp2 <- tmp[grepl("t0_vs_t1", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t0_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t1_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)
```
### ... > ruraldiet
```{r}
tmp <- list_dam_all_dream[grepl("rural_diet", names(list_dam_all_dream))]

tmp2 <- tmp[grepl("t0_vs_t1", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t0_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t1_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)
```
### ... > mbege
```{r}
tmp <- list_dam_all_dream[grepl("rural_diet_ctr|MBEGE", names(list_dam_all_dream))]

tmp2 <- tmp[grepl("t0_vs_t1", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t0_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)

tmp2 <- tmp[grepl("t1_vs_t2", names(tmp))]
upset(fromList(tmp2), sets = names(tmp2), keep.order = T)
```



`

### mark in res_dream_all DAM in control

```{r}
res_dream_all$DAM_in_ctr <- NA
res_dream_all$DAM_in_ctr_up <- NA

for(i in unique(res_dream_all$diet)){
  print(i)
  
  for (j in unique(res_dream_all$comparison)){
    print(j)
    
    if(i == "MBEGE"){
      lookup <- paste0("rural_diet","_ctr.",j)
    } else{
      lookup <- paste0(i, "_ctr.",j)
    }
    
    dam_ctr <-  as.character(unique(unlist(list_dam_all_dream[grepl(lookup, names(list_dam_all_dream))])))
    
    dam_ctr_up <- as.character(unique(unlist(list_dam_all_dream[grepl(paste0(lookup,".up"), names(list_dam_all_dream))])))

    
    res_dream_all[which(res_dream_all$diet == i & res_dream_all$comparison == j & res_dream_all$regulation != "ns"), "DAM_in_ctr"] <- ifelse(as.character(res_dream_all[which(res_dream_all$diet == i & res_dream_all$comparison == j & res_dream_all$regulation != "ns"), "ionMz"]) %in% dam_ctr, "yes","")
    
    res_dream_all[which(res_dream_all$diet == i & res_dream_all$comparison == j & res_dream_all$regulation != "ns"), "DAM_in_ctr_up"] <- ifelse(res_dream_all[which(res_dream_all$diet == i & res_dream_all$comparison == j & res_dream_all$regulation != "ns"), "DAM_in_ctr"] == "yes" &  as.character( res_dream_all[which(res_dream_all$diet == i & res_dream_all$comparison == j & res_dream_all$regulation != "ns"), "ionMz"]) %in% dam_ctr_up, "up","") 
    
  }
}
```
```{r}
res_dream_all$DAM_in_ctr_direction <- ifelse(res_dream_all$DAM_in_ctr == "yes" & res_dream_all$DAM_in_ctr_up == "","down", res_dream_all$DAM_in_ctr_up)
## which values are NA or "" will be ns

res_dream_all$DAM_in_ctr_direction <- ifelse(is.na(res_dream_all$DAM_in_ctr_direction)| res_dream_all$DAM_in_ctr_direction =="","ns", res_dream_all$DAM_in_ctr_direction)

unique(res_dream_all$DAM_in_ctr_direction )

```
add log2FC2
```{r}

```

### ... plot intersection with control
```{r}
metabo_in_ctr <- unique(subset(res_dream_all, DAM_in_ctr_direction != "ns")$HMDB)
plot_intersect <- subset(res_dream_all, HMDB %in% metabo_in_ctr)

plot_intersect <- reshape2::melt(plot_intersect[,c("comparison", "diet","regulation","DAM_in_ctr_direction","CompoundName")], id.vars = c("comparison", "diet","CompoundName" ))

plot_intersect$V1 <- ifelse(plot_intersect$variable == "regulation", "diet", "ctr")

head(plot_intersect)
```
```{r fig.width=10, fig.height=6}
plot_intersect$comparison <- factor(plot_intersect$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

plot_intersect <- subset(plot_intersect, value !="ns")
# plot_intersect <- subset(plot_intersect, !(value %in% c("","ns")))
# 
plot_intersect <- as.data.frame(plot_intersect %>% group_by(diet, CompoundName) %>% mutate(sum_ctr = sum(V1 == "ctr")))
# 
# plot_intersect <- subset(plot_intersect, sum_indiet >1)

#plot_intersect <- plot_intersect[!is.na(plot_intersect$value),]

ggplot(subset(plot_intersect, sum_ctr >0), aes(x= V1, y = CompoundName, shape = value, fill = value))+
  geom_point(size = 2)+
  scale_shape_manual(values = c("up" = 24, "down" = 25) )+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue", "ns" = "white"))+
  facet_grid(diet ~ comparison, scales = "free", space = "free")+
  theme(axis.title = element_blank())+
  ggtitle("DAM in controls")
```

### plot intersect with control including the FC
```{r}
intersect_metabo <- unique(subset(plot_intersect, sum_ctr >0)$CompoundName)

res_dream_ruraldiet.ctr$diet <- "rural_diet_ctr"
res_dream_urbandiet.ctr$diet <- "urban_diet_ctr"

tmp_intersect <- rbind(all_sig_df[,c("diet","comparison","log2FC2","regulation","CompoundName","adj.P.Val")],res_dream_ruraldiet.ctr[,c("diet","comparison","log2FC2","regulation","CompoundName","adj.P.Val")], res_dream_urbandiet.ctr[,c("diet","comparison","log2FC2","regulation","CompoundName","adj.P.Val")])

head(tmp_intersect)
```

```{r,  fig.width=10, fig.height=6}
tmp_intersect$diet_comparison <- paste0(tmp_intersect$diet,"_",tmp_intersect$comparison)

tmp_intersect <- as.data.frame(tmp_intersect %>% mutate(
  group = case_when(
    diet == "urban_diet" ~ "urabn_diet",
    diet == "urban_diet_ctr" ~ "urban_diet", 
    diet == "rural_diet" ~ "rural_diet", 
    diet == "rural_diet_ctr" ~ "rural_diet",
    diet == "MBEGE" ~ "mbege"
  )
))

tmp_intersect <- as.data.frame(tmp_intersect %>% mutate(
  treatment = case_when(
    diet %in% c("urban_diet","rural_diet","MBEGE") ~ "diet",
    diet %in% c("urban_diet_ctr","rural_diet_ctr") ~ "ctr"
  )
))

tmp_intersect$label <- str_trunc(tmp_intersect$CompoundName,24)

ggplot(subset(tmp_intersect, group == "urban_diet" & regulation != "ns"), aes(x= treatment, y= label, shape = regulation, fill = log2FC2))+
  geom_point(size = 2)+
  scale_shape_manual(values = c("up" = 24, "down" = 25) )+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  facet_wrap(. ~ comparison,  nrow = 1)+
  theme(axis.title = element_blank())+
  ggtitle("DAM in controls")

```
```{r fig.height=10}
ggplot(subset(tmp_intersect, group == "rural_diet" & regulation != "ns"), aes(x= treatment, y= label, shape = regulation, fill = log2FC2))+
  geom_point(size = 2)+
  scale_shape_manual(values = c("up" = 24, "down" = 25) )+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  facet_wrap(. ~ comparison,  nrow = 1)+
  theme(axis.title = element_blank())+
  ggtitle("DAM in controls")
```

```{r fig.height=20}
tmp_intersect$merged <- paste0(tmp_intersect$CompoundName,"_", tmp_intersect$comparison)

tmp_intersect$label <- str_trunc(tmp_intersect$CompoundName, 24)

tmp_intersect <- as.data.frame(tmp_intersect %>% group_by(merged) %>% mutate(count = n()) )

ggplot(subset(tmp_intersect, count >2 & diet != "MBEGE"), aes(x= comparison, y = label, shape = regulation, fill = log2FC2))+
  geom_point(size = 2)+
  scale_shape_manual(values = c("up" = 24, "down" = 25) )+
  facet_grid(diet ~ comparison, scales = "free", space = "free")+
  theme(axis.title = element_blank())+
  ggtitle("DAM in controls")
```


# ===============================

## 



# ------------------------
## 7.2. pathways
#### ... annotation
```{r}
anno_pathways <- data.frame()
for (i in unique(res_dream_all$diet)){
  print(i)
  
  for (j in unique(res_dream_all$comparison)){
    dfup <- Getpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "up")$HMDB)))
    dfup$regulation <- "up"
    
  dfdown <- Getpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "down")$HMDB)))
  dfdown$regulation <- "down"
  
  tmpdf <- rbind(dfup, dfdown)
  tmpdf$comparison <- j
  tmpdf$diet <- i
  
  anno_pathways<- rbind(anno_pathways, tmpdf)
  }

}


head(anno_pathways)
```

```{r}
#write.csv(anno_pathways,"./240913_annotation_metabolites_pathway.csv")
```

```{r}
# enrich_pathways$count <- apply(enrich_pathways, 1, function(x){
#  length(unlist(str_split(x[["commonName"]],",")))
# })
```

```{r}
pathway_count <- as.data.frame(anno_pathways %>% group_by(diet, comparison, regulation, pathwayName) %>% summarize(count = n()))

head(pathway_count)
```
```{r}
pathway_count$pathwayName_low <- tolower(pathway_count$pathwayName)

pathway_count$merged <- paste0(pathway_count$pathwayName_low,"_", pathway_count$comparison,"_", pathway_count$regulation)
```

```{r fig.height=8, fig.width=15}
pathway_count$diet <- factor(pathway_count$diet, levels = c("urban_diet","rural_diet","MBEGE"))

pathway_count$comparison <- factor(pathway_count$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(pathway_count, count >3), aes(x= regulation, y= pathwayName))+
  geom_point(aes(size = count, fill = regulation), shape = 21)+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  facet_wrap(diet~comparison, nrow = 1)+
  scale_size_continuous(range = c(3,8))
```

#### ... fisher pathway
#### .... > enrich by direction
```{r}
fisher_pathways <- data.frame()

for (i in unique(res_dream_all$diet)){
  print(i)
  
  for (j in unique(res_dream_all$comparison)){
    dfup <- Getfisherpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "up")$HMDB)))$fishresults
    dfup$regulation <- "up"
    if(length(dfup)>0){
    dfup$regulation <- "up"
  } else {
    dfup <- NULL
  }
    
  dfdown <- Getfisherpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation == "down")$HMDB)))$fishresults
  if(length(dfdown)>0){
    dfdown$regulation <- "down"
  } else {
    dfdown <- NULL
  }
  
  # dfall <- Getfisherpathway_RAMPdb(hmdbinput = paste0("hmdb:", unique(subset(res_dream_all, diet == i & comparison == j & regulation != "ns")$HMDB)))$fishresults
  #   dfall$regulation <- "all"
  #   if(length(dfall)>0){
  #   dfall$regulation <- "all"
  # } else {
  #   dfall <- NULL
  # }
  
  
  tmpdf <- rbind(dfup, dfdown)
  tmpdf$comparison <- j
  tmpdf$diet <- i
  
  fisher_pathways<- rbind(fisher_pathways, tmpdf)
  }

}


head(fisher_pathways)


```

```{r}
fisher_pathways$pathwayName_low <- tolower(fisher_pathways$pathwayName)

fisher_pathways$merged <- paste0(fisher_pathways$diet, "_",fisher_pathways$pathwayName_low,"_", fisher_pathways$comparison,"_", fisher_pathways$regulation)

fisher_pathways_sig <- subset(fisher_pathways, Pval_FDR <= 0.05)

```


```{r fig.height=6, fig.width=15}
fisher_pathways_sig$diet <- factor(fisher_pathways_sig$diet, levels = c("urban_diet","rural_diet","MBEGE"))

fisher_pathways_sig$comparison <- factor(fisher_pathways_sig$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

top_8_enrichpath <- as.data.frame(subset(fisher_pathways_sig,pathwaySource %in% c("wiki"))  %>% group_by(diet, comparison, regulation) %>% slice_min(abs(Pval_FDR), n = 8))

ggplot(subset(top_8_enrichpath, regulation != "all"), 
       aes(x= regulation, y= pathwayName))+
  geom_point(aes(size = Num_In_Path, fill = regulation), shape = 21)+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  facet_wrap(diet~comparison, nrow = 1)+
   scale_size_continuous(range = c(2,8))+
  theme_bw()+
  theme(axis.title = element_blank())
```
```{r fig.height=6, fig.width=15}
fisher_pathways_sig$diet <- factor(fisher_pathways_sig$diet, levels = c("urban_diet","rural_diet","MBEGE"))

fisher_pathways_sig$comparison <- factor(fisher_pathways_sig$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

top_8_enrichpath <- as.data.frame(subset(fisher_pathways_sig,pathwaySource  == "kegg")  %>% group_by(diet, comparison, regulation) %>% slice_min(abs(Pval_FDR), n = 8))

ggplot(subset(top_8_enrichpath, regulation != "all"), 
       aes(x= regulation, y= pathwayName))+
  geom_point(aes(size = Num_In_Path, fill = regulation), shape = 21)+
  scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
  facet_wrap(diet~comparison, nrow = 1)+
   scale_size_continuous(range = c(2,8))+
  theme_bw()+
  theme(axis.title = element_blank())
```

```{r}
# ggplot(subset(top_8_enrichpath, regulation == "all"), 
#        aes(x= regulation, y= pathwayName))+
#   geom_point(aes(size = Num_In_Path, x= comparison, fill = Pval_FDR), shape = 21)+
#   #scale_fill_manual(values = c("up" = "tomato3","down" = "cornflowerblue"))+
#   facet_wrap(.~diet, nrow = 1)+
#    scale_size_continuous(range = c(2,8))+
#   theme_bw()+
#   theme(axis.title = element_blank())+
#   scale_fill_gradientn(colors = c("orange","red","blue"))
```

```{r}
#write.csv(plotdf, "../Figures/Fig. Metabolomics/Revision/240823_triangleplot_info_metabolomics.csv")
```

##### ... > hm enrichment
```{r}
top_4_enrichpath <- as.data.frame(subset(fisher_pathways_sig,pathwaySource  == "wiki")  %>% group_by(diet, comparison, regulation) %>% slice_min(abs(Pval_FDR), n = 8))

hmdf_enrich <- top_4_enrichpath

hmdf_enrich<- hmdf_enrich[!duplicated(hmdf_enrich$merged),]

#wide_anno <- top_8_enrichpath
wide_anno <- data.frame()

wide_anno <- wide_anno[]
for(i in unique(hmdf_enrich$merged)){
  
  tmp <- str_split_fixed(hmdf_enrich[hmdf_enrich$merged == i,]$analytes,
                         ";", 
                         n = hmdf_enrich[hmdf_enrich$merged == i,]$Num_In_Path)
  
  tmp <- as.character(tmp[1,])
  
  tmp <- unlist(lapply(sapply(tmp, str_split,","),"[",1))
  
  tmp <- data.frame(metbolite = tmp, merged = i)
  
  wide_anno<- rbind(wide_anno, tmp)
  
}

head(wide_anno)
```
```{r}
wide_anno$diet <- hmdf_enrich[match(wide_anno$merged, hmdf_enrich$merged),"diet"] 

wide_anno$pathway <-hmdf_enrich[match(wide_anno$merged, hmdf_enrich$merged),"pathwayName"] 

wide_anno$comparison <- hmdf_enrich[match(wide_anno$merged, hmdf_enrich$merged),"comparison"] 

wide_anno$regulation <- hmdf_enrich[match(wide_anno$merged, hmdf_enrich$merged),"regulation"] 
```

```{r}
## add unique ID for metabolite and get log2FC2 values
wide_anno$HMDB <- melt_annotation_tal_new[match(wide_anno$metbolite, melt_annotation_tal_new$CompoundName), ]$CompoundID_new

wide_anno[is.na(wide_anno$HMDB),c("metbolite","merged")]
```
```{r}
## some metabolites were not found and will be inputted manuall
wide_anno$CompoundName <- NA

wide_anno[wide_anno$metbolite == "Ferulic acid 4-O-sulfate","CompoundName"] <- "Ferulic acid 4-sulfate"

wide_anno[wide_anno$metbolite == "alpha-Linolenic acid","CompoundName"] <- "Alpha-Linolenic acid"

wide_anno[wide_anno$metbolite == "O5;Arachidonic acid","CompoundName"] <- "Arachidonic acid"

wide_anno[wide_anno$metbolite == "Malic acid","CompoundName"] <- "L-Malic acid"

wide_anno[wide_anno$metbolite == "LysoPA(0:0/16:0);LysoPA(0:0/18:0)","CompoundName"] <- "LPA(0:0/16:0)"

wide_anno[is.na(wide_anno$HMDB),"HMDB"] <- melt_annotation_tal_new[match(wide_anno[is.na(wide_anno$HMDB),"CompoundName"], melt_annotation_tal_new$CompoundName), ]$CompoundID_new

wide_anno[is.na(wide_anno$HMDB),c("metbolite","merged")]
```
```{r}
wide_anno$uniqueID <- paste0(wide_anno$HMDB,"_", wide_anno$diet,"_", wide_anno$comparison,"_", wide_anno$regulation)

all_sig_df$merged <- paste0(all_sig_df$HMDB,"_", all_sig_df$diet,"_", all_sig_df$comparison,"_", all_sig_df$regulation)

wide_anno$log2FC2 <- all_sig_df[match(wide_anno$uniqueID, all_sig_df$merged),"log2FC2"]

wide_anno$log2FC2 [1:10]
```
```{r}
wide_anno$label <- ifelse(is.na(wide_anno$metbolite), wide_anno$CompoundName, wide_anno$metbolite)
```



```{r fig.height=12, fig.width=15}
## replace NA with 0
wide_anno[is.na(wide_anno$log2FC2),"log2FC2"] <- 0

ggplot(wide_anno, aes(x = label, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colors = c("blue","white","red"))+
  facet_grid(comparison ~ diet, scales = "free", space = "free")+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r fig.width=10, fig.height=6}
ggplot(subset(wide_anno, diet == "urban_diet"), aes(x = label, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colors = c("blue","white","red"))+
  facet_grid(comparison~ ., scales = "free", space = "free")+
  theme(axis.text.x =  element_text(angle = 90))
```
```{r fig.height=4, fig.width=8}
dfx <- subset(wide_anno, pathway %in% c("Biomarkers for urea cycle disorders","Glucose homeostasis","Flavan−3−ol metabolic pathway","Omega−3 / omega−6 fatty acid synthesis","Prostaglandin and leukotriene metabolism in senescence") & diet == "urban_diet")


max.val <- round(max(abs(dfx$log2FC2), na.rm = T),1)*1.2

p.enrich.urban <- ggplot(dfx, aes(x = label, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black") )+
  facet_grid(comparison~ ., scales = "free", space = "free")+
  theme(axis.text.x =  element_text(angle = 40, hjust = 1), axis.title = element_blank())+ggtitle("Urban diet")

p.enrich.urban
```
```{r fig.height=3, fig.width=8}
dfx <- subset(wide_anno, pathway %in% c("Incretin synthesis, secretion, and inactivation","Omega-9 fatty acid synthesis","Omega−3 / omega−6 fatty acid synthesis","Neurotransmitter release cycle") & diet == "rural_diet")

max.val <- round(max(abs(dfx$log2FC2), na.rm = T),1)*1.2

p.enrich.rural <- ggplot(dfx, aes(x = label, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black") )+
  facet_grid(comparison~ ., scales = "free", space = "free")+
  theme(axis.text.x =  element_text(angle = 40, hjust = 1), axis.title = element_blank())+ggtitle("rural diet")

p.enrich.rural
```

```{r fig.height=4, fig.width=8}
dfx <- subset(wide_anno, pathway %in% c("Incretin synthesis, secretion, and inactivation","Omega-9 fatty acid synthesis","Omega−3 / omega−6 fatty acid synthesis","Neurotransmitter release cycle") & diet == "rural_diet")

max.val <- round(max(abs(dfx$log2FC2), na.rm = T),1)*1.2

p.enrich.rural <- ggplot(dfx, aes(x = label, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black") )+
  facet_grid(comparison~ ., scales = "free", space = "free")+
  theme(axis.text.x =  element_text(angle = 40, hjust = 1), axis.title = element_blank())+ggtitle("rural diet")

p.enrich.rural
```

```{r fig.height=4, fig.width=8}
dfx <- subset(wide_anno, pathway %in% c("Flavan−3−ol metabolic pathway","Omega−3 / omega−6 fatty acid synthesis","Regulation of lipid metabolism by PPARalpha","Activation of gene expression by SREBF (SREBP)") & diet == "MBEGE")

dfx$labelx_x <- str_trunc(dfx$label,23)

max.val <- round(max(abs(dfx$log2FC2), na.rm = T),1)*1.2

p.enrich.mbege <- ggplot(dfx, aes(x = labelx_x, y = pathway, fill = log2FC2))+
  geom_tile(color = "black")+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black") )+
  facet_grid(comparison~ ., scales = "free", space = "free")+
  theme(axis.text.x =  element_text(angle = 40, hjust = 1), axis.title = element_blank())+ggtitle("Mbege")

p.enrich.mbege
```


# =======================================

# 8. Export

## 8.1. EXPORT results
```{r}
export_cols <- c("CompoundName","HMDB", "ionMz","class_name", "comparison","diet","log2FC","log2FC2","AveExpr","t","P.Value","adj.P.Val","adj.P.Val.signif", "regulation","DAM_in_ctr")

export_file <- res_dream_all[,export_cols]

## controls
tmp1 <- res_dream_ruraldiet.ctr
tmp1$diet <- "controls_u"
tmp2 <- res_dream_urbandiet.ctr
tmp2$diet <- "controls_r"
res_dream_ctr <- rbind(tmp1,tmp2)
res_dream_ctr$DAM_in_ctr <- ""
res_dream_ctr$class_name <- rampdb__all_class_ClassyFire_class[match(res_dream_ctr$HMDB, rampdb__all_class_ClassyFire_class$HMDB),"class_name"]
res_dream_ctr <- res_dream_ctr[,export_cols]


export_file <- rbind(export_file, res_dream_ctr)

colnames(export_file) <- c("metabolite","HMDB_id","ionMz","RAMPDb_chemical_class","comparison","diet","log2FC","log2FC2","AveExpr","t","P.Value","adj.P.Val","adj.P.Val.signif", "regulation","DAM_in_ctr")

export_file$Diet <- ifelse(export_file$diet == "urban_diet", "Western Diet",
                                          ifelse(export_file$diet == "rural_diet","Traditional Diet", 
                                          ifelse(export_file$diet == "MBEGE" ,"Fermented banana beverage",
                                          ifelse(export_file$diet== "controls_r", "Rural Dwellers Control",
                                          ifelse(export_file$diet == "controls_u" , "Urban Dwellers Control", NA)))))

export_file$diet <- NULL

export_file$Diet <- factor(export_file$Diet, levels = c("Western Diet","Rural Dwellers Control","Traditional Diet","Fermented banana beverage","Urban Dwellers Control"))  

export_file$DAM_in_ctr <- factor(export_file$DAM_in_ctr, levels = c("","yes"))

export_file$regulation <- factor(export_file$regulation, levels = c("up","down","ns"))


export_file <- export_file[order(export_file$Diet, export_file$comparison, export_file$regulation, export_file$adj.P.Val,  decreasing = F),]

#write.xlsx(export_file, paste0("./",Sys.Date(),"_metabolomics_DAM.xlsx"), asTable = T)

```

## 8.2. Export list ionMz DAM
```{r}

# write.csv(subset(res_dream_all, regulation != "ns"), "../multiomics/Data/240909_sig_metabolites_dream.csv")
```

## 8.3. export enrichment pathway RAMP-DB results
```{r}
export_fisher <- subset(fisher_pathways_sig, pathwaySource == "wiki")

export_fisher<- export_fisher %>%
  mutate(Diet = case_when(
    diet == "urban_diet" ~ "Western Diet",
    diet == "rural_diet" ~ "Traditional Diet",
    diet == "MBEGE" ~ "Fermented banana beverage", 
    diet == "controls_u" ~ "Urban Dwellers Control", 
    diet == "controls_r" ~ "Rural Dwellers Control"
  ))

export_fisher$diet <- NULL

export_fisher$Diet <- factor(export_fisher$Diet, levels = c("Western Diet", 
                                                        "Traditional Diet", 
                                                        "Fermented banana beverage", 
                                                        "Urban Dwellers Control", 
                                                        "Rural Dwellers Control"))

export_fisher$regulation <- factor(export_fisher$regulation, levels = c("up","down","ns"))




export_fisher <- export_fisher[order(export_fisher$Diet, export_fisher$comparison, export_fisher$regulation, export_fisher$Pval_FDR, decreasing = F),]


head(export_fisher)
#write.xlsx(subset(export_fisher, pathwaySource == "wiki"), paste0("./", Sys.Date(), "_Rampdb_fisher_wiki.xlsx"), asTable = T)
```


# ----------------------
<!-- ## 6.4. ggraph network -->
<!-- ### 6.4.1. create network_df -->
<!-- ```{r} -->
<!-- export_rampdb_sig_all_class <- export_rampdb_sig_all_class %>% group_by(class_name) %>% mutate(count_class = n()) -->
<!-- export_rampdb_sig_all_class <- as.data.frame(export_rampdb_sig_all_class) -->

<!-- ## keep only pathways with more than 3 metabolites -->
<!-- network_df <- export_rampdb_sig_all_class[,c("HMDB","class_name","count_class")] -->

<!-- network_df$class_name_new <- ifelse(network_df$count_class >=4, network_df$class_name, NA) -->

<!-- ## see how many pathways each metabolite has -->
<!-- network_df <- network_df %>% group_by(HMDB) %>% mutate(count_path = n_distinct(class_name_new)) -->


<!-- ## for metabolite with count_path = 1 & class_name_new = NA --> these should get class "other" -->

<!-- network_df$class_plot <- ifelse(network_df$count_path == 1 & is.na(network_df$class_name_new), "other", network_df$class_name_new) -->

<!-- ## remove rows with class_plot == NA -->
<!-- network_df <- network_df[!is.na(network_df$class_plot),] -->

<!-- ``` -->

<!-- ```{r} -->
<!-- ## check others -->
<!-- others_HMDB <-as.character(network_df[network_df$class_plot == "other",]$HMDB) -->

<!-- others <- network_df[which(network_df$HMDB %in% others_HMDB),] -->
<!-- ``` -->

<!-- ### 6.4.2. Network to each diet -->

<!-- ##### Urban diet -->
<!-- #### ... get comparison results table and transform to wide format -->
<!-- ```{r} -->
<!-- sig_mix.mod_urbandiet_df <- summary_table_mix.mod_urbandiet$summary_table -->

<!-- sig_mix.mod_urbandiet_df <- subset(sig_mix.mod_urbandiet_df, regulation != "ns") -->

<!-- sig_mix.mod_urbandiet_df$HMDB <- melt_annotation_tal_new[match(as.character(sig_mix.mod_urbandiet_df$ionMZ), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"] -->

<!-- dim(sig_mix.mod_urbandiet_df) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- sig_mix.mod_urbandiet_df_wide <- as.data.frame(spread(sig_mix.mod_urbandiet_df[,c("HMDB","log2FC2","comparison")], key = "comparison", value = "log2FC2" )) -->

<!-- rownames(sig_mix.mod_urbandiet_df_wide) <- sig_mix.mod_urbandiet_df_wide$HMDB -->

<!-- sig_mix.mod_urbandiet_df_wide$HMDB <- NULL -->

<!-- # change order of cols -->

<!-- sig_mix.mod_urbandiet_df_wide <- sig_mix.mod_urbandiet_df_wide[,c("t0_vs_t1","t1_vs_t2","t0_vs_t2")] -->

<!-- head(sig_mix.mod_urbandiet_df_wide) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # 1. add fold changes of mbege -->
<!-- network_df.urbandiet <- merge(network_df, sig_mix.mod_urbandiet_df_wide,  -->
<!--                               by.x = "HMDB", by.y = "row.names", all.x = T) -->

<!-- # 2. create a data frame for edges -->
<!-- net_edges.urbandiet <- network_df.urbandiet[,c("HMDB","class_plot")] -->
<!-- ## re-name colnames to from and to -->
<!-- colnames(net_edges.urbandiet) <- c("from","to") -->

<!-- # 3. Create a data frame for vertices -->
<!-- net_vertices.urbandiet <- data.frame(name = c(network_df.urbandiet$HMDB,  -->
<!--                                     network_df.urbandiet$class_plot),  -->
<!--                         class_count = c(rep(NA, length(network_df.urbandiet$HMDB)), -->
<!--                                         network_df.urbandiet$count_path), -->
<!--                        t0_vs_t1 = c(network_df.urbandiet$t0_vs_t1,  -->
<!--                                     rep(0, length(network_df.urbandiet$class_plot))),  -->
<!--                        t1_vs_t2 = c(network_df.urbandiet$t1_vs_t2,  -->
<!--                                     rep(0, length(network_df.urbandiet$class_plot))),  -->
<!--                        t0_vs_t2 = c(network_df.urbandiet$t0_vs_t2,  -->
<!--                                     rep(0, length(network_df.urbandiet$class_plot))), -->
<!--                        type = c(rep("metabolite", length(network_df.urbandiet$HMDB)),  -->
<!--                                 rep("pathway", length(network_df.urbandiet$class_plot)))) -->

<!-- ## remove duplicated names -->
<!-- net_vertices.urbandiet <- net_vertices.urbandiet[!duplicated(net_vertices.urbandiet$name),] -->

<!-- ## factor type -->
<!-- net_vertices.urbandiet$type <- factor(net_vertices.urbandiet$type, levels = c("pathway","metabolite")) -->


<!-- ## label only pathway -->
<!-- net_vertices.urbandiet$mylabel <- ifelse(net_vertices.urbandiet$type == "pathway", -->
<!--                                          net_vertices.urbandiet$name, "") -->

<!-- head(net_vertices.urbandiet) -->
<!-- ```  -->
<!-- ```{r} -->
<!-- nrow(net_edges.urbandiet[net_edges.urbandiet$to == "other",]) -->
<!-- ``` -->

<!-- ##### Rural diet -->
<!-- #### ... get comparison results table and transform to wide format -->
<!-- ```{r} -->
<!-- sig_mix.mod_ruraldiet_df <- summary_table_mix.mod_ruraldiet$summary_table -->

<!-- sig_mix.mod_ruraldiet_df <- subset(sig_mix.mod_ruraldiet_df, regulation != "ns") -->

<!-- sig_mix.mod_ruraldiet_df$HMDB <- melt_annotation_tal_new[match(as.character(sig_mix.mod_ruraldiet_df$ionMZ), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"] -->

<!-- dim(sig_mix.mod_ruraldiet_df) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- sig_mix.mod_ruraldiet_df_wide <- as.data.frame(spread(sig_mix.mod_ruraldiet_df[,c("HMDB","log2FC2","comparison")], key = "comparison", value = "log2FC2" )) -->

<!-- rownames(sig_mix.mod_ruraldiet_df_wide) <- sig_mix.mod_ruraldiet_df_wide$HMDB -->

<!-- sig_mix.mod_ruraldiet_df_wide$HMDB <- NULL -->

<!-- # change order of cols -->

<!-- sig_mix.mod_ruraldiet_df_wide <- sig_mix.mod_ruraldiet_df_wide[,c("t0_vs_t1","t1_vs_t2","t0_vs_t2")] -->

<!-- head(sig_mix.mod_ruraldiet_df_wide) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # 1. add fold changes of mbege -->
<!-- network_df.ruraldiet <- merge(network_df, sig_mix.mod_ruraldiet_df_wide,  -->
<!--                               by.x = "HMDB", by.y = "row.names", all.x = T) -->

<!-- # 2. create a data frame for edges -->
<!-- net_edges.ruraldiet <- network_df.ruraldiet[,c("HMDB","class_plot")] -->
<!-- ## re-name colnames to from and to -->
<!-- colnames(net_edges.ruraldiet) <- c("from","to") -->

<!-- # 3. Create a data frame for vertices -->
<!-- net_vertices.ruraldiet <- data.frame(name = c(network_df.ruraldiet$HMDB,  -->
<!--                                     network_df.ruraldiet$class_plot),  -->
<!--                         class_count = c(rep(NA, length(network_df.ruraldiet$HMDB)), -->
<!--                                         network_df.ruraldiet$count_path), -->
<!--                        t0_vs_t1 = c(network_df.ruraldiet$t0_vs_t1,  -->
<!--                                     rep(NA, length(network_df.ruraldiet$class_plot))),  -->
<!--                        t1_vs_t2 = c(network_df.ruraldiet$t1_vs_t2,  -->
<!--                                     rep(NA, length(network_df.ruraldiet$class_plot))),  -->
<!--                        t0_vs_t2 = c(network_df.ruraldiet$t0_vs_t2,  -->
<!--                                     rep(NA, length(network_df.ruraldiet$class_plot))), -->
<!--                        type = c(rep("metabolite", length(network_df.ruraldiet$HMDB)),  -->
<!--                                 rep("pathway", length(network_df.ruraldiet$class_plot)))) -->

<!-- ## remove duplicated names -->
<!-- net_vertices.ruraldiet <- net_vertices.ruraldiet[!duplicated(net_vertices.ruraldiet$name),] -->

<!-- ## factor type -->
<!-- net_vertices.ruraldiet$type <- factor(net_vertices.ruraldiet$type, levels = c("pathway","metabolite")) -->


<!-- ## label only pathway -->
<!-- net_vertices.ruraldiet$mylabel <- ifelse(net_vertices.ruraldiet$type == "pathway", -->
<!--                                          net_vertices.ruraldiet$name, "") -->

<!-- head(net_vertices.ruraldiet) -->
<!-- ```  -->

<!-- ##### Mbege -->
<!-- #### ... get comparison results table and transform to wide format -->
<!-- ```{r} -->
<!-- sig_mix.mod_mbege_df <- summary_table_mix.mod_mbege$summary_table -->

<!-- sig_mix.mod_mbege_df <- subset(sig_mix.mod_mbege_df, regulation != "ns") -->

<!-- sig_mix.mod_mbege_df$HMDB <- melt_annotation_tal_new[match(as.character(sig_mix.mod_mbege_df$ionMZ), as.character( melt_annotation_tal_new$ionMz)),"CompoundID_new"] -->

<!-- dim(sig_mix.mod_mbege_df) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- sig_mix.mod_mbege_df_wide <- as.data.frame(spread(sig_mix.mod_mbege_df[,c("HMDB","log2FC2","comparison")], key = "comparison", value = "log2FC2" )) -->

<!-- rownames(sig_mix.mod_mbege_df_wide) <- sig_mix.mod_mbege_df_wide$HMDB -->

<!-- sig_mix.mod_mbege_df_wide$HMDB <- NULL -->

<!-- # change order of cols -->

<!-- sig_mix.mod_mbege_df_wide <- sig_mix.mod_mbege_df_wide[,c("t0_vs_t1","t1_vs_t2","t0_vs_t2")] -->

<!-- head(sig_mix.mod_mbege_df_wide) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # 1. add fold changes of mbege -->
<!-- network_df.mbege <- merge(network_df, sig_mix.mod_mbege_df_wide,  -->
<!--                               by.x = "HMDB", by.y = "row.names", all.x = T) -->

<!-- # 2. create a data frame for edges -->
<!-- net_edges.mbege <- network_df.mbege[,c("HMDB","class_plot")] -->
<!-- ## re-name colnames to from and to -->
<!-- colnames(net_edges.mbege) <- c("from","to") -->

<!-- # 3. Create a data frame for vertices -->
<!-- net_vertices.mbege <- data.frame(name = c(network_df.mbege$HMDB,  -->
<!--                                     network_df.mbege$class_plot),  -->
<!--                         class_count = c(rep(NA, length(network_df.mbege$HMDB)), -->
<!--                                         network_df.mbege$count_path), -->
<!--                        t0_vs_t1 = c(network_df.mbege$t0_vs_t1,  -->
<!--                                     rep(NA, length(network_df.mbege$class_plot))),  -->
<!--                        t1_vs_t2 = c(network_df.mbege$t1_vs_t2,  -->
<!--                                     rep(NA, length(network_df.mbege$class_plot))),  -->
<!--                        t0_vs_t2 = c(network_df.mbege$t0_vs_t2,  -->
<!--                                     rep(NA, length(network_df.mbege$class_plot))), -->
<!--                        type = c(rep("metabolite", length(network_df.mbege$HMDB)),  -->
<!--                                 rep("pathway", length(network_df.mbege$class_plot)))) -->

<!-- ## remove duplicated names -->
<!-- net_vertices.mbege <- net_vertices.mbege[!duplicated(net_vertices.mbege$name),] -->

<!-- ## factor type -->
<!-- net_vertices.mbege$type <- factor(net_vertices.mbege$type, levels = c("pathway","metabolite")) -->


<!-- ## label only pathway -->
<!-- net_vertices.mbege$mylabel <- ifelse(net_vertices.mbege$type == "pathway", -->
<!--                                          net_vertices.mbege$name, "") -->

<!-- head(net_vertices.mbege) -->
<!-- ```  -->



<!-- ## Create all the networks -->
<!-- create the network -->
<!-- ```{r} -->
<!-- ## create network -->
<!-- ### Urban diet -->
<!-- newgraph_urbandiet <- graph_from_data_frame(net_edges.urbandiet, vertices = net_vertices.urbandiet) -->

<!-- ### Rural diet -->
<!-- newgraph_ruraldiet <- graph_from_data_frame(net_edges.ruraldiet, vertices = net_vertices.ruraldiet) -->

<!-- ### Mbege -->
<!-- newgraph_mbege <- graph_from_data_frame(net_edges.mbege, vertices = net_vertices.mbege) -->
<!-- ``` -->

<!-- ### Define variables for the networks -->
<!-- general parameters for the graph -->
<!-- ```{r} -->
<!-- color_network <- c("metabolite" = "black", "pathway" = "#f1c232") -->
<!-- shapes_network <- c("metabolite" = 15, "pathway" = 21) -->

<!-- # Gradient color -->
<!-- palbrown <- brewer.pal(n = 11, name = "BrBG") -->
<!-- pal <- wes_palette("Zissou1", 100, type = "continuous") -->
<!-- ``` -->

<!-- fold change -->
<!-- ```{r} -->
<!-- # find max log2FC2 value  -->
<!-- comp_cols <- c("t0_vs_t1","t1_vs_t2","t0_vs_t2") -->
<!-- max.value.network <- rbind(net_vertices.urbandiet[,comp_cols], -->
<!--                            net_vertices.ruraldiet[,comp_cols], -->
<!--                            net_vertices.mbege[,comp_cols]) -->

<!-- max.value.network <- max(abs(max.value.network), na.rm = T) -->

<!-- max.value.network.round <- ceiling(max.value.network*1.2) -->

<!-- max.value.network*1.2 -->
<!-- max.value.network.round -->
<!-- ``` -->

<!-- ```{r} -->

<!-- # find max log2FC2 value for urban -->
<!-- comp_cols <- c("t0_vs_t1","t1_vs_t2","t0_vs_t2") -->
<!-- max.value.network.urbandiet <- net_vertices.urbandiet[,comp_cols] -->

<!-- max.value.network.urbandiet <- max(abs(max.value.network.urbandiet), na.rm = T) -->

<!-- max.value.network.urbandiet.round <- ceiling(max.value.network.urbandiet*1.2) -->

<!-- max.urbandiet <- max.value.network.urbandiet*1.2 -->
<!-- max.urbandiet -->
<!-- max.value.network.urbandiet.round -->
<!-- ``` -->

<!-- ```{r} -->

<!-- ## t0t1 -->
<!-- set.seed(1) -->
<!-- t0t1_urbandiet <- ggraph(newgraph_urbandiet, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t1), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t1, urbandiet") -->

<!-- ## t1t2 -->
<!-- set.seed(1) -->
<!-- t1t2_urbandiet <- ggraph(newgraph_urbandiet, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t1_vs_t2), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t1 - t2, urbandiet") -->


<!-- ## t0t2 -->
<!-- set.seed(1) -->
<!-- t0t2_urbandiet <- ggraph(newgraph_urbandiet, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t2), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t2, urbandiet") -->

<!-- ``` -->


<!-- # ---------------- -->

<!-- ### 6.4.3. Alltogether network -->
<!-- ```{r} -->
<!-- tmp1 <- sig_mix.mod_urbandiet_df_wide -->
<!-- #tmp1$diet <- "urban_diet" -->
<!-- colnames(tmp1) <- paste0("urbandiet.",colnames(tmp1)) -->

<!-- tmp2 <- sig_mix.mod_ruraldiet_df_wide -->
<!-- #tmp2$diet <- "rural_diet" -->
<!-- colnames(tmp2) <- paste0("ruraldiet.",colnames(tmp2)) -->

<!-- tmp3 <- sig_mix.mod_mbege_df_wide -->
<!-- #tmp3$diet <- "MBEGE" -->
<!-- colnames(tmp3) <- paste0("mbege.",colnames(tmp3)) -->

<!-- sig_mix.mod_all_df_wide <- merge(tmp1, tmp2, by = "row.names", all.x = T, all.y = T) -->
<!-- rownames(sig_mix.mod_all_df_wide) <- sig_mix.mod_all_df_wide$Row.names -->
<!-- sig_mix.mod_all_df_wide$Row.names <- NULL -->

<!-- sig_mix.mod_all_df_wide <- merge(sig_mix.mod_all_df_wide, tmp3, by = "row.names",  all.x = T, all.y = T) -->
<!-- rownames(sig_mix.mod_all_df_wide) <- sig_mix.mod_all_df_wide$Row.names -->
<!-- sig_mix.mod_all_df_wide$Row.names <- NULL -->

<!-- colnames(sig_mix.mod_all_df_wide) -->
<!-- ``` -->
<!-- #### ... create network df.all -->
<!-- ```{r} -->
<!-- export_rampdb_sig_all_class <- export_rampdb_sig_all_class %>% group_by(class_name) %>% mutate(count_class = n()) -->
<!-- export_rampdb_sig_all_class <- as.data.frame(export_rampdb_sig_all_class) -->

<!-- ## keep only pathways with more than 3 metabolites -->
<!-- network_df <- export_rampdb_sig_all_class[,c("HMDB","class_name","count_class")] -->

<!-- network_df$class_name_new <- ifelse(network_df$count_class >=5, network_df$class_name, NA) -->

<!-- ## see how many pathways each metabolite has -->
<!-- network_df <- network_df %>% group_by(HMDB) %>% mutate(count_path = n_distinct(class_name_new)) -->


<!-- ## for metabolite with count_path = 1 & class_name_new = NA --> these should get class "other" -->

<!-- network_df$class_plot <- ifelse(network_df$count_path == 1 & is.na(network_df$class_name_new), "other", network_df$class_name_new) -->

<!-- ## remove rows with class_plot == NA -->
<!-- network_df <- network_df[!is.na(network_df$class_plot),] -->

<!-- ``` -->


<!-- ```{r} -->
<!-- # 1. add fold changes of mbege -->
<!-- network_df.all <- merge(network_df, sig_mix.mod_all_df_wide,  -->
<!--                               by.x = "HMDB", by.y = "row.names", all.x = T, all.y = T) -->

<!-- # 2. create a data frame for edges -->
<!-- net_edges.all <- network_df.all[,c("HMDB","class_plot")] -->
<!-- ## re-name colnames to from and to -->
<!-- colnames(net_edges.all) <- c("from","to") -->

<!-- # 3. Create a data frame for vertices -->
<!-- net_vertices.all <- data.frame(name = c(network_df.all$HMDB,  -->
<!--                                     network_df.all$class_plot),  -->
<!--                         class_count = c(rep(NA, length(network_df.all$HMDB)), -->
<!--                                         network_df.all$count_path), -->
<!--                         ## urbandiet -->
<!--                        t0_vs_t1.urband = c(network_df.all$urbandiet.t0_vs_t1,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t1_vs_t2.urband = c(network_df.all$urbandiet.t1_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t0_vs_t2.urband = c(network_df.all$urbandiet.t0_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))), -->
<!--                        ## ruraldiet -->
<!--                        t0_vs_t1.rurald = c(network_df.all$ruraldiet.t0_vs_t1,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t1_vs_t2.rurald = c(network_df.all$ruraldiet.t1_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t0_vs_t2.rurald = c(network_df.all$ruraldiet.t0_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))), -->
<!--                        ## mbege -->
<!--                        t0_vs_t1.mbege = c(network_df.all$mbege.t0_vs_t1,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t1_vs_t2.mbege = c(network_df.all$mbege.t1_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))),  -->
<!--                        t0_vs_t2.mbege = c(network_df.all$mbege.t0_vs_t2,  -->
<!--                                     rep(NA, length(network_df.all$class_plot))), -->
<!--                        ## type -->
<!--                        type = c(rep("metabolite", length(network_df.all$HMDB)),  -->
<!--                                 rep("pathway", length(network_df.all$class_plot)))) -->

<!-- ## remove duplicated names -->
<!-- net_vertices.all <- net_vertices.all[!duplicated(net_vertices.all$name),] -->

<!-- ## factor type -->
<!-- net_vertices.all$type <- factor(net_vertices.all$type, levels = c("pathway","metabolite")) -->


<!-- ## label only pathway -->
<!-- net_vertices.all$mylabel <- ifelse(net_vertices.all$type == "pathway", -->
<!--                                          net_vertices.all$name, "") -->

<!-- head(net_vertices.all) -->
<!-- ```  -->

<!-- #### create network -->
<!-- ```{r} -->
<!-- ### all -->
<!-- newgraph_all <- graph_from_data_frame(net_edges.all, vertices = net_vertices.all, directed = FALSE) -->

<!-- ``` -->

<!-- #### clusters -->
<!-- ```{r} -->
<!-- lou <- cluster_louvain(newgraph_all) -->

<!-- lou_df <- communities(lou) -->
<!-- lou_df <- as.data.frame(do.call(rbind, Map(cbind, cluster = names(lou_df), lou_df))) -->

<!-- lou_df$cluster <- paste0("cluster_",lou_df$cluster) -->
<!-- colnames(lou_df) <- c("cluster","name") -->

<!-- ## check if there are duplicated values -->
<!-- sum(duplicated(lou_df$name)) -->

<!-- ``` -->
<!-- #### add cluster info to the network information and recreate network -->
<!-- ```{r} -->
<!-- net_vertices.all$cluster <- lou_df[match(net_vertices.all$name, lou_df$name),"cluster"] -->
<!-- net_vertices.all$weight <- 1 -->

<!-- ## direction -->
<!-- for ( i in unique(colnames(net_vertices.all)[grep("_vs_", colnames(net_vertices.all))])){ -->
<!--   net_vertices.all[[i]] <- ifelse(!is.na(net_vertices.all[[i]]) & net_vertices.all[[i]] >0, "up",  -->
<!--                                   ifelse(!is.na(net_vertices.all[[i]]) & net_vertices.all[[i]] <0, "down", "ns")) -->
<!-- } -->


<!-- unique(net_vertices.all$cluster) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- ### all -->
<!-- newgraph_all <- graph_from_data_frame(net_edges.all, vertices = net_vertices.all, directed = FALSE) -->
<!-- set.seed(12) -->
<!-- LO = layout_with_fr(newgraph_all) -->

<!-- set.seed(12) -->
<!-- plot(lou, newgraph_all, vertex.label = NA, layout = LO ) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- testplot <- ggraph(newgraph_all, layout = "manual", -->
<!--                   x = LO[, 1], -->
<!--                   y = LO[, 2]) + -->
<!--    geom_mark_hull( -->
<!--     aes(x, y, group = cluster, fill = cluster), -->
<!--     concavity = 4, -->
<!--     expand = unit(4, "mm"), -->
<!--     alpha = 0.25, color = "darkgray" -->
<!--   )+ -->
<!--   ## add new fill scale -->
<!--   new_scale_fill() + -->
<!--   scale_fill_brewer(palette = "Set1")+ -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type), size = 3, color = "#404040") + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("all") -->


<!-- testplot -->
<!-- ``` -->

<!-- ```{r} -->
<!-- shapes_network2 <- c("metabolite" = 21, "pathway" = 19 ) -->
<!-- color_network2 <- c("metabolite" = "white", "pathway" = "#f1c232") -->
<!-- fill_direction <- c("up" = "darkred", "down" = "darkblue", "ns" = NA) -->


<!-- testplot2 <- ggraph(newgraph_all, layout = "manual", -->
<!--                   x = LO[, 1], -->
<!--                   y = LO[, 2]) + -->
<!--   #  geom_mark_hull( -->
<!--   #   aes(x, y, group = cluster, fill = cluster), -->
<!--   #   concavity = 4, -->
<!--   #   expand = unit(4, "mm"), -->
<!--   #   alpha = 0.25, color = "darkgray" -->
<!--   # )+ -->
<!--   # scale_fill_brewer(palette = "Paired")+ -->
<!--   #  ## add new fill scale -->
<!--   # new_scale_fill() + -->
<!--   geom_edge_fan( color = "darkgray", size = 0.5) + -->
<!--   geom_node_point(aes( fill = t0_vs_t1.urband, shape = type, color = type), size = 3) + -->
<!--    scale_color_manual(values = color_network2)+ -->
<!--   scale_shape_manual(values =  shapes_network2)+ -->
<!--   scale_fill_manual(values = fill_direction)+ -->
<!--   # scale_fill_gradientn(colours = c("purple","yellow","darkgreen"), -->
<!--   #                      #limits = c(-max.urbandiet, max.urbandiet), -->
<!--   #                      #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--   #                      limits = c(-1.5,1.5), # put range of log2FC2 to -2,2 -->
<!--   #                      oob = scales::squish, # how to handle values outside of the limits rang -->
<!--   #                      guide = guide_colorbar(frame.colour = "black", -->
<!--   #                                             ticks.colour = "black"), na.value = "#e5e5e5")+ -->
<!--   #geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 vs t1, urbandiet") -->


<!-- testplot2 -->
<!-- ``` -->
<!-- ```{r} -->
<!-- color_network2 <- c("metabolite" = "white", "pathway" = "#f1c232") -->

<!-- testplot2 <- ggraph(newgraph_all, layout = "manual", -->
<!--                   x = LO[, 1], -->
<!--                   y = LO[, 2]) + -->
<!--   geom_edge_fan( color = "darkgray", size = 0.5) + -->
<!--   geom_node_point(aes( fill = t0_vs_t1.rurald, shape = type, color = type), size = 3) + -->
<!--    scale_color_manual(values = color_network2)+ -->
<!--   scale_shape_manual(values =  shapes_network2)+ -->
<!--   scale_fill_manual(values = fill_direction)+ -->
<!--   # scale_fill_gradientn(colours = c("purple","yellow","darkgreen"), -->
<!--   #                      #limits = c(-max.urbandiet, max.urbandiet), -->
<!--   #                      #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--   #                      limits = c(-1.5,1.5), # put range of log2FC2 to -2,2 -->
<!--   #                      oob = scales::squish, # how to handle values outside of the limits rang -->
<!--   #                      guide = guide_colorbar(frame.colour = "black", -->
<!--   #                                             ticks.colour = "black"), na.value = "#e5e5e5")+ -->
<!--  # geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 vs t1, ruraldiet") -->


<!-- testplot2 -->
<!-- ``` -->

<!-- ```{r} -->
<!-- testplot2 <- ggraph(newgraph_all, layout = "manual", -->
<!--                   x = LO[, 1], -->
<!--                   y = LO[, 2]) + -->
<!--   geom_edge_fan( color = "darkgray", size = 0.5) + -->
<!--   geom_node_point(aes( fill = t0_vs_t1.mbege, shape = type, color = type), size = 3) + -->
<!--    scale_color_manual(values = color_network2)+ -->
<!--   scale_shape_manual(values =  shapes_network2)+ -->
<!--   scale_fill_manual(values = fill_direction)+ -->
<!--   # scale_fill_gradientn(colours = c("purple","yellow","darkgreen"), -->
<!--   #                      #limits = c(-max.urbandiet, max.urbandiet), -->
<!--   #                      #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--   #                      limits = c(-1.5,1.5), # put range of log2FC2 to -2,2 -->
<!--   #                      oob = scales::squish, # how to handle values outside of the limits rang -->
<!--   #                      guide = guide_colorbar(frame.colour = "black", -->
<!--   #                                             ticks.colour = "black"), na.value = "#e5e5e5")+ -->
<!--   #geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 vs t1, mbege") -->


<!-- testplot2 -->
<!-- ``` -->

<!-- ```{r} -->
<!-- testplot2 <- ggraph(newgraph_all, layout = "manual", -->
<!--                   x = LO[, 1], -->
<!--                   y = LO[, 2]) + -->
<!--   geom_edge_fan( color = "darkgray", size = 0.5) + -->
<!--   geom_node_point(aes( fill = t0_vs_t2.urband, shape = type, color = type), size = 3) + -->
<!--    scale_color_manual(values = color_network2)+ -->
<!--   scale_shape_manual(values =  shapes_network2)+ -->
<!--   scale_fill_manual(values = fill_direction)+ -->
<!--   # scale_fill_gradientn(colours = c("purple","yellow","darkgreen"), -->
<!--   #                      #limits = c(-max.urbandiet, max.urbandiet), -->
<!--   #                      #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--   #                      limits = c(-1.5,1.5), # put range of log2FC2 to -2,2 -->
<!--   #                      oob = scales::squish, # how to handle values outside of the limits rang -->
<!--   #                      guide = guide_colorbar(frame.colour = "black", -->
<!--   #                                             ticks.colour = "black"), na.value = "#e5e5e5")+ -->
<!--   #geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 vs t1, mbege") -->


<!-- testplot2 -->
<!-- ``` -->

<!-- ```{r} -->
<!-- library(graphlayouts) -->
<!-- gnetwork <- graph_from_data_frame(net_edges.all, vertices = net_vertices.all, directed = FALSE) -->
<!-- gnetwork <- igraph::simplify(gnetwork) -->

<!-- bb <- layout_as_backbone(gnetwork, keep = 1) -->
<!-- E(gnetwork)$col <- F -->
<!-- E(gnetwork)$col[bb$backbone] <- T -->
<!-- ``` -->


<!-- ### Plot diets -->
<!-- #### Urban diet -->
<!-- ```{r} -->

<!-- t0t1_urbandiet <- ggraph(gnetwork, layout = "manual", -->
<!--                   x = bb$xy[, 1], -->
<!--                   y = bb$xy[, 2]) + -->
<!--    geom_mark_hull( -->
<!--     aes(x, y, group = cluster, fill = cluster), -->
<!--     concavity = 4, -->
<!--     expand = unit(4, "mm"), -->
<!--     alpha = 0.25, color = "darkgray" -->
<!--   )+ -->
<!--   ## add new fill scale -->
<!--   new_scale_fill() + -->
<!--   scale_fill_brewer(palette = "Set1")+ -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t1.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t1, urbandiet") -->


<!-- t1t2_urbandiet <- ggraph(gnetwork, layout = "manual", -->
<!--                   x = bb$xy[, 1], -->
<!--                   y = bb$xy[, 2]) + -->
<!--    geom_mark_hull( -->
<!--     aes(x, y, group = cluster, fill = cluster), -->
<!--     concavity = 4, -->
<!--     expand = unit(2, "mm"), -->
<!--     alpha = 0.25, color = "darkgray" -->
<!--   )+ -->
<!--   ## add new fill scale -->
<!--   new_scale_fill() + -->
<!--   scale_fill_brewer(palette = "Set1")+ -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t1_vs_t2.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t1 - t2, urbandiet") -->


<!-- t0t2_urbandiet <- ggraph(gnetwork, layout = "manual", -->
<!--                   x = bb$xy[, 1], -->
<!--                   y = bb$xy[, 2]) + -->
<!--    geom_mark_hull( -->
<!--     aes(x, y, group = cluster, fill = cluster), -->
<!--     concavity = 4, -->
<!--     expand = unit(2, "mm"), -->
<!--     alpha = 0.25, color = "darkgray" -->
<!--   )+ -->
<!--   ## add new fill scale -->
<!--   new_scale_fill() + -->
<!--   scale_fill_brewer(palette = "Set1")+ -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t2.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t2, urbandiet") -->




<!-- ``` -->


<!-- #### ... Urban diet -->
<!-- ```{r} -->

<!-- # find max log2FC2 value for urban -->
<!-- comp_cols <- paste0( c("t0_vs_t1","t1_vs_t2","t0_vs_t2"), ".urband") -->
<!-- max.value.network.urbandiet <- net_vertices.all[,comp_cols] -->

<!-- max.value.network.urbandiet <- max(abs(max.value.network.urbandiet), na.rm = T) -->

<!-- max.value.network.urbandiet.round <- ceiling(max.value.network.urbandiet*1.2) -->

<!-- max.urbandiet <- max.value.network.urbandiet -->
<!-- max.urbandiet -->
<!-- max.value.network.urbandiet.round -->
<!-- ``` -->

<!-- ```{r} -->

<!-- ## t0t1 -->
<!-- set.seed(1) -->
<!-- t0t1_urbandiet <- ggraph(newgraph_all, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t1.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t1, urbandiet") -->

<!-- ## t1t2 -->
<!-- set.seed(1) -->
<!-- t1t2_urbandiet <- ggraph(newgraph_all, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t1_vs_t2.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t1 - t2, urbandiet") -->


<!-- ## t0t2 -->
<!-- set.seed(1) -->
<!-- t0t2_urbandiet <- ggraph(newgraph_all, layout = "graphopt") + -->
<!--   geom_edge_link( color = "lightgray") + -->
<!--   geom_node_point(aes( fill = type, shape = type, color = t0_vs_t2.urband), size = 4) + -->
<!--    scale_fill_manual(values = color_network)+ -->
<!--   scale_shape_manual(values =  shapes_network)+ -->
<!--   scale_color_gradientn(colours = c("#0000e5","lightyellow","#e50000"), -->
<!--                        #limits = c(-max.urbandiet, max.urbandiet), -->
<!--                        #values = scales::rescale(., from =  c(-max.urbandiet, max.urbandiet), to = c(-2,2) ), -->
<!--                        limits = c(-2,2), # put range of log2FC2 to -2,2 -->
<!--                        oob = scales::squish, # how to handle values outside of the limits rang -->
<!--                        guide = guide_colorbar(frame.colour = "black", -->
<!--                                               ticks.colour = "black"), na.value = "white")+ -->
<!--   geom_node_text(aes(label = mylabel), repel = TRUE, size = 2, max.overlaps = 50 ) + -->
<!--   theme_graph()+ggtitle("t0 - t2, urbandiet") -->

 <!-- ``` -->



# ==================================

# 8. Enrichment of class
## 8.1. Urbandiet
```{r}
tmpinput <-NULL 

rampdb_sig_enrichchem_urbandiet <- list()

for (i in names(list_sig_dream_urbandiet_diet_hmdb)){
  rampdb_sig_enrichchem_urbandiet[[i]]$up <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_urbandiet_diet_hmdb[[i]]$up))[["ClassyFire_sub_class"]]
  
   rampdb_sig_enrichchem_urbandiet[[i]]$down <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_urbandiet_diet_hmdb[[i]]$down))[["ClassyFire_sub_class"]]
}
```
```{r}
# rbind the data frames within each comparison
  rampdb_sig_enrichchem_urbandiet <- lapply(rampdb_sig_enrichchem_urbandiet, function(res){
    do.call(rbind, Map(cbind, regulation = names(res), res))
  })
  
  rampdb_sig_enrichchem_urbandiet <- do.call(rbind, Map(cbind, comparison = names(rampdb_sig_enrichchem_urbandiet), rampdb_sig_enrichchem_urbandiet))
  
  rampdb_sig_enrichchem_urbandiet <- subset(rampdb_sig_enrichchem_urbandiet, adjP_BH <= 0.05)
  
  rampdb_sig_enrichchem_urbandiet <- rampdb_sig_enrichchem_urbandiet %>% add_significance(., p.col = "adjP_BH")
  
  head(rampdb_sig_enrichchem_urbandiet)
```

```{r}
rampdb_sig_enrichchem_urbandiet$comparison <- factor(rampdb_sig_enrichchem_urbandiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(rampdb_sig_enrichchem_urbandiet, met_hits >2), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(.~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())
```
## 8.2. Rural diet
```{r}
rampdb_sig_enrichchem_rurladiet <- list()

for (i in names(list_sig_dream_ruraldiet_diet_hmdb)){
  rampdb_sig_enrichchem_rurladiet[[i]]$up <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_ruraldiet_diet_hmdb[[i]]$up))[["ClassyFire_sub_class"]]
  
   rampdb_sig_enrichchem_rurladiet[[i]]$down <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_ruraldiet_diet_hmdb[[i]]$down))[["ClassyFire_sub_class"]]
}
```
```{r}
# rbind the data frames within each comparison
  rampdb_sig_enrichchem_rurladiet <- lapply(rampdb_sig_enrichchem_rurladiet, function(res){
    do.call(rbind, Map(cbind, regulation = names(res), res))
  })
  
  rampdb_sig_enrichchem_rurladiet <- do.call(rbind, Map(cbind, comparison = names(rampdb_sig_enrichchem_rurladiet), rampdb_sig_enrichchem_rurladiet))
  
  rampdb_sig_enrichchem_rurladiet <- subset(rampdb_sig_enrichchem_rurladiet, adjP_BH <= 0.05)
  
  rampdb_sig_enrichchem_rurladiet <- rampdb_sig_enrichchem_rurladiet %>% add_significance(., p.col = "adjP_BH")
  
  head(rampdb_sig_enrichchem_rurladiet)
```

```{r}
rampdb_sig_enrichchem_rurladiet$comparison <- factor(rampdb_sig_enrichchem_rurladiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(rampdb_sig_enrichchem_rurladiet, met_hits >2), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(.~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())+
  ggtitle("Rural diet")
```
## 8.3. Mbege
```{r}
rampdb_sig_enrichchem_mbege <- list()

for (i in names(list_sig_dream_mbege_diet_hmdb)){
  rampdb_sig_enrichchem_mbege[[i]]$up <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_mbege_diet_hmdb[[i]]$up))[["ClassyFire_sub_class"]]
  
   rampdb_sig_enrichchem_mbege[[i]]$down <- Getchemenrich_RAMPdb(hmdbinput =  paste0("hmdb:", list_sig_dream_mbege_diet_hmdb[[i]]$down))[["ClassyFire_sub_class"]]
}
```
```{r}
# rbind the data frames within each comparison
  rampdb_sig_enrichchem_mbege <- lapply(rampdb_sig_enrichchem_mbege, function(res){
    do.call(rbind, Map(cbind, regulation = names(res), res))
  })
  
  rampdb_sig_enrichchem_mbege <- do.call(rbind, Map(cbind, comparison = names(rampdb_sig_enrichchem_mbege), rampdb_sig_enrichchem_mbege))
  
  rampdb_sig_enrichchem_mbege <- subset(rampdb_sig_enrichchem_mbege, adjP_BH <= 0.05)
  
  rampdb_sig_enrichchem_mbege <- rampdb_sig_enrichchem_mbege %>% add_significance(., p.col = "adjP_BH")
  
  head(rampdb_sig_enrichchem_mbege)
```

```{r}
rampdb_sig_enrichchem_mbege$comparison <- factor(rampdb_sig_enrichchem_mbege$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(rampdb_sig_enrichchem_mbege, met_hits >2), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(.~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())+
  ggtitle("Mbege diet")
```

## 8.4. Plot all together
### enrichment
```{r}
rampdb_sig_enrichchem_urbandiet$diet <- "urban_diet" 

rampdb_sig_enrichchem_rurladiet$diet <- "rural_diet"

rampdb_sig_enrichchem_mbege$diet <- "MBEGE"

rampdb_sig_enrichchem_all <- rbind(rampdb_sig_enrichchem_urbandiet, rampdb_sig_enrichchem_rurladiet, rampdb_sig_enrichchem_mbege)

rampdb_sig_enrichchem_all$diet <- factor(rampdb_sig_enrichchem_all$diet, levels = c("urban_diet","rural_diet","MBEGE"))

## sum up the count per each metabolite
```


```{r fig.width=15, fig.height=5}
ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(diet~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())
```

```{r fig.width=10, fig.height=5}
top3 <- as.data.frame(subset(rampdb_sig_enrichchem_all, met_hits >1) %>% group_by(diet, comparison) %>%  slice_min(adjP_BH, n = 10) %>%
  mutate(label = class_name))

ggplot(top3, aes(x= regulation, y= reorder(class_name, met_hits )))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(diet~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())
```


### ... counts da
```{r}
count_da_all <- as.data.frame(all_sig_df %>% 
                                      group_by(diet, comparison) %>% 
                                      summarize(up = sum(regulation == "up"),
                                                down = sum(regulation == "down")))


count_da_all <- reshape2::melt(count_da_all)

count_da_all$diet <- factor(count_da_all$diet, levels = c("urban_diet","rural_diet","MBEGE"))

count_da_all
```


```{r fig.height=5 }
p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison == "t0_vs_t1"), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(.~ diet, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

p2 <- ggplot(subset(count_da_all, comparison == "t0_vs_t1"), aes(x= diet, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black")+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))+plot_annotation(title = "t1 vs t0")
```
```{r fig.height=5 }
p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison == "t0_vs_t2"), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(.~ diet, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

p2 <- ggplot(subset(count_da_all, comparison == "t0_vs_t2"), aes(x= diet, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black")+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))+plot_annotation(title = "t2 vs t0")
```
```{r fig.height=5, fig.width=10 }
p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(comparison~ diet, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

p2 <- ggplot(subset(count_da_all, comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= diet, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black")+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  facet_wrap(.~comparison, nrow = 1)+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))+plot_annotation(title = "t1 vs t0, t2 vs t0")
```
```{r fig.height=5, fig.width=10 }
p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(diet~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

p2 <- ggplot(subset(count_da_all, comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black")+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  facet_wrap(.~diet, nrow = 1)+
  geom_text(vjust = 1, position = position_stack(), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))+plot_annotation(title = "t1 vs t0, t2 vs t0")
```
```{r fig.height=5, fig.width=10 }
## metabolites ratio
rampdb_sig_enrichchem_all$metabo_ratio <- as.numeric(rampdb_sig_enrichchem_all$met_hits)/ as.numeric(rampdb_sig_enrichchem_all$met_size)

rampdb_sig_enrichchem_all$regulation <- factor(rampdb_sig_enrichchem_all$regulation, levels = c("up","down"))

p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = metabo_ratio), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(diet~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

p2 <- ggplot(subset(count_da_all, comparison %in% c("t0_vs_t1", "t0_vs_t2")), aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black", position = position_dodge())+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  facet_wrap(.~diet, nrow = 1)+
  geom_text(vjust = 1, position = position_dodge(width = 1), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))+plot_annotation(title = "t1 vs t0, t2 vs t0")
```
```{r fig.height=5, fig.width=10 }
## metabolites ratio
rampdb_sig_enrichchem_all$metabo_ratio <- as.numeric(rampdb_sig_enrichchem_all$met_hits)/ as.numeric(rampdb_sig_enrichchem_all$met_size)

rampdb_sig_enrichchem_all$regulation <- factor(rampdb_sig_enrichchem_all$regulation, levels = c("up","down"))

p1 <- ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 ), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = metabo_ratio), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_wrap(diet~ comparison, nrow =1)+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())

count_da_all$comparison <- factor(count_da_all$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

p2 <- ggplot(count_da_all, aes(x= comparison, y= value, fill = variable))+
  geom_bar(stat = "identity", color = "black", position = position_dodge())+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  theme_void()+
  facet_wrap(.~diet, nrow = 1)+
  geom_text(vjust = 1, position = position_dodge(width = 1), aes(label = value))

p2 + p1 + plot_layout(ncol = 1, guides = "collect", heights = c(2,9))
```

```{r fig.height=7, fig.width=7}
ggplot(subset(rampdb_sig_enrichchem_all, met_hits >1 & comparison %in%  c("t0_vs_t1", "t0_vs_t2")), aes(x= regulation, y= class_name ))+
  geom_point(aes(shape = regulation, fill = regulation, size = met_hits), shape = 21)+
  geom_text(aes(label = adjP_BH.signif))+
  scale_fill_manual(values = c("up" = "tomato3", "down" = "cornflowerblue"))+
  #scale_shape_manual(values = c("up" = 24, "down" = 25))+
  facet_grid(comparison~ diet, space = "free", scales = "free")+
  theme_bw()+
  scale_size_continuous(range = c(2,8))+
  theme(axis.title = element_blank())
```
### count chemical class
```{r}
count_chem_class_all <- as.data.frame(all_sig_df %>% group_by(diet, comparison, regulation, class_name) %>% summarize(count = n()))

## remove na classes
count_chem_class_all <- count_chem_class_all[!is.na(count_chem_class_all$class_name),]
```

## 7.5 violin plots

### urban diet
```{r}
metbo_2_plot <- c("Dopamine 4-sulfate" ,"Dopamine" ,"Docosahexaenoic acid","Eicosapentaenoic acid","12-Oxo-20-carboxy-leukotriene B4","Leukotriene B4" )
names(metbo_2_plot) <- res_dream_all[match(metbo_2_plot, res_dream_all$CompoundName),"ionMz"]

tmp_urbandiet_data <- metabo_log2_urbandiet[names(metbo_2_plot),]

tmp_urbandiet_data <- reshape2::melt(tmp_urbandiet_data, id.vars = "row.names")

colnames(tmp_urbandiet_data) <- c("ionMz","ID","value")

tmp_urbandiet_data$CompoundName <- res_dream_all[match(tmp_urbandiet_data$ionMz, res_dream_all$ionMz),"CompoundName"]

tmp_urbandiet_data <- merge(tmp_urbandiet_data, sample_table_urbandiet, by = "ID", all.x = T)

head(tmp_urbandiet_data)
```
```{r}
tmp_urbandiet_data$CompoundName <- factor(tmp_urbandiet_data$CompoundName, levels = rev(metbo_2_plot))

vln.plot.urbandiet <- ggplot(tmp_urbandiet_data, aes(x= timepoint, y= value))+
  geom_violin(aes(fill = timepoint))+
  geom_boxplot(fill = "white", width = 0.2, outlier.shape = NA)+
  geom_point(size =0.5)+
  geom_line(aes(group = PID), alpha = 0.5, size = 0.2)+
  scale_fill_manual(values = col_timepoint)+
  theme_bw()+
  facet_wrap(.~CompoundName, scales = "free")+
  ggtitle("urban diet")

vln.plot.urbandiet
```
### rural diet
```{r}
metbo_2_plot <- c("Fumaric acid","Eicosapentaenoic acid","Arachidonic acid","Docosapentaenoic acid (22n-6)","Dodecanoic acid","Docosahexaenoic acid")

names(metbo_2_plot) <- res_dream_all[match(metbo_2_plot, res_dream_all$CompoundName),"ionMz"]

tmp_ruraldiet_data <- metabo_log2_ruraldiet[names(metbo_2_plot),]

tmp_ruraldiet_data <- reshape2::melt(tmp_ruraldiet_data, id.vars = "row.names")

colnames(tmp_ruraldiet_data) <- c("ionMz","ID","value")

tmp_ruraldiet_data$CompoundName <- res_dream_all[match(tmp_ruraldiet_data$ionMz, res_dream_all$ionMz),"CompoundName"]

tmp_ruraldiet_data <- merge(tmp_ruraldiet_data, sample_table_ruraldiet, by = "ID", all.x = T)

head(tmp_ruraldiet_data)
```
```{r fig.width=14, fig.height=3}
tmp_ruraldiet_data$CompoundName <- factor(tmp_ruraldiet_data$CompoundName, levels = rev(metbo_2_plot))

vln.plot.ruraldiet <- ggplot(tmp_ruraldiet_data, aes(x= timepoint, y= value))+
  geom_violin(aes(fill = timepoint))+
  geom_boxplot(fill = "white", width = 0.2, outlier.shape = NA)+
  geom_point(size = 0.5)+
  geom_line(aes(group = PID), alpha = 0.5, size = 0.2)+
  scale_fill_manual(values = col_timepoint)+
  theme_bw()+
  facet_wrap(.~CompoundName, scales = "free", nrow = 1)+
  ggtitle("Rural diet")

vln.plot.ruraldiet
```

### Mbege
```{r}
# metbo_2_plot <- c("LysoPE(0:0/16:0)","LysoPE(0:0/18:2(9Z,12Z))","LysoPE(0:0/18:1(11Z))","LysoPE(0:0/20:4(5Z,8Z,11Z,14Z))")

metbo_2_plot <- c("Cysteinylglycine","Tyrosol","4-ethylphenylsulfate","Hippuric acid")

names(metbo_2_plot) <- res_dream_all[match(metbo_2_plot, res_dream_all$CompoundName),"ionMz"]

tmp_mbege_data <- metabo_log2_mbege[names(metbo_2_plot),]

tmp_mbege_data <- reshape2::melt(tmp_mbege_data, id.vars = "row.names")

colnames(tmp_mbege_data) <- c("ionMz","ID","value")

tmp_mbege_data$CompoundName <- res_dream_all[match(tmp_mbege_data$ionMz, res_dream_all$ionMz),"CompoundName"]

tmp_mbege_data <- merge(tmp_mbege_data, sample_table_mbege, by = "ID", all.x = T)

head(tmp_mbege_data)
```
```{r}
tmp_mbege_data$CompoundName <- factor(tmp_mbege_data$CompoundName, levels = rev(metbo_2_plot))

vln.plot.mbege <- ggplot(tmp_mbege_data, aes(x= timepoint, y= value))+
  geom_violin(aes(fill = timepoint))+
  geom_boxplot(fill = "white", width = 0.2, outlier.shape = NA)+
  geom_point(size = 0.5)+
  geom_line(aes(group = PID), alpha = 0.5, size = 0.2)+
  scale_fill_manual(values = col_timepoint)+
  theme_bw()+
  facet_wrap(.~CompoundName, scales = "free")+
  ggtitle("Mbege")

vln.plot.mbege
```
```{r}

```


## 7.6. Heatmaps for classes
### assign metabolites to classes
##### ... all sig list
```{r}

all_sig_hmdb <- unique(subset(all_sig_df, regulation != "ns")$HMDB)

str(all_sig_hmdb)
```

#### ... Pathway- RampDB


```{r}
tmpinput <- paste0("hmdb:", all_sig_hmdb)
rampdb_sig_all <- Getpathway_RAMPdb(hmdbinput =  tmpinput)

length(unique(tmpinput))
length(unique(rampdb_sig_all$inputId))

```
only 89 of the 356 sig metabolties were annotated to pathways

##### .... chemical class
```{r}
tmpinput <- paste0("hmdb:", all_sig_hmdb)
rampdb_sig_all_class <- Getchemclass_RAMPdb(hmdbinput =  tmpinput)
rampdb_sig_all_class$HMDB <- gsub("hmdb:","", rampdb_sig_all_class$sourceId)

length(tmpinput)
length(unique(rampdb_sig_all_class$sourceId))
```

```{r}
chemical_annotation <- as.data.frame.matrix(table(rampdb_sig_all_class[,c("HMDB","class_name")]))

head(chemical_annotation)
```


```{r}
ClassyFire_sub_class <- subset(rampdb_sig_all_class, class_level_name == "ClassyFire_sub_class")

all_sig_df$class_name <-ClassyFire_sub_class[match(all_sig_df$HMDB, ClassyFire_sub_class$HMDB),"class_name"]

res_dream_all$class_name <-ClassyFire_sub_class[match(res_dream_all$HMDB, ClassyFire_sub_class$HMDB),"class_name"]

head(all_sig_df)
head(res_dream_all)
```

#### ... food
```{r}
food <- read.csv("./foodb/metabolites-2024-09-11_food_HMDB.csv")

head(food)
```
```{r}
res_dream_all$food <- ifelse(res_dream_all$HMDB %in% food$HMDB_ID, "food",NA)

all_sig_df$food <- ifelse(all_sig_df$HMDB %in% food$HMDB_ID, "food",NA)
```


```{r}
all_sig_df$diet <- factor(all_sig_df$diet, levels = c("urban_diet","rural_diet","MBEGE"))

all_sig_df$comparison <- factor(all_sig_df$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(all_sig_df, aes(x= regulation, fill = food))+
  geom_bar(stat = "count", color = "black")+
  scale_fill_manual(values = c("food" = "maroon3"), na.value = "lightgray")+
  facet_wrap(diet ~ comparison)
```
```{r}
ggplot(all_sig_df, aes(x= regulation, fill = food))+
  geom_bar(stat = "count", color = "black", position = "fill")+
  scale_fill_manual(values = c("food" = "maroon3"), na.value = "lightgray")+
  facet_wrap(diet ~ comparison)
```

#### ... hetamps
##### ...Fatty acids and conjugates

```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Fatty acids and conjugates" & diet == "urban_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

log2_urbandiet_tmp <- metabo_log2_urbandiet[which(rownames(metabo_log2_urbandiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_urbandiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_urbandiet_tmp <- log2_urbandiet_tmp[,match(tmp_sample_table$ID, colnames(log2_urbandiet_tmp))]

log2_urbandiet_tmp_scale <- t(scale(t(log2_urbandiet_tmp)))

rownames(log2_urbandiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_urbandiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno) <- tmp_sample_table$ID

pheatmap(log2_urbandiet_tmp_scale, cluster_cols = F, treeheight_row = 2,
         annotation_col = tmp_anno, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_urbandiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_urbandiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, urban diet")
```
```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Fatty acids and conjugates" & diet == "rural_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_ruraldiet_tmp <- metabo_log2_ruraldiet[which(rownames(metabo_log2_ruraldiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_ruraldiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_ruraldiet_tmp <- log2_ruraldiet_tmp[,match(tmp_sample_table$ID, colnames(log2_ruraldiet_tmp))]

log2_ruraldiet_tmp_scale <- as.data.frame(t(scale(t(log2_ruraldiet_tmp))))

rownames(log2_ruraldiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_ruraldiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_ruraldiet <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_ruraldiet) <- tmp_sample_table$ID

pheatmap(log2_ruraldiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_ruraldiet, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_ruraldiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_ruraldiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, rural diet")
```

```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Fatty acids and conjugates" & diet == "MBEGE")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_mbege_tmp <- metabo_log2_mbege[which(rownames(metabo_log2_mbege) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_mbege
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_mbege_tmp <- log2_mbege_tmp[,match(tmp_sample_table$ID, colnames(log2_mbege_tmp))]

log2_mbege_tmp_scale <- as.data.frame(t(scale(t(log2_mbege_tmp))))

rownames(log2_mbege_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_mbege_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_mbege <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_mbege) <- tmp_sample_table$ID

pheatmap(log2_mbege_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_mbege, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_mbege_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_mbege_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, Mbege")
```


```{r}
tmp_hm_df_fa <- merge(log2_urbandiet_tmp_scale, log2_ruraldiet_tmp_scale, by = "row.names", all.x = T, all.y = T)

rownames(tmp_hm_df_fa) <- tmp_hm_df_fa$Row.names
tmp_hm_df_fa$Row.names <- NULL

tmp_hm_df_fa <- merge(tmp_hm_df_fa, log2_mbege_tmp_scale, by = "row.names", all.x = T, all.y = T)
rownames(tmp_hm_df_fa) <- tmp_hm_df_fa$Row.names
tmp_hm_df_fa$Row.names <- NULL

head(tmp_hm_df_fa)
```
```{r}
anno_tmp_all_fa <- rbind(tmp_anno, tmp_anno_ruraldiet, tmp_anno_mbege)
```

```{r}
## replace NA with 0
tmp_hm_df_fa[is.na(tmp_hm_df_fa)] <- 0
```


```{r fig.height=10, fig.width=10}
pmap_fattyacids <- pheatmap(tmp_hm_df_fa, cluster_cols = F, 
         annotation_col = anno_tmp_all_fa, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = tmp_hm_df_fa, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = tmp_hm_df_fa, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, all")
```

##### ...Carbohydrates and carbohydrate conjugates


```{r fig.height=8}
## Carbohydrates and carbohydrate conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Carbohydrates and carbohydrate conjugates" & diet == "urban_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

log2_urbandiet_tmp <- metabo_log2_urbandiet[which(rownames(metabo_log2_urbandiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_urbandiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_urbandiet_tmp <- log2_urbandiet_tmp[,match(tmp_sample_table$ID, colnames(log2_urbandiet_tmp))]

log2_urbandiet_tmp_scale <- t(scale(t(log2_urbandiet_tmp)))

rownames(log2_urbandiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_urbandiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno) <- tmp_sample_table$ID

pheatmap(log2_urbandiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_urbandiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_urbandiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Carbohydrates and carbohydrate conjugates, urban diet")
```
```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Carbohydrates and carbohydrate conjugates" & diet == "rural_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_ruraldiet_tmp <- metabo_log2_ruraldiet[which(rownames(metabo_log2_ruraldiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_ruraldiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_ruraldiet_tmp <- log2_ruraldiet_tmp[,match(tmp_sample_table$ID, colnames(log2_ruraldiet_tmp))]

log2_ruraldiet_tmp_scale <- as.data.frame(t(scale(t(log2_ruraldiet_tmp))))

rownames(log2_ruraldiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_ruraldiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_ruraldiet <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_ruraldiet) <- tmp_sample_table$ID

pheatmap(log2_ruraldiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_ruraldiet, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_ruraldiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_ruraldiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Carbohydrates and carbohydrate conjugates, rural diet")
```

```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Carbohydrates and carbohydrate conjugates" & diet == "MBEGE")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_mbege_tmp <- metabo_log2_mbege[which(rownames(metabo_log2_mbege) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_mbege
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_mbege_tmp <- log2_mbege_tmp[,match(tmp_sample_table$ID, colnames(log2_mbege_tmp))]

log2_mbege_tmp_scale <- as.data.frame(t(scale(t(log2_mbege_tmp))))

rownames(log2_mbege_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_mbege_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_mbege <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_mbege) <- tmp_sample_table$ID

pheatmap(log2_mbege_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_mbege, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_mbege_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_mbege_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Carbohydrates and carbohydrate conjugates, Mbege")
```


```{r}
tmp_hm_df_carbo <- merge(log2_urbandiet_tmp_scale, log2_ruraldiet_tmp_scale, by = "row.names", all.x = T, all.y = T)

rownames(tmp_hm_df_carbo) <- tmp_hm_df_carbo$Row.names
tmp_hm_df_carbo$Row.names <- NULL

tmp_hm_df_carbo <- merge(tmp_hm_df_carbo, log2_mbege_tmp_scale, by = "row.names", all.x = T, all.y = T)
rownames(tmp_hm_df_carbo) <- tmp_hm_df_carbo$Row.names
tmp_hm_df_carbo$Row.names <- NULL

head(tmp_hm_df_carbo)
```
```{r}
anno_tmp_all_carbo <- rbind(tmp_anno, tmp_anno_ruraldiet, tmp_anno_mbege)
```

```{r}
## replace NA with 0
tmp_hm_df_carbo[is.na(tmp_hm_df_carbo)] <- 0
```


```{r fig.height=10, fig.width=10}
pmap_carbo <- pheatmap(tmp_hm_df_carbo, cluster_cols = F, 
         annotation_col = anno_tmp_all_carbo, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = tmp_hm_df_carbo, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = tmp_hm_df_carbo, 
                               maxvalue = 2)[["color"]], main = "Carbohydrates and carbohydrate conjugates, all")
```
##### ...	Amino acids, peptides, and analogues


```{r fig.height=8}
## Carbohydrates and carbohydrate conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Amino acids, peptides, and analogues" & diet == "urban_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

log2_urbandiet_tmp <- metabo_log2_urbandiet[which(rownames(metabo_log2_urbandiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_urbandiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_urbandiet_tmp <- log2_urbandiet_tmp[,match(tmp_sample_table$ID, colnames(log2_urbandiet_tmp))]

log2_urbandiet_tmp_scale <- t(scale(t(log2_urbandiet_tmp)))

rownames(log2_urbandiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_urbandiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno) <- tmp_sample_table$ID

pheatmap(log2_urbandiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_urbandiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_urbandiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Amino acids, peptides, and analogues, urban diet")
```
```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Amino acids, peptides, and analogues" & diet == "rural_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_ruraldiet_tmp <- metabo_log2_ruraldiet[which(rownames(metabo_log2_ruraldiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_ruraldiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_ruraldiet_tmp <- log2_ruraldiet_tmp[,match(tmp_sample_table$ID, colnames(log2_ruraldiet_tmp))]

log2_ruraldiet_tmp_scale <- as.data.frame(t(scale(t(log2_ruraldiet_tmp))))

rownames(log2_ruraldiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_ruraldiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_ruraldiet <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_ruraldiet) <- tmp_sample_table$ID

pheatmap(log2_ruraldiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_ruraldiet, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_ruraldiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_ruraldiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Amino acids, peptides, and analogues, rural diet")
```

```{r fig.height=8}
## Fatty acids and conjugates

chosen_metabolites <- unique(subset(all_sig_df, class_name == "Amino acids, peptides, and analogues" & diet == "MBEGE")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

chosen_metabolites

log2_mbege_tmp <- metabo_log2_mbege[which(rownames(metabo_log2_mbege) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_mbege
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_mbege_tmp <- log2_mbege_tmp[,match(tmp_sample_table$ID, colnames(log2_mbege_tmp))]

log2_mbege_tmp_scale <- as.data.frame(t(scale(t(log2_mbege_tmp))))

rownames(log2_mbege_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_mbege_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno_mbege <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno_mbege) <- tmp_sample_table$ID

pheatmap(log2_mbege_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno_mbege, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = log2_mbege_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_mbege_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Amino acids, peptides, and analogues, Mbege")
```


```{r}
tmp_hm_df_aa <- merge(log2_urbandiet_tmp_scale, log2_ruraldiet_tmp_scale, by = "row.names", all.x = T, all.y = T)

rownames(tmp_hm_df_aa) <- tmp_hm_df_aa$Row.names
tmp_hm_df_aa$Row.names <- NULL

tmp_hm_df_aa <- merge(tmp_hm_df_aa, log2_mbege_tmp_scale, by = "row.names", all.x = T, all.y = T)
rownames(tmp_hm_df_aa) <- tmp_hm_df_aa$Row.names
tmp_hm_df_aa$Row.names <- NULL

head(tmp_hm_df_aa)
```
```{r}
anno_tmp_all_aa <- rbind(tmp_anno, tmp_anno_ruraldiet, tmp_anno_mbege)
```

```{r}
## replace NA with 0
tmp_hm_df_aa[is.na(tmp_hm_df_aa)] <- 0
```


```{r fig.height=10, fig.width=10}
pmap_aa <- pheatmap(tmp_hm_df_aa, cluster_cols = F, 
         annotation_col = anno_tmp_all_aa, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_colors = list(timepoint = col_timepoint), 
         breaks = scaleColors(data = tmp_hm_df_aa, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = tmp_hm_df_aa, 
                               maxvalue = 2)[["color"]], main = "Amino acids, peptides, and analogues, all")
```
#### ... combine plots by residence
```{r}
color_paths <- c("Fatty acids and conjugates" = "purple", 
                 "Amino acids, peptides, and analogues" = "orange2", 
                 "Carbohydrates and carbohydrate conjugates" = "seagreen")

color_paths <- ggsci::pal_uchicago()(3)
names(color_paths) <- c("Fatty acids and conjugates",
                 "Amino acids, peptides, and analogues" ,
                 "Carbohydrates and carbohydrate conjugates" )
show_col(color_paths)
```

#### ... !!! triangle plots
```{r fig.height=15, fig.width=5}
plotdf <- subset(res_dream_all, class_name %in% c("Fatty acids and conjugates", "Amino acids, peptides, and analogues", "Carbohydrates and carbohydrate conjugates", "Eicosanoids") & regulation != "ns")

plotdf$diet <- factor(plotdf$diet, levels = c("urban_diet","rural_diet","MBEGE"))

plotdf <- plotdf[order(plotdf$class_name),]

plotdf$comparison <- factor(plotdf$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

plotdf$y <- str_trunc(plotdf$CompoundName,18)

max.val <- round(max(abs(plotdf$log2FC2), na.rm = T),1)*1.2

ggplot(plotdf, aes(x= comparison, y= y))+
  geom_point(aes(shape = regulation, fill = log2FC2), size = 3)+
   scale_shape_manual(values = c("up" = 24, "down" = 25))+
  geom_text(aes(label = adj.P.Val.signif))+
  scale_fill_gradientn(colours = c("blue","white","red"), 
                       limits = c(-max.val, max.val),
                       guide = guide_colorbar(frame.colour = "black", 
                                              ticks.colour = "black"))+
  facet_grid(class_name~ diet, scales = "free", space = "free")+
  theme_bw()
```


#### ... >> urban diet
```{r fig.height=10, fig.width=15}
chosen_metabolites <- unique(subset(all_sig_df, class_name %in% c("Fatty acids and conjugates","Amino acids, peptides, and analogues","Carbohydrates and carbohydrate conjugates") & diet == "urban_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_urbandiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

log2_urbandiet_tmp <- metabo_log2_urbandiet[which(rownames(metabo_log2_urbandiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_urbandiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_urbandiet_tmp <- log2_urbandiet_tmp[,match(tmp_sample_table$ID, colnames(log2_urbandiet_tmp))]

log2_urbandiet_tmp_scale <- t(scale(t(log2_urbandiet_tmp)))

rownames(log2_urbandiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_urbandiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno) <- tmp_sample_table$ID

metabo_anno <- data.frame(class = all_sig_df[match(rownames(log2_urbandiet_tmp_scale), all_sig_df$CompoundName),"class_name"])
rownames(metabo_anno) <- rownames(log2_urbandiet_tmp_scale)

pheatmap(log2_urbandiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_row = metabo_anno, fontsize = 10, treeheight_row = 1,
         annotation_colors = list(timepoint = col_timepoint, class = color_paths), 
         breaks = scaleColors(data = log2_urbandiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_urbandiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, urban diet")
```
#### ... >> rural diet
```{r fig.height=10, fig.width=15}
chosen_metabolites <- unique(subset(all_sig_df, class_name %in% c("Fatty acids and conjugates","Amino acids, peptides, and analogues","Carbohydrates and carbohydrate conjugates") & diet == "rural_diet")$CompoundName)
  
#   unique(c("Betaine","L-Tyrosine","Ornithine",
# "L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
# "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Alpha-Linolenic acid","Eicosapentaenoic acid","Docosahexaenoic acid", labels_ruraldiet$metabolite ))

## find ionMz 
names(chosen_metabolites) <- melt_annotation_tal_new[match(chosen_metabolites, melt_annotation_tal_new$CompoundName),"ionMz"]

log2_ruraldiet_tmp <- metabo_log2_ruraldiet[which(rownames(metabo_log2_ruraldiet) %in% names(chosen_metabolites)),]

## organize columns
tmp_sample_table <- sample_table_ruraldiet
tmp_sample_table <- tmp_sample_table[order(tmp_sample_table$timepoint, tmp_sample_table$PID),]

log2_ruraldiet_tmp <- log2_ruraldiet_tmp[,match(tmp_sample_table$ID, colnames(log2_ruraldiet_tmp))]

log2_ruraldiet_tmp_scale <- t(scale(t(log2_ruraldiet_tmp)))

rownames(log2_ruraldiet_tmp_scale) <- melt_annotation_tal_new[match(rownames(log2_ruraldiet_tmp_scale),as.character( melt_annotation_tal_new$ionMz)),"CompoundName"]

tmp_anno <- tmp_sample_table[,c("timepoint","PID")]
rownames(tmp_anno) <- tmp_sample_table$ID

metabo_anno <- data.frame(class = all_sig_df[match(rownames(log2_ruraldiet_tmp_scale), all_sig_df$CompoundName),"class_name"])
rownames(metabo_anno) <- rownames(log2_ruraldiet_tmp_scale)

pheatmap(log2_ruraldiet_tmp_scale, cluster_cols = F, 
         annotation_col = tmp_anno, cellheight = 10,cellwidth = 2,  show_colnames = F,
         annotation_row = metabo_anno, fontsize = 10,
         annotation_colors = list(timepoint = col_timepoint, class = color_paths), 
         breaks = scaleColors(data = log2_ruraldiet_tmp_scale, 
                              maxvalue = 2)[["breaks"]],
           color = scaleColors(data = log2_ruraldiet_tmp_scale, 
                               maxvalue = 2)[["color"]], main = "Fatty acids and conjugates, rural_diet ")
```

## 7.6. RMAPDb pathway
### Retrieve Pathways From Input Analyte(s)
### ... Urban diet
### comparison + direction
```{r}
list_rampdb_urbandiet <- list()
for (i in names(list_sig_dream_urbandiet_diet_hmdb)){
  print(i)
  
  for (j in names(list_sig_dream_urbandiet_diet_hmdb[[i]])){
    print(j)
    
    myinput <- unlist(list_sig_dream_urbandiet_diet_hmdb[[i]][[j]])
    
     myinput <- paste0("hmdb:",myinput)
    print(myinput[1:5])
  
  list_rampdb_urbandiet[[i]][[j]] <- Getpathway_RAMPdb(hmdbinput =  myinput)
    
  }
   
  }


```
```{r}
count_rampdb_urbandiet <- data.frame()

for (i in names(list_rampdb_urbandiet)){
  print(i)
  
  for (j in names(list_rampdb_urbandiet[[i]])){
    
    print(j)
    
    myinput <- list_rampdb_urbandiet[[i]][[j]]
     
  pathcount <-  myinput %>% 
    group_by(pathwayId, pathwayName) %>% 
    summarize(count = n_distinct(inputId)) 
  
  myinput$CompoundName <- melt_annotation_tal_new[match(gsub("hmdb:","",myinput$inputId),melt_annotation_tal_new$CompoundID_new),"CompoundName"]
  
  hmdb_ids <- split(
    myinput$CompoundName,
    myinput$pathwayId
  )
  hmdb_ids <- lapply(hmdb_ids, paste, collapse = "/")
  
  hmdb_ids <- t(as.data.frame(bind_rows(hmdb_ids)))
  
  pathcount$compoundIds <-as.character(hmdb_ids[match(pathcount$pathwayId, rownames(hmdb_ids)),1])
  
  pathcount$comparison <- i
  
  pathcount$regulation <- j
  
  count_rampdb_urbandiet <- rbind(count_rampdb_urbandiet, pathcount)
    
  }
 
     
  }
 


head(count_rampdb_urbandiet)
```
```{r}
count_rampdb_urbandiet <- count_rampdb_urbandiet[order(count_rampdb_urbandiet$count, decreasing = T),]
# exclude
count_rampdb_urbandiet <- count_rampdb_urbandiet[which(!count_rampdb_urbandiet$pathwayId %in% c("R-HSA-1430728","WP3604","R-HSA-382551","R-HSA-1643685","WP3602")),]
head(count_rampdb_urbandiet)
```
```{r fig.height=6}
count_rampdb_urbandiet$comparison <- factor(count_rampdb_urbandiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(count_rampdb_urbandiet, count >3), aes(x = regulation, y= reorder(pathwayName, count), fill = regulation, size = count))+
  geom_point(shape = 21)+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ comparison, nrow = 1)+
  theme_bw()+ ggtitle("Urban diet")+
  theme(axis.text.y = element_text(size = 10))
```

### ... Rural diet
### comparison + direction
```{r}
list_rampdb_ruraldiet <- list()
for (i in names(list_sig_dream_ruraldiet_diet_hmdb)){
  print(i)
  
  for (j in names(list_sig_dream_ruraldiet_diet_hmdb[[i]])){
    print(j)
    
    myinput <- unlist(list_sig_dream_ruraldiet_diet_hmdb[[i]][[j]])
    
     myinput <- paste0("hmdb:",myinput)
    print(myinput[1:5])
  
  list_rampdb_ruraldiet[[i]][[j]] <- Getpathway_RAMPdb(hmdbinput =  myinput)
    
  }
   
  }


```
```{r}
count_rampdb_ruraldiet <- data.frame()

for (i in names(list_rampdb_ruraldiet)){
  print(i)
  
  for (j in names(list_rampdb_ruraldiet[[i]])){
    
    print(j)
    
    myinput <- list_rampdb_ruraldiet[[i]][[j]]
     
  pathcount <-  myinput %>% 
    group_by(pathwayId, pathwayName) %>% 
    summarize(count = n_distinct(inputId)) 
  
  myinput$CompoundName <- melt_annotation_tal_new[match(gsub("hmdb:","",myinput$inputId),melt_annotation_tal_new$CompoundID_new),"CompoundName"]
  
  hmdb_ids <- split(
    myinput$CompoundName,
    myinput$pathwayId
  )
  hmdb_ids <- lapply(hmdb_ids, paste, collapse = "/")
  
  hmdb_ids <- t(as.data.frame(bind_rows(hmdb_ids)))
  
  pathcount$compoundIds <-as.character(hmdb_ids[match(pathcount$pathwayId, rownames(hmdb_ids)),1])
  
  pathcount$comparison <- i
  
  pathcount$regulation <- j
  
  count_rampdb_ruraldiet <- rbind(count_rampdb_ruraldiet, pathcount)
    
  }
 
     
  }
 


head(count_rampdb_ruraldiet)
```
```{r}
count_rampdb_ruraldiet <- count_rampdb_ruraldiet[order(count_rampdb_ruraldiet$count, decreasing = T),]
# exclude
count_rampdb_ruraldiet <- count_rampdb_ruraldiet[which(!count_rampdb_ruraldiet$pathwayId %in% c("R-HSA-1430728","WP3604","R-HSA-382551","R-HSA-1643685","WP3602")),]
head(count_rampdb_ruraldiet)
```
```{r}
count_rampdb_ruraldiet$comparison <- factor(count_rampdb_ruraldiet$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(count_rampdb_ruraldiet, count >3), aes(x = regulation, y= reorder(pathwayName, count), fill = regulation, size = count))+
  geom_point(shape = 21)+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ comparison, nrow = 1)+
  theme_bw()+ ggtitle("Rural diet")+
  theme(axis.text.y = element_text(size = 10))
```

### ... Mbege
### comparison + direction
```{r}
list_rampdb_mbege <- list()
for (i in names(list_sig_dream_mbege_diet_hmdb)){
  print(i)
  
  for (j in names(list_sig_dream_mbege_diet_hmdb[[i]])){
    print(j)
    
    myinput <- unlist(list_sig_dream_mbege_diet_hmdb[[i]][[j]])
    
     myinput <- paste0("hmdb:",myinput)
    print(myinput[1:5])
  
  list_rampdb_mbege[[i]][[j]] <- Getpathway_RAMPdb(hmdbinput =  myinput)
    
  }
   
  }


```
```{r}
count_rampdb_mbege <- data.frame()

for (i in names(list_rampdb_mbege)){
  print(i)
  
  for (j in names(list_rampdb_mbege[[i]])){
    
    print(j)
    
    myinput <- list_rampdb_mbege[[i]][[j]]
     
  pathcount <-  myinput %>% 
    group_by(pathwayId, pathwayName) %>% 
    summarize(count = n_distinct(inputId)) 
  
  myinput$CompoundName <- melt_annotation_tal_new[match(gsub("hmdb:","",myinput$inputId),melt_annotation_tal_new$CompoundID_new),"CompoundName"]
  
  hmdb_ids <- split(
    myinput$CompoundName,
    myinput$pathwayId
  )
  hmdb_ids <- lapply(hmdb_ids, paste, collapse = "/")
  
  hmdb_ids <- t(as.data.frame(bind_rows(hmdb_ids)))
  
  pathcount$compoundIds <-as.character(hmdb_ids[match(pathcount$pathwayId, rownames(hmdb_ids)),1])
  
  pathcount$comparison <- i
  
  pathcount$regulation <- j
  
  count_rampdb_mbege <- rbind(count_rampdb_mbege, pathcount)
    
  }
 
     
  }
 


head(count_rampdb_mbege)
```
```{r}
count_rampdb_mbege <- count_rampdb_mbege[order(count_rampdb_mbege$count, decreasing = T),]
# exclude
count_rampdb_mbege <- count_rampdb_mbege[which(!count_rampdb_mbege$pathwayId %in% c("R-HSA-1430728","WP3604","R-HSA-382551","R-HSA-1643685","WP3602")),]
head(count_rampdb_mbege)
```
```{r}
count_rampdb_mbege$comparison <- factor(count_rampdb_mbege$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))

ggplot(subset(count_rampdb_mbege, count >=3), aes(x = regulation, y= reorder(pathwayName, count), fill = regulation, size = count))+
  geom_point(shape = 21)+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ comparison, nrow = 1)+
  theme_bw()+ ggtitle("Mbege diet")+
  theme(axis.text.y = element_text(size = 10))
```

### Rmpdb all diets
```{r}
count_rampdb_urbandiet$diet <- "urban_diet"
count_rampdb_ruraldiet$diet <- "rural_diet"
count_rampdb_mbege$diet <- "MBEGE"

count_rampdb_all <- rbind(count_rampdb_urbandiet, count_rampdb_ruraldiet, count_rampdb_mbege)

head(count_rampdb_all)
```


```{r fig.width=9}
count_rampdb_all$comparison <- factor(count_rampdb_all$comparison, levels = c("t0_vs_t1","t1_vs_t2","t0_vs_t2"))
count_rampdb_all$diet <- factor(count_rampdb_all$diet, levels = c("urban_diet","rural_diet","MBEGE"))

ggplot(subset(count_rampdb_all, count >=4 & comparison == "t0_vs_t1"), aes(x = regulation, y= reorder(pathwayName, count), fill = regulation, size = count))+
  geom_point(shape = 21)+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ diet, nrow = 1)+
  theme_bw()+ ggtitle("Mbege diet")+
  theme(axis.text.y = element_text(size = 10))+
  scale_size_area(max_size = 7)
```
```{r fig.width=12}
ggplot(subset(count_rampdb_all, count >=3), aes(x = comparison, y= reorder(pathwayName, count), fill = regulation, shape = regulation))+
  geom_point(size = 4)+
  scale_shape_manual(values = c("up" = 24, "down" = 25))+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ diet, nrow = 1)+
  theme_bw()+ ggtitle("t1 vs t0")+
  theme(axis.text.y = element_text(size = 10))+
  scale_size_area(max_size = 7)
```




```{r fig.width=12}
ggplot(subset(count_rampdb_all, count >=3 & comparison != "t1_vs_t2"), aes(x = comparison, y= reorder(pathwayName, count), fill = regulation, shape = regulation))+
  geom_point(size = 4)+
  scale_shape_manual(values = c("up" = 24, "down" = 25))+
  scale_fill_manual(values= c("up"= "tomato3", "down" = "cornflowerblue"))+
  facet_wrap(.~ diet, nrow = 1)+
  theme_bw()+ ggtitle("t1 vs t0 , t2 vs t0")+
  theme(axis.text.y = element_text(size = 10))+
  scale_size_area(max_size = 7)
```

### ... Violin plot selected
#### ... >> urbandiet
```{r}
selected_urbandiet <- unique(c("Betaine","L-Tyrosine","Ornithine",
"L-Isoleucine","L-Methionine","L-Arginine","Citrulline","Agmatine","L-Tryptophan",
 "Capric acid","Myristic acid", "Arachidonic acid","Leukotriene B4","Eicosapentaenoic acid","Docosahexaenoic acid"))
### check if all selected are sig

all(selected_urbandiet %in%  all_sig_df[which(all_sig_df$diet == "urban_diet" & all_sig_df$regulation != "ns"), ]$CompoundName)

```

# =======================================
# 8. List of sig metabolites
```{r}
all_sig_df
```

# ==========================================

# 9. Microbial metabolites
## ** Import conversions of hmdb of sig metabolites to other ids
```{r}
#write.table(unique(all_sig_df$HMDB),"./240913_metabolites_sig_all.txt",sep = "\t", quote = F, row.names = F)
```

```{r}
sig_metabolites_ids_sep2024 <- read_csv("./pathbank/240913_hmdb_map_otherids.csv")

head(sig_metabolites_ids_sep2024)
```
```{r}
sig_metabolites_ids_sep2024[is.na(sig_metabolites_ids_sep2024$HMDB),]
```

```{r}
list_id_hmdb_sig <- reshape2::melt(sig_metabolites_ids_sep2024[,c("Query","PubChem","")])
```

```{r}
keggs <- sig_metabolites_ids_sep2024$KEGG
keggs <- keggs[!is.na(keggs)]
hmdbs <- sig_metabolites_ids_sep2024$HMDB[!is.na(sig_metabolites_ids_sep2024$HMDB)]
chebis <- sig_metabolites_ids_sep2024$ChEBI[!is.na(sig_metabolites_ids_sep2024$ChEBI)]

mypaths <- pathbkank[which(pathbkank$HMDB_ID %in% hmdbs | pathbkank$KEGG_ID %in% keggs| pathbkank$ChEBI_ID %in% chebis),]

## filter disease, filter animals 
mypaths <- mypaths[which(mypaths$Pathway_Subject != "Disease"),]
 
mypaths <- mypaths[which(!mypaths$Species %in% c("Mus musculus","Caenorhabditis elegans","Rattus norvegicus","Drosophila melanogaster")),]

head(mypaths)
```
```{r fig.height=6}
mypaths_specie <- as.data.frame(mypaths %>% group_by(Metabolite_Name, Species) %>% summarize(count = n()))

mypaths_specie$Species <- factor(mypaths_specie$Species, levels = c("Homo sapiens","Arabidopsis thaliana", "Bos taurus","Escherichia coli","Pseudomonas aeruginosa","Saccharomyces cerevisiae"))

ggplot(mypaths_specie, aes(x= Species, y= Metabolite_Name))+
         geom_point(size = 3, shape = 15)+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 30, vjust =1), axis.title = element_blank())+
  theme(aspect.ratio = 1.5)
```

```{r}
mypaths$metboid_specie <- paste0(mypaths$Metabolite_ID,"_", mypaths$Species)

mypaths_wide <- as.data.frame(mypaths[,c()])
```

